{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "36f0b20b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Obtaining dependency information for biopython from https://files.pythonhosted.org/packages/2a/3c/10f22f3599acdf12b2a7a687589e3722b521147fe0e7846bd11f294eaf90/biopython-1.83-cp38-cp38-win_amd64.whl.metadata\n",
      "  Downloading biopython-1.83-cp38-cp38-win_amd64.whl.metadata (13 kB)\n",
      "Requirement already satisfied: numpy in c:\\users\\gabriel\\anaconda3\\lib\\site-packages (from biopython) (1.24.3)\n",
      "Downloading biopython-1.83-cp38-cp38-win_amd64.whl (2.7 MB)\n",
      "   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--\n",
      "   -- ------------------------------------- 0.1/2.7 MB 1.7 MB/s eta 0:00:02\n",
      "   ----- ---------------------------------- 0.4/2.7 MB 3.1 MB/s eta 0:00:01\n",
      "   -------- ------------------------------- 0.6/2.7 MB 3.4 MB/s eta 0:00:01\n",
      "   ----------- ---------------------------- 0.8/2.7 MB 3.6 MB/s eta 0:00:01\n",
      "   -------------- ------------------------- 1.0/2.7 MB 3.7 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.0/2.7 MB 3.9 MB/s eta 0:00:01\n",
      "   --------------- ------------------------ 1.1/2.7 MB 1.8 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 1.3/2.7 MB 2.0 MB/s eta 0:00:01\n",
      "   --------------------- ------------------ 1.5/2.7 MB 2.2 MB/s eta 0:00:01\n",
      "   ------------------------- -------------- 1.7/2.7 MB 2.4 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 2.0/2.7 MB 2.5 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 2.3/2.7 MB 2.7 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.6/2.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.7/2.7 MB 2.8 MB/s eta 0:00:00\n",
      "Installing collected packages: biopython\n",
      "Successfully installed biopython-1.83\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "#pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8d5ba4ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting wget\n",
      "  Downloading wget-3.2.zip (10 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Building wheels for collected packages: wget\n",
      "  Building wheel for wget (setup.py): started\n",
      "  Building wheel for wget (setup.py): finished with status 'done'\n",
      "  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9680 sha256=241fcbfb51aa35ba1e2f8babd09b178f61a9bb5623c2f77cfdd30d0b31cdb460\n",
      "  Stored in directory: c:\\users\\gabriel\\appdata\\local\\pip\\cache\\wheels\\bd\\a8\\c3\\3cf2c14a1837a4e04bd98631724e81f33f462d86a1d895fae0\n",
      "Successfully built wget\n",
      "Installing collected packages: wget\n",
      "Successfully installed wget-3.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "#comando para descargar los archivos FASTA introduciendo únicamente la url\n",
    "#pip install wget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "038150c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio:\n",
      "\n",
      "NAME\n",
      "    Bio - Collection of modules for dealing with biological data in Python.\n",
      "\n",
      "DESCRIPTION\n",
      "    The Biopython Project is an international association of developers\n",
      "    of freely available Python tools for computational molecular biology.\n",
      "    \n",
      "    https://biopython.org\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    Affy (package)\n",
      "    Align (package)\n",
      "    AlignIO (package)\n",
      "    Alphabet (package)\n",
      "    Application (package)\n",
      "    Blast (package)\n",
      "    CAPS (package)\n",
      "    Cluster (package)\n",
      "    Compass (package)\n",
      "    Data (package)\n",
      "    Emboss (package)\n",
      "    Entrez (package)\n",
      "    ExPASy (package)\n",
      "    File\n",
      "    GenBank (package)\n",
      "    Geo (package)\n",
      "    Graphics (package)\n",
      "    HMM (package)\n",
      "    KEGG (package)\n",
      "    LogisticRegression\n",
      "    MarkovModel\n",
      "    MaxEntropy\n",
      "    Medline (package)\n",
      "    NMR (package)\n",
      "    NaiveBayes\n",
      "    Nexus (package)\n",
      "    PDB (package)\n",
      "    Pathway (package)\n",
      "    Phylo (package)\n",
      "    PopGen (package)\n",
      "    Restriction (package)\n",
      "    SCOP (package)\n",
      "    SVDSuperimposer (package)\n",
      "    SearchIO (package)\n",
      "    Seq\n",
      "    SeqFeature\n",
      "    SeqIO (package)\n",
      "    SeqRecord\n",
      "    SeqUtils (package)\n",
      "    Sequencing (package)\n",
      "    SwissProt (package)\n",
      "    TogoWS (package)\n",
      "    UniGene (package)\n",
      "    UniProt (package)\n",
      "    _utils\n",
      "    bgzf\n",
      "    codonalign (package)\n",
      "    cpairwise2\n",
      "    kNN\n",
      "    motifs (package)\n",
      "    pairwise2\n",
      "    phenotype (package)\n",
      "\n",
      "CLASSES\n",
      "    builtins.Exception(builtins.BaseException)\n",
      "        MissingExternalDependencyError\n",
      "            MissingPythonDependencyError(MissingExternalDependencyError, builtins.ImportError)\n",
      "    builtins.ValueError(builtins.Exception)\n",
      "        StreamModeError\n",
      "    builtins.Warning(builtins.Exception)\n",
      "        BiopythonWarning\n",
      "            BiopythonDeprecationWarning\n",
      "            BiopythonExperimentalWarning\n",
      "            BiopythonParserWarning\n",
      "    \n",
      "    class BiopythonDeprecationWarning(BiopythonWarning)\n",
      "     |  Biopython deprecation warning.\n",
      "     |  \n",
      "     |  Biopython uses this warning instead of the built in DeprecationWarning\n",
      "     |  since those are ignored by default since Python 2.7.\n",
      "     |  \n",
      "     |  To silence all our deprecation warning messages, use:\n",
      "     |  \n",
      "     |  >>> import warnings\n",
      "     |  >>> from Bio import BiopythonDeprecationWarning\n",
      "     |  >>> warnings.simplefilter('ignore', BiopythonDeprecationWarning)\n",
      "     |  \n",
      "     |  Code marked as deprecated is likely to be removed in a future version\n",
      "     |  of Biopython. To avoid removal of this code, please contact the Biopython\n",
      "     |  developers via the mailing list or GitHub.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BiopythonDeprecationWarning\n",
      "     |      BiopythonWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors inherited from BiopythonWarning:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class BiopythonExperimentalWarning(BiopythonWarning)\n",
      "     |  Biopython experimental code warning.\n",
      "     |  \n",
      "     |  Biopython uses this warning for experimental code ('alpha' or 'beta'\n",
      "     |  level code) which is released as part of the standard releases to mark\n",
      "     |  sub-modules or functions for early adopters to test & give feedback.\n",
      "     |  \n",
      "     |  Code issuing this warning is likely to change (or even be removed) in\n",
      "     |  a subsequent release of Biopython. Such code should NOT be used for\n",
      "     |  production/stable code. It should only be used if:\n",
      "     |  \n",
      "     |  - You are running the latest release of Biopython, or ideally the\n",
      "     |    latest code from our repository.\n",
      "     |  - You are subscribed to the biopython-dev mailing list to provide\n",
      "     |    feedback on this code, and to be alerted of changes to it.\n",
      "     |  \n",
      "     |  If all goes well, experimental code would be promoted to stable in\n",
      "     |  a subsequent release, and this warning removed from it.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BiopythonExperimentalWarning\n",
      "     |      BiopythonWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors inherited from BiopythonWarning:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class BiopythonParserWarning(BiopythonWarning)\n",
      "     |  Biopython parser warning.\n",
      "     |  \n",
      "     |  Some in-valid data files cannot be parsed and will trigger an exception.\n",
      "     |  Where a reasonable interpretation is possible, Biopython will issue this\n",
      "     |  warning to indicate a potential problem. To silence these warnings, use:\n",
      "     |  \n",
      "     |  >>> import warnings\n",
      "     |  >>> from Bio import BiopythonParserWarning\n",
      "     |  >>> warnings.simplefilter('ignore', BiopythonParserWarning)\n",
      "     |  \n",
      "     |  Consult the warnings module documentation for more details.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BiopythonParserWarning\n",
      "     |      BiopythonWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors inherited from BiopythonWarning:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class BiopythonWarning(builtins.Warning)\n",
      "     |  Biopython warning.\n",
      "     |  \n",
      "     |  Biopython should use this warning (or subclasses of it), making it easy to\n",
      "     |  silence all our warning messages should you wish to:\n",
      "     |  \n",
      "     |  >>> import warnings\n",
      "     |  >>> from Bio import BiopythonWarning\n",
      "     |  >>> warnings.simplefilter('ignore', BiopythonWarning)\n",
      "     |  \n",
      "     |  Consult the warnings module documentation for more details.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BiopythonWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Warning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class MissingExternalDependencyError(builtins.Exception)\n",
      "     |  Missing an external dependency.\n",
      "     |  \n",
      "     |  Used for things like missing command line tools. Important for our unit\n",
      "     |  tests to allow skipping tests with missing external dependencies.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      MissingExternalDependencyError\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.Exception:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Exception:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class MissingPythonDependencyError(MissingExternalDependencyError, builtins.ImportError)\n",
      "     |  Missing an external python dependency (subclass of ImportError).\n",
      "     |  \n",
      "     |  Used for missing Python modules (rather than just a typical ImportError).\n",
      "     |  Important for our unit tests to allow skipping tests with missing external\n",
      "     |  python dependencies, while also allowing the exception to be caught as an\n",
      "     |  ImportError.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      MissingPythonDependencyError\n",
      "     |      MissingExternalDependencyError\n",
      "     |      builtins.ImportError\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.ImportError:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.ImportError:\n",
      "     |  \n",
      "     |  msg\n",
      "     |      exception message\n",
      "     |  \n",
      "     |  name\n",
      "     |      module name\n",
      "     |  \n",
      "     |  path\n",
      "     |      module path\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.Exception:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class StreamModeError(builtins.ValueError)\n",
      "     |  Incorrect stream mode (text vs binary).\n",
      "     |  \n",
      "     |  This error should be raised when a stream (file or file-like object)\n",
      "     |  argument is in text mode while the receiving function expects binary mode,\n",
      "     |  or vice versa.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      StreamModeError\n",
      "     |      builtins.ValueError\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.ValueError:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.ValueError:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "\n",
      "VERSION\n",
      "    1.83\n",
      "\n",
      "FILE\n",
      "    c:\\users\\gabriel\\anaconda3\\lib\\site-packages\\bio\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "help(Bio)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab8bcbb4",
   "metadata": {},
   "source": [
    "## Lectura de archivos FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c56a01a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6eacf4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module wget:\n",
      "\n",
      "NAME\n",
      "    wget - Download utility as an easy way to get file from the net\n",
      "\n",
      "DESCRIPTION\n",
      "      python -m wget <URL>\n",
      "      python wget.py <URL>\n",
      "    \n",
      "    Downloads: http://pypi.python.org/pypi/wget/\n",
      "    Development: http://bitbucket.org/techtonik/python-wget/\n",
      "    \n",
      "    wget.py is not option compatible with Unix wget utility,\n",
      "    to make command line interface intuitive for new people.\n",
      "    \n",
      "    Public domain by anatoly techtonik <techtonik@gmail.com>\n",
      "    Also available under the terms of MIT license\n",
      "    Copyright (c) 2010-2015 anatoly techtonik\n",
      "\n",
      "FUNCTIONS\n",
      "    bar_adaptive(current, total, width=80)\n",
      "        Return progress bar string for given values in one of three\n",
      "        styles depending on available width:\n",
      "        \n",
      "            [..  ] downloaded / total\n",
      "            downloaded / total\n",
      "            [.. ]\n",
      "        \n",
      "        if total value is unknown or <= 0, show bytes counter using two\n",
      "        adaptive styles:\n",
      "        \n",
      "            %s / unknown\n",
      "            %s\n",
      "        \n",
      "        if there is not enough space on the screen, do not display anything\n",
      "        \n",
      "        returned string doesn't include control characters like \r",
      " used to\n",
      "        place cursor at the beginning of the line to erase previous content.\n",
      "        \n",
      "        this function leaves one free character at the end of string to\n",
      "        avoid automatic linefeed on Windows.\n",
      "    \n",
      "    bar_thermometer(current, total, width=80)\n",
      "        Return thermometer style progress bar string. `total` argument\n",
      "        can not be zero. The minimum size of bar returned is 3. Example:\n",
      "        \n",
      "            [..........            ]\n",
      "        \n",
      "        Control and trailing symbols (\r",
      " and spaces) are not included.\n",
      "        See `bar_adaptive` for more information.\n",
      "    \n",
      "    callback_progress(blocks, block_size, total_size, bar_function)\n",
      "        callback function for urlretrieve that is called when connection is\n",
      "        created and when once for each block\n",
      "        \n",
      "        draws adaptive progress bar in terminal/console\n",
      "        \n",
      "        use sys.stdout.write() instead of \"print,\", because it allows one more\n",
      "        symbol at the line end without linefeed on Windows\n",
      "        \n",
      "        :param blocks: number of blocks transferred so far\n",
      "        :param block_size: in bytes\n",
      "        :param total_size: in bytes, can be -1 if server doesn't return it\n",
      "        :param bar_function: another callback function to visualize progress\n",
      "    \n",
      "    detect_filename(url=None, out=None, headers=None, default='download.wget')\n",
      "        Return filename for saving file. If no filename is detected from output\n",
      "        argument, url or headers, return default (download.wget)\n",
      "    \n",
      "    download(url, out=None, bar=<function bar_adaptive at 0x0000029AEE628CA0>)\n",
      "        High level function, which downloads URL into tmp file in current\n",
      "        directory and then renames it to filename autodetected from either URL\n",
      "        or HTTP headers.\n",
      "        \n",
      "        :param bar: function to track download progress (visualize etc.)\n",
      "        :param out: output filename or directory\n",
      "        :return:    filename where URL is downloaded to\n",
      "    \n",
      "    filename_fix_existing(filename)\n",
      "        Expands name portion of filename with numeric ' (x)' suffix to\n",
      "        return filename that doesn't exist already.\n",
      "    \n",
      "    filename_from_headers(headers)\n",
      "        Detect filename from Content-Disposition headers if present.\n",
      "        http://greenbytes.de/tech/tc2231/\n",
      "        \n",
      "        :param: headers as dict, list or string\n",
      "        :return: filename from content-disposition header or None\n",
      "    \n",
      "    filename_from_url(url)\n",
      "        :return: detected filename as unicode or None\n",
      "    \n",
      "    get_console_width()\n",
      "        Return width of available window area. Autodetection works for\n",
      "        Windows and POSIX platforms. Returns 80 for others\n",
      "        \n",
      "        Code from http://bitbucket.org/techtonik/python-pager\n",
      "    \n",
      "    to_unicode(filename)\n",
      "        :return: filename decoded from utf-8 to unicode\n",
      "    \n",
      "    win32_unicode_console()\n",
      "        # enable unicode output to windows console\n",
      "        # https://stackoverflow.com/questions/878972/windows-cmd-encoding-change-causes-python-crash\n",
      "    \n",
      "    win32_utf8_argv()\n",
      "        Uses shell32.GetCommandLineArgvW to get sys.argv as a list of Unicode\n",
      "        strings.\n",
      "        \n",
      "        Versions 2.x of Python don't support Unicode in sys.argv on\n",
      "        Windows, with the underlying Windows API instead replacing multi-byte\n",
      "        characters with '?'.\n",
      "\n",
      "DATA\n",
      "    PY3K = True\n",
      "    usage = 'usage: wget.py [options] URL\\n\\noptions:\\n  -o --ou...ut file...\n",
      "\n",
      "VERSION\n",
      "    3.2\n",
      "\n",
      "FILE\n",
      "    c:\\users\\gabriel\\anaconda3\\lib\\site-packages\\wget.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import wget\n",
    "help(wget)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "037085a9",
   "metadata": {},
   "source": [
    "La función wget se utiliza para descargar archivos desde la web en sistemas Unix/Linux a través de la línea de comandos. Si estás trabajando en Python y necesitas descargar archivos desde la web, puedes utilizar el módulo urllib.request que viene con Python de forma predeterminada.\n",
    "\n",
    "Aquí tienes un ejemplo básico de cómo descargar un archivo utilizando urllib.request:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8a501ad9",
   "metadata": {},
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "HTTP Error 404: Not Found",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 6\u001b[0m\n\u001b[0;32m      3\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttps://example.com/archivo.txt\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      4\u001b[0m destino_local \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124marchivo_local.txt\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m----> 6\u001b[0m \u001b[43murllib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlretrieve\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdestino_local\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:247\u001b[0m, in \u001b[0;36murlretrieve\u001b[1;34m(url, filename, reporthook, data)\u001b[0m\n\u001b[0;32m    230\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    231\u001b[0m \u001b[38;5;124;03mRetrieve a URL into a temporary location on disk.\u001b[39;00m\n\u001b[0;32m    232\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    243\u001b[0m \u001b[38;5;124;03mdata file as well as the resulting HTTPMessage object.\u001b[39;00m\n\u001b[0;32m    244\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    245\u001b[0m url_type, path \u001b[38;5;241m=\u001b[39m _splittype(url)\n\u001b[1;32m--> 247\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m contextlib\u001b[38;5;241m.\u001b[39mclosing(\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m) \u001b[38;5;28;01mas\u001b[39;00m fp:\n\u001b[0;32m    248\u001b[0m     headers \u001b[38;5;241m=\u001b[39m fp\u001b[38;5;241m.\u001b[39minfo()\n\u001b[0;32m    250\u001b[0m     \u001b[38;5;66;03m# Just return the local path and the \"headers\" for file://\u001b[39;00m\n\u001b[0;32m    251\u001b[0m     \u001b[38;5;66;03m# URLs. No sense in performing a copy unless requested.\u001b[39;00m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:222\u001b[0m, in \u001b[0;36murlopen\u001b[1;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[0;32m    220\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    221\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[1;32m--> 222\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:531\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    529\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m processor \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_response\u001b[38;5;241m.\u001b[39mget(protocol, []):\n\u001b[0;32m    530\u001b[0m     meth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(processor, meth_name)\n\u001b[1;32m--> 531\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mmeth\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    533\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:640\u001b[0m, in \u001b[0;36mHTTPErrorProcessor.http_response\u001b[1;34m(self, request, response)\u001b[0m\n\u001b[0;32m    637\u001b[0m \u001b[38;5;66;03m# According to RFC 2616, \"2xx\" code indicates that the client's\u001b[39;00m\n\u001b[0;32m    638\u001b[0m \u001b[38;5;66;03m# request was successfully received, understood, and accepted.\u001b[39;00m\n\u001b[0;32m    639\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;241m200\u001b[39m \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m code \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m300\u001b[39m):\n\u001b[1;32m--> 640\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparent\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43merror\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    641\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhttp\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhdrs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    643\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:569\u001b[0m, in \u001b[0;36mOpenerDirector.error\u001b[1;34m(self, proto, *args)\u001b[0m\n\u001b[0;32m    567\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m http_err:\n\u001b[0;32m    568\u001b[0m     args \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mdict\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefault\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttp_error_default\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;241m+\u001b[39m orig_args\n\u001b[1;32m--> 569\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:502\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    500\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    501\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 502\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    503\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    504\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\urllib\\request.py:649\u001b[0m, in \u001b[0;36mHTTPDefaultErrorHandler.http_error_default\u001b[1;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[0;32m    648\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttp_error_default\u001b[39m(\u001b[38;5;28mself\u001b[39m, req, fp, code, msg, hdrs):\n\u001b[1;32m--> 649\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m HTTPError(req\u001b[38;5;241m.\u001b[39mfull_url, code, msg, hdrs, fp)\n",
      "\u001b[1;31mHTTPError\u001b[0m: HTTP Error 404: Not Found"
     ]
    }
   ],
   "source": [
    "import urllib.request\n",
    "\n",
    "url = 'https://example.com/archivo.txt'\n",
    "destino_local = 'archivo_local.txt'\n",
    "\n",
    "urllib.request.urlretrieve(url, destino_local)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cdc251a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function parse in module Bio.SeqIO:\n",
      "\n",
      "parse(handle, format, alphabet=None)\n",
      "    Turn a sequence file into an iterator returning SeqRecords.\n",
      "    \n",
      "    Arguments:\n",
      "     - handle   - handle to the file, or the filename as a string\n",
      "       (note older versions of Biopython only took a handle).\n",
      "     - format   - lower case string describing the file format.\n",
      "     - alphabet - no longer used, should be None.\n",
      "    \n",
      "    Typical usage, opening a file to read in, and looping over the record(s):\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> filename = \"Fasta/sweetpea.nu\"\n",
      "    >>> for record in SeqIO.parse(filename, \"fasta\"):\n",
      "    ...    print(\"ID %s\" % record.id)\n",
      "    ...    print(\"Sequence length %i\" % len(record))\n",
      "    ID gi|3176602|gb|U78617.1|LOU78617\n",
      "    Sequence length 309\n",
      "    \n",
      "    For lazy-loading file formats such as twobit, for which the file contents\n",
      "    is read on demand only, ensure that the file remains open while extracting\n",
      "    sequence data.\n",
      "    \n",
      "    If you have a string 'data' containing the file contents, you must\n",
      "    first turn this into a handle in order to parse it:\n",
      "    \n",
      "    >>> data = \">Alpha\\nACCGGATGTA\\n>Beta\\nAGGCTCGGTTA\\n\"\n",
      "    >>> from Bio import SeqIO\n",
      "    >>> from io import StringIO\n",
      "    >>> for record in SeqIO.parse(StringIO(data), \"fasta\"):\n",
      "    ...     print(\"%s %s\" % (record.id, record.seq))\n",
      "    Alpha ACCGGATGTA\n",
      "    Beta AGGCTCGGTTA\n",
      "    \n",
      "    Use the Bio.SeqIO.read(...) function when you expect a single record\n",
      "    only.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(SeqIO.parse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d462505",
   "metadata": {},
   "source": [
    "En caso de trabajar con datos de grandes dimensiones:\n",
    "\n",
    "1-Debemos descagar los datos (utilizar urllib.request)\n",
    "\n",
    "2- Ya con ello, deberíamos descomprimirlos para poder trabajar con ellos (utilizar paquete gzip)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a04b65a1",
   "metadata": {},
   "source": [
    "**SimpleFastaParser vs SeqIO.parse()**\n",
    "En caso de que el tiempo de ejecución sea importante( o manejando grandes datos), se deberá utilizar antes el método de SimpleFastaParser \n",
    "(from Bio.SeqIO.FastaIO import SimpleFastaParser) que SeqIO.parse (from Bio import SeqIO), ya que este primero \n",
    "tarda un menor tiempo en leer los archivos pasados"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9db5d373",
   "metadata": {},
   "source": [
    "## Calculo del contenido de GC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "686ab839",
   "metadata": {},
   "source": [
    "El contenido en G+C del DNA es la proporción de citosinas y guaninas respecto al total de nucleótidos en el genoma. Es un carácter ampliamente utilizado en las descripciones taxonómicas de microorganismos procariotas.\n",
    "\n",
    "El contenido GC se utiliza a veces para clasificar organismos en taxonomía. Por ejemplo, las Actinobacteria se caracterizan por ser “bacterias de GC alto”. En Streptomyces coelicolor el GC es del 72%, en la levadura Saccharomyces cerevisiae del 38%, mientras que en el organismo modelo Arabidopsis thaliana es del 36%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "157b373b",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'gc_content.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[5], line 5\u001b[0m\n\u001b[0;32m      3\u001b[0m max_sequence_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m      4\u001b[0m max_gc_content \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m seq_record \u001b[38;5;129;01min\u001b[39;00m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgc_content.fasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m:\n\u001b[0;32m      6\u001b[0m     sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(seq_record\u001b[38;5;241m.\u001b[39mseq) \u001b[38;5;66;03m#hay que convertir la secuencia a string\u001b[39;00m\n\u001b[0;32m      7\u001b[0m     sequence_id \u001b[38;5;241m=\u001b[39m seq_record\u001b[38;5;241m.\u001b[39mid\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\Bio\\SeqIO\\__init__.py:613\u001b[0m, in \u001b[0;36mparse\u001b[1;34m(handle, format, alphabet)\u001b[0m\n\u001b[0;32m    611\u001b[0m iterator_generator \u001b[38;5;241m=\u001b[39m _FormatToIterator\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mformat\u001b[39m)\n\u001b[0;32m    612\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m iterator_generator:\n\u001b[1;32m--> 613\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43miterator_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    614\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mformat\u001b[39m \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39m_FormatToIterator:\n\u001b[0;32m    615\u001b[0m     \u001b[38;5;66;03m# Use Bio.AlignIO to read in the alignments\u001b[39;00m\n\u001b[0;32m    616\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m (r \u001b[38;5;28;01mfor\u001b[39;00m alignment \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39mparse(handle, \u001b[38;5;28mformat\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m alignment)\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\Bio\\SeqIO\\FastaIO.py:190\u001b[0m, in \u001b[0;36mFastaIterator.__init__\u001b[1;34m(self, source, alphabet)\u001b[0m\n\u001b[0;32m    188\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m alphabet \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    189\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe alphabet argument is no longer supported\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m--> 190\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfmt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mFasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\Bio\\SeqIO\\Interfaces.py:58\u001b[0m, in \u001b[0;36mSequenceIterator.__init__\u001b[1;34m(self, source, alphabet, mode, fmt)\u001b[0m\n\u001b[0;32m     56\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe alphabet argument is no longer supported\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     57\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(source, _PathLikeTypes):\n\u001b[1;32m---> 58\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     59\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshould_close_stream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m     60\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'gc_content.fasta'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "max_sequence_id = None\n",
    "max_gc_content = 0\n",
    "for seq_record in SeqIO.parse(\"gc_content.fasta\",\"fasta\"):\n",
    "    sequence = str(seq_record.seq) #hay que convertir la secuencia a string\n",
    "    sequence_id = seq_record.id\n",
    "    gc_content = (sequence.count(\"C\")+ sequence.count(\"G\")) / len(sequence) * 100\n",
    "    if gc_content > max_gc_content:\n",
    "        max_sequence_id=sequence_id\n",
    "        max_gc_content=gc_content\n",
    "        \n",
    "print(max_sequence_id)\n",
    "print(max_gc_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00853ce2",
   "metadata": {},
   "source": [
    "## Objeto SEQ: DNA->RNA->PROTEINA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be866236",
   "metadata": {},
   "source": [
    "Es el dogma central de la biología molecular: ADN->Transcripción->ARN->Traducción->Proteína.\n",
    "\n",
    "LAS OPERACIONES DE TRANSCRIPCIÓN, TRADUCCIÓN Y REPLICACIÓN ESTÁN DISPONIBLES EN EL PAQUETE DE BIOPYTHON."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b388ffa2",
   "metadata": {},
   "source": [
    "**objeto seq**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97b6c117",
   "metadata": {},
   "source": [
    "El objeto seq en biopython es la cadena de caracteres de python con métodos biológicos\n",
    "\n",
    "OJO: porque permite más letras que solo ATCG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "90468961",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTAGTCGACTGAT\n",
      "<class 'Bio.Seq.Seq'>\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "secuencia= Seq(\"ACTAGTCGACTGAT\")\n",
    "print(secuencia)\n",
    "print(type(secuencia))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "da0db231",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class Seq in module Bio.Seq:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data: Union[str, bytes, bytearray, Bio.Seq._SeqAbstractBaseClass, Bio.Seq.SequenceDataAbstractBaseClass, dict, NoneType], length: Union[int, NoneType] = None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data: Union[str, bytes, bytearray, Bio.Seq._SeqAbstractBaseClass, Bio.Seq.SequenceDataAbstractBaseClass, dict, NoneType], length: Union[int, NoneType] = None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None or a dictionary (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_sequence = Seq(None, 20)\n",
      " |      >>> my_undefined_sequence\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_sequence)\n",
      " |      20\n",
      " |      >>> print(my_undefined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |      \n",
      " |      If the sequence contents is known for parts of the sequence only, use\n",
      " |      a dictionary for the data argument to pass the known sequence segments:\n",
      " |      \n",
      " |      >>> my_partially_defined_sequence = Seq({3: \"ACGT\"}, 10)\n",
      " |      >>> my_partially_defined_sequence\n",
      " |      Seq({3: 'ACGT'}, length=10)\n",
      " |      >>> len(my_partially_defined_sequence)\n",
      " |      10\n",
      " |      >>> print(my_partially_defined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined\n",
      " |      >>> my_partially_defined_sequence[3:7]\n",
      " |      Seq('ACGT')\n",
      " |      >>> print(my_partially_defined_sequence[3:7])\n",
      " |      ACGT\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  __annotations__ = {'_data': typing.Union[bytes, Bio.Seq.SequenceDataAb...\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Return an iterable of the sequence.\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement(self, inplace=False)\n",
      " |      Return the complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement()\n",
      " |      Seq('GCTAA')\n",
      " |      \n",
      " |      In contrast, ``complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement()\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement(inplace=True)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 while ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  islower(self)\n",
      " |      Return True if all ASCII characters in data are lowercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  isupper(self)\n",
      " |      Return True if all ASCII characters in data are uppercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  removeprefix(self, prefix, inplace=False)\n",
      " |      Return a new Seq object with prefix (left) removed.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      e.g. Removing a start Codon:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"ATGGTGTGTGT\")\n",
      " |      >>> my_seq\n",
      " |      Seq('ATGGTGTGTGT')\n",
      " |      >>> my_seq.removeprefix('ATG')\n",
      " |      Seq('GTGTGTGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``removeprefix`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the removesuffix method.\n",
      " |  \n",
      " |  removesuffix(self, suffix, inplace=False)\n",
      " |      Return a new Seq object with suffix (right) removed.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      e.g. Removing a stop codon:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"GTGTGTGTTAG\")\n",
      " |      >>> my_seq\n",
      " |      Seq('GTGTGTGTTAG')\n",
      " |      >>> stop_codon = Seq(\"TAG\")\n",
      " |      >>> my_seq.removesuffix(stop_codon)\n",
      " |      Seq('GTGTGTGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``removesuffix`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the removeprefix method.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement(self, inplace=False)\n",
      " |      Return the reverse complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement()\n",
      " |      Seq('AATCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement()\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement(inplace=True)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('TCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  search(self, subs)\n",
      " |      Search the substrings subs in self and yield the index and substring found.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - subs - a list of strings, Seq, MutableSeq, bytes, or bytearray\n",
      " |         objects containing the substrings to search for.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> dna = Seq(\"GTCATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAGTTG\")\n",
      " |      >>> matches = dna.search([\"CC\", Seq(\"ATTG\"), \"ATTG\", Seq(\"CCC\")])\n",
      " |      >>> for index, substring in matches:\n",
      " |      ...     print(index, substring)\n",
      " |      ...\n",
      " |      7 CC\n",
      " |      9 ATTG\n",
      " |      20 CC\n",
      " |      34 CC\n",
      " |      34 CCC\n",
      " |      35 CC\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      Following the usual convention, the sequence is interpreted as the\n",
      " |      coding strand of the DNA double helix, not the template strand. This\n",
      " |      means we can get the RNA sequence just by switching T to U.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  defined\n",
      " |      Return True if the sequence is defined, False if undefined or partially defined.\n",
      " |      \n",
      " |      Zero-length sequences are always considered to be defined.\n",
      " |  \n",
      " |  defined_ranges\n",
      " |      Return a tuple of the ranges where the sequence contents is defined.\n",
      " |      \n",
      " |      The return value has the format ((start1, end1), (start2, end2), ...).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __array_ufunc__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb61828",
   "metadata": {},
   "source": [
    "**algunos métodos relacionados**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "05e65269",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "#count()\n",
    "#método similar al presentado por las cadenas de carácteres\n",
    "secuencia= Seq(\"AAAA\")\n",
    "print(secuencia.count(\"AA\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6c48f44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function count_overlap in module Bio.Seq:\n",
      "\n",
      "count_overlap(self, sub, start=None, end=None)\n",
      "    Return an overlapping count.\n",
      "    \n",
      "    Returns an integer, the number of occurrences of substring\n",
      "    argument sub in the (sub)sequence given by [start:end].\n",
      "    Optional arguments start and end are interpreted as in slice\n",
      "    notation.\n",
      "    \n",
      "    Arguments:\n",
      "     - sub - a string or another Seq object to look for\n",
      "     - start - optional integer, slice start\n",
      "     - end - optional integer, slice end\n",
      "    \n",
      "    e.g.\n",
      "    \n",
      "    >>> from Bio.Seq import Seq\n",
      "    >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      "    3\n",
      "    >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      "    4\n",
      "    >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      "    1\n",
      "    \n",
      "    For a non-overlapping search, use the ``count`` method:\n",
      "    \n",
      "    >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      "    2\n",
      "    \n",
      "    Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      "    the ``count`` method:\n",
      "    \n",
      "    >>> from Bio.Seq import Seq\n",
      "    >>> my_seq = Seq(\"AAAATGA\")\n",
      "    >>> print(my_seq.count_overlap(\"A\"))\n",
      "    5\n",
      "    >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      "    True\n",
      "    >>> print(my_seq.count_overlap(\"ATG\"))\n",
      "    1\n",
      "    >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      "    True\n",
      "    >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      "    1\n",
      "    >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      "    True\n",
      "    >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      "    1\n",
      "    >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      "    True\n",
      "    \n",
      "    HOWEVER, do not use this method for such cases because the\n",
      "    count() method is much for efficient.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Seq.count_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4961290d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "secuencia.count_overlap(\"AA\")\n",
    "#que sería como: AAAA, y de ahí, buscar todos los AA posibles, AA-AA,A-AA-A Y AA-AA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bb5b6890",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class Seq in module Bio.Seq:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data: Union[str, bytes, bytearray, Bio.Seq._SeqAbstractBaseClass, Bio.Seq.SequenceDataAbstractBaseClass, dict, NoneType], length: Union[int, NoneType] = None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data: Union[str, bytes, bytearray, Bio.Seq._SeqAbstractBaseClass, Bio.Seq.SequenceDataAbstractBaseClass, dict, NoneType], length: Union[int, NoneType] = None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None or a dictionary (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_sequence = Seq(None, 20)\n",
      " |      >>> my_undefined_sequence\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_sequence)\n",
      " |      20\n",
      " |      >>> print(my_undefined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |      \n",
      " |      If the sequence contents is known for parts of the sequence only, use\n",
      " |      a dictionary for the data argument to pass the known sequence segments:\n",
      " |      \n",
      " |      >>> my_partially_defined_sequence = Seq({3: \"ACGT\"}, 10)\n",
      " |      >>> my_partially_defined_sequence\n",
      " |      Seq({3: 'ACGT'}, length=10)\n",
      " |      >>> len(my_partially_defined_sequence)\n",
      " |      10\n",
      " |      >>> print(my_partially_defined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined\n",
      " |      >>> my_partially_defined_sequence[3:7]\n",
      " |      Seq('ACGT')\n",
      " |      >>> print(my_partially_defined_sequence[3:7])\n",
      " |      ACGT\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  __annotations__ = {'_data': typing.Union[bytes, Bio.Seq.SequenceDataAb...\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Return an iterable of the sequence.\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement(self, inplace=False)\n",
      " |      Return the complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement()\n",
      " |      Seq('GCTAA')\n",
      " |      \n",
      " |      In contrast, ``complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement()\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement(inplace=True)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 while ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  islower(self)\n",
      " |      Return True if all ASCII characters in data are lowercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  isupper(self)\n",
      " |      Return True if all ASCII characters in data are uppercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  removeprefix(self, prefix, inplace=False)\n",
      " |      Return a new Seq object with prefix (left) removed.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      e.g. Removing a start Codon:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"ATGGTGTGTGT\")\n",
      " |      >>> my_seq\n",
      " |      Seq('ATGGTGTGTGT')\n",
      " |      >>> my_seq.removeprefix('ATG')\n",
      " |      Seq('GTGTGTGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``removeprefix`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the removesuffix method.\n",
      " |  \n",
      " |  removesuffix(self, suffix, inplace=False)\n",
      " |      Return a new Seq object with suffix (right) removed.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      e.g. Removing a stop codon:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"GTGTGTGTTAG\")\n",
      " |      >>> my_seq\n",
      " |      Seq('GTGTGTGTTAG')\n",
      " |      >>> stop_codon = Seq(\"TAG\")\n",
      " |      >>> my_seq.removesuffix(stop_codon)\n",
      " |      Seq('GTGTGTGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``removesuffix`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the removeprefix method.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement(self, inplace=False)\n",
      " |      Return the reverse complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement()\n",
      " |      Seq('AATCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement()\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement(inplace=True)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('TCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |      \n",
      " |      See the ``search`` method to find the locations of multiple subsequences\n",
      " |      at the same time.\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  search(self, subs)\n",
      " |      Search the substrings subs in self and yield the index and substring found.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - subs - a list of strings, Seq, MutableSeq, bytes, or bytearray\n",
      " |         objects containing the substrings to search for.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> dna = Seq(\"GTCATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAGTTG\")\n",
      " |      >>> matches = dna.search([\"CC\", Seq(\"ATTG\"), \"ATTG\", Seq(\"CCC\")])\n",
      " |      >>> for index, substring in matches:\n",
      " |      ...     print(index, substring)\n",
      " |      ...\n",
      " |      7 CC\n",
      " |      9 ATTG\n",
      " |      20 CC\n",
      " |      34 CC\n",
      " |      34 CCC\n",
      " |      35 CC\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      Following the usual convention, the sequence is interpreted as the\n",
      " |      coding strand of the DNA double helix, not the template strand. This\n",
      " |      means we can get the RNA sequence just by switching T to U.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  defined\n",
      " |      Return True if the sequence is defined, False if undefined or partially defined.\n",
      " |      \n",
      " |      Zero-length sequences are always considered to be defined.\n",
      " |  \n",
      " |  defined_ranges\n",
      " |      Return a tuple of the ranges where the sequence contents is defined.\n",
      " |      \n",
      " |      The return value has the format ((start1, end1), (start2, end2), ...).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __array_ufunc__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6c98c8f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34.78260869565217"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Para calcular el contenido de GC (en esta versión parece no haber el método GC)\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "secuencia = Seq(\"GACATATATAGCTAGCAAAGCTA\")\n",
    "gc_fraction(secuencia)*100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82b0a82a",
   "metadata": {},
   "source": [
    "**Transcripción**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453e89d6",
   "metadata": {},
   "source": [
    "recordemos que es el paso de DNA a RNA, si se trascribe la cadena 3'->5' entonces obtendremos la cadena 5'->3' como ARN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c3067235",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TAGCTAAACACAACAGTTACGAATCGATCGAAGTTGATCGATCAT')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_codificante= Seq(\"ATGATCGATCAACTTCGATCGATTCGTAACTGTTGTGTTTAGCTA\")\n",
    "dna_plantilla = dna_codificante.reverse_complement()\n",
    "dna_plantilla #generalmente la secuencia de nucleótidos suele ser leída en dirección 5' a 3' \n",
    "# mientras que la cadena molde es mostrada la reversa, aunq en sentido 5' a 3' en vez de como suele verse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "12cdaafb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGAUCGAUCAACUUCGAUCGAUUCGUAACUGUUGUGUUUAGCUA')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#transcripción\n",
    "# al final solo cambiamos las T por U, al transcribir la molde obtenemos la cadena de RNA de dna_codificante\n",
    "rna_mensajero=dna_codificante.transcribe()\n",
    "rna_mensajero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6b322348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCGATCAACTTCGATCGATTCGTAACTGTTGTGTTTAGCTA')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# el objeto aeq tambíen incluye un método de retro-transcripción para que, partiendo de mRNA obtengamos la cadena \n",
    "#correspondiente de DNA\n",
    "dna_codificante = rna_mensajero.back_transcribe()\n",
    "dna_codificante"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d84ca193",
   "metadata": {},
   "source": [
    "**Traducción**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "643fd788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIDQLRSIRNCCV*L')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#primera posibilidad, a partir del mRNA\n",
    "rna_mensajero.translate()\n",
    "#para este caso * == codón de STOP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6074141a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIDQLRSIRNCCV*L')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#segunda posibilidad, directamente con el DNA codificante (OJO= CODIFICANTE)\n",
    "#aquí se mantiene que * == codón de STOP\n",
    "dna_codificante.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb79252",
   "metadata": {},
   "source": [
    "## FASTQ y comprobación de la calidad de las lecturas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8a4922c",
   "metadata": {},
   "source": [
    "\n",
    "Fastq: formato para secuencias de secuenciación de siguiente generación\n",
    "Secuencia fasta + detalles calidad de la información (la Q es de Quality).\n",
    "\n",
    "Línea 1: Encabezado (Header): comienza con @. Sigue el identificador (identifier). Si son datos crudos contiene info del secuenciador que identifica a esta secuencia y el read pair (/1 o /2), si son datos ya procesados en SRA contiene una descripción de la secuencia.\n",
    "\n",
    "Línea 2: la secuencia.\n",
    "\n",
    "Línea 3: Comienza con +. Puede ser sólo el símbolo + o repetir la info del Header.\n",
    "\n",
    "Línea 4: Información de la calidad de secuenciación de cada base. Cada letra o símbolo representa a una base de la secuencia codificado en formato ASCII.\n",
    "\n",
    "La info de calidad se codifica en ASCII porque esto permite en 1 sólo caracter codificar un valor de calidad. Por lo tanto la línea 2 y la 4 tienen el mismo número de caracteres.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dc7836e",
   "metadata": {},
   "source": [
    "**Descarga de archivos FASTQ**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d544633a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "El sistema no puede encontrar la ruta especificada.\n",
      "El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos.\n"
     ]
    }
   ],
   "source": [
    "!rm -f SRR003265.filt.fastq.gz 2>/dev/null\n",
    "!wget:-nd https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA18489/sequence_read/SRR003265.filt.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "0fa8aa56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('SRR003265.filt.fastq.gz', <http.client.HTTPMessage at 0x230471643d0>)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import urllib.request\n",
    "\n",
    "# Eliminar el archivo si existe\n",
    "archivo_local = 'SRR003265.filt.fastq.gz'\n",
    "if os.path.exists(archivo_local):\n",
    "    os.remove(archivo_local)\n",
    "\n",
    "# Descargar el archivo desde la URL\n",
    "url = 'https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA18489/sequence_read/SRR003265.filt.fastq.gz'\n",
    "urllib.request.urlretrieve(url, archivo_local)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1c89f0",
   "metadata": {},
   "source": [
    "\n",
    "En este código:\n",
    "\n",
    "1. Se utiliza `os.path.exists()` para verificar si el archivo local (`SRR003265.filt.fastq.gz`) ya existe. Si existe, se elimina con `os.remove()`.\n",
    "\n",
    "2. Se utiliza `urllib.request.urlretrieve()` para descargar el archivo desde la URL especificada y guardarlo localmente con el nombre `SRR003265.filt.fastq.gz`.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44ad2ee",
   "metadata": {},
   "source": [
    "**Lectura de archivos FASTQ**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6277bc35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: SRR003265.31\n",
      "Name: SRR003265.31\n",
      "Description: SRR003265.31 3042NAAXX:3:1:1252:1819 length=51\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('GGGAAAAGAAAAACAAACAAACAAAAACAAAACACAGAAACAAAAAAACCA')\n",
      "SRR003265.31 SRR003265.31 3042NAAXX:3:1:1252:1819 length=51 GGGAAAAGAAAAACAAACAAACAAAAACAAAACACAGAAACAAAAAAACCA\n",
      "{'phred_quality': [40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 30, 23, 40, 32, 35, 29, 40, 16, 40, 40, 32, 35, 31, 40, 40, 39, 22, 40, 24, 20, 28, 31, 12, 31, 10, 22, 28, 13, 26, 20, 23, 23]}\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict #Se utiliza para crear un diccionario que asigna automáticamente un valor predeterminado a una nueva clave si aún no existe.\n",
    "import gzip #Se utiliza para leer archivos comprimidos en formato gzip.\n",
    "\n",
    "%matplotlib inline\n",
    "import seaborn as sns #Se utilizan para crear gráficos y visualizaciones.\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO #Se utiliza para leer y manipular datos de secuencias biológicas.\n",
    "recs = SeqIO.parse(gzip.open ('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "rec = next(recs)\n",
    "print(rec)\n",
    "print(rec.id, rec.description, rec.seq)\n",
    "print(rec.letter_annotations) # quality scores of our reads, per letter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f398847",
   "metadata": {},
   "source": [
    "**OJO**\n",
    "\n",
    "-rec es un objeto de tipo SeqRecord que contiene información sobre una secuencia biológica individual en formato fastq.\n",
    "\n",
    "-rec.id: Obtiene el identificador único de la secuencia.\n",
    "\n",
    "\n",
    "-rec.description: Obtiene la descripción de la secuencia.\n",
    "\n",
    "\n",
    "-rec.seq: Obtiene la secuencia de letras que representa la secuencia biológica.\n",
    "\n",
    "\n",
    "-rec.letter_annotations: Contiene información adicional asociada a las letras de la secuencia, como puntajes de calidad en el caso de datos de secuenciación.\n",
    "\n",
    "A continuación mostramos una imagen con la tabla de calidad de Phe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cafa4d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import Image\n",
    "\n",
    "# Ruta de la imagen\n",
    "ruta_imagen = 'C:\\\\Users\\\\Gabriel\\\\Desktop\\\\TFG\\\\Imágenes\\\\notebooks\\\\tablaphecalidad.png'\n",
    "\n",
    "# Abrir la imagen\n",
    "imagen = Image.open(ruta_imagen)\n",
    "\n",
    "# Mostrar la imagen\n",
    "imagen.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2a792f9",
   "metadata": {},
   "source": [
    "**CUIDADO**\n",
    "\n",
    "Nunca debemos hacer list(recs), debido a que sobrecarga la memoria, mostrando tooodas las secuencias presentadas en nuestro archivo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "454b2410",
   "metadata": {},
   "source": [
    "**Distribución de los nucleotidos de las lecturas**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "66259de6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G: 20.68 5359334\n",
      "A: 28.60 7411965\n",
      "C: 21.00 5444053\n",
      "T: 29.58 7666885\n",
      "N: 0.14 37289\n"
     ]
    }
   ],
   "source": [
    "#Con el siguiente código realizo un recuento de el número de bases presentes en todas las lecturas(\n",
    "#recordemos que cada read o lectura es una secuencia)\n",
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "count = defaultdict(int)\n",
    "for rec in recs:\n",
    "    for letter in rec.seq: \n",
    "        count [letter] += 1\n",
    "tot = sum(count.values())\n",
    "for letter, count in count.items():\n",
    "    print('%s: %.2f %d'% (letter, 100. * count / tot, count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b843e749",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defaultdict(<class 'int'>, {})\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABSwAAAL0CAYAAAAGIXkoAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAACmqUlEQVR4nOzdd3hchZ3v/8+Z0ahLoy7hgptkG7AxYIMLbizNEELq5SawTlhYIDe5EDbk5i6b3w1J9gYS9i6QhIQAmwUSWiqpG1MSXIQbBgwYsCX3qi6N+mjK+f0xc8YWbhppZs4Z6f16Hj8Plo5mvrqBvfLH32KYpmkKAAAAAAAAABzAZXcBAAAAAAAAAGAhsAQAAAAAAADgGASWAAAAAAAAAByDwBIAAAAAAACAYxBYAgAAAAAAAHAMAksAAAAAAAAAjkFgCQAAAAAAAMAxCCwBAAAAAAAAOEaG3QWki3A4rMOHD6ugoECGYdhdDgAAAAAAAJBWTNNUV1eXxo0bJ5fr5H2UBJZDdPjwYU2cONHuMgAAAAAAAIC0duDAAU2YMOGknyewHKKCggJJkf8HLSwstLkaAAAAAAAAIL10dnZq4sSJsZztZAgsh8gaAy8sLCSwBAAAAAAAAIbpdOsWOboDAAAAAAAAwDEILAEAAAAAAAA4BoElAAAAAAAAAMcgsAQAAAAAAADgGASWAAAAAAAAAByDwBIAAAAAAACAYxBYAgAAAAAAAHAMAksAAAAAAAAAjkFgCQAAAAAAAMAxCCwBAAAAAAAAOAaBJQAAAAAAAADHILAEAAAAAAAA4BgElgAAAAAAAAAcg8ASAAAAAAAAgGMQWAIAAAAAAABwDAJLAAAAAAAAAI5BYAkAAAAAAADAMQgsAQAAAAAAADgGgSUAAAAAAAAAxyCwBAAAAAAAAOAYBJYAAAAAAAAAHIPAEgAAAAAAAIBjEFgCAAAAAAAAcAwCSwAAAAAAAACOQWAJAAAAAAAAwDEILAEAAAAAAAA4BoElAAAAAAAAAMcgsAQAAAAAG9TWt+imJ1/XEV+f3aUAAOAoBJYAAAAAYIN/f3mH/ra9SU+t32d3KQAAOAqBJQAAAACkmK8voLcPdEiS1tU321sMAAAOQ2AJAAAAACm2YVerwmbkn9873KnWbr+9BQEA4CAElgAAAACQYq/tbBn0+9oP/R4AgLGMwBIAAAAAUswKKKsr8iVJ6+oJLAEAsBBYAgAAAEAKHWzv1Z6WHrldhu66fLqkyB5L0zRtrgwAAGcgsAQAAACAFKqNdlPOmeDVJTMrlJXhUmOnXzubum2uDAAAZyCwBAAAAIAUssbBF9eUK9vj1kVTSiRJaxkLBwBAEoElAAAAAKRMOGxq/a5WSdKSmjJJ0tKackmRsXAAAEBgCQAAAAAp8/6RTrX1DCgv063zJhZJkpZMjwSXm3a3yR8M2VgdAADOQGAJAAAAAClijYMvmFoqjzvyx7EZlQUqL8hSXyCkN/a121keAACOQGAJAAAAACliHdxZHB0HlyTDMLSkOvL7deyxBACAwBIAAAAAUqE/ENLmvW2SpMXVZYM+Z42Fs8cSAAACSwAAAABIiS172zUQDKuyMEvVFfmDPndxNMB873CnWrv9dpQHAIBjEFgCAAAAQAqs2xnpnlxcXS7DMAZ9rqIgWzOrCmSa0mvRK+IAAIxVBJYAAAAAkAJH91eWnvDzS6eXS5LW1TEWDgAY2wgsAQAAACDJ2noG9N7hTklHx78/bEn0EE/tzhaZppmy2gAAcBoCSwAAAABIstd2RrorZ1YVqKIg+4TPXDi5RJkZLh3x9WtXc3cqywMAwFEILAEAAAAgyaxx8JN1V0pStset+VNKJElr61pSUhcAAE5EYAkAAAAASWSapmp3WvsrTx5YSoPHwgEAGKsILAEAAAAgifa29upQR58y3a5YB+XJLK6OHN7ZsKtV/mAoFeUBAOA4BJYAAAAAkES19ZGr3xdMKlJuZsYpn51ZVaCy/Cz1BUJ6c19HCqoDAMB5CCwBAAAAIInWRfdXLj7F/kqLy2XExsLXRYNOAADGGgJLAAAAAEiSYCisDbtbJUmLa8qH9DXssQQAjHUElgAAAACQJO8c8qmrP6jC7AzNHu8d0tdYnZjvHvKprWcgmeUBAOBIBJYAAAAAkCS10XHwRdPK5HYZQ/qaisJszawqkGlKr9FlCQAYgwgsAQAAACBJrLHuxTWn3195rNhYeD2BJQBg7CGwBAAAAIAk6PEH9db+dklHA8ihsvZdrqtvlmmaCa8NAAAnI7AEAAAAgCTYtKdVgZCpCcU5OrMkN66vvWhyiTIzXDrs69eu5p4kVQgAgDMRWAIAAABAEtTWR66DL6kpk2EMbX+lJSfTrYsml0iKdFkCADCWEFgCAAAAQBLU7owEjYury4f19eyxBACMVQSWAAAAAJBgjZ39qmvslmFIi6aVDus1rEM9G3a3aiAYTmR5AAA4GoElAAAAACTYa9Hr4LPGeVWclzms1zirqlBl+ZnqHQjpzejxHgAAxgICSwAAAABIMGuMe3Gc18GP5XIZWlzNWDgAYOwhsAQAAACABDJNU7XRDssl1cMPLCVpcU1k/yWHdwAAYwmBJQAAAAAkUF1jt5q6/MrKcOmCScUjei3r8M47h3xq7xlIRHkAADgegSUAAAAAJJDVXXnRlBJle9wjeq3KwmzNqCyQaUrrd7UmojwAAByPwBIAAAAAEqg2Or69ZAT7K49l7cFkLBwAMFYQWAIAAABAggwEw9q0p02SdPEI91dalsQCyxaZppmQ1wQAwMkILAEAAAAgQd7a367egZBK8zJ1VlVhQl5z/pRSZbpdOtTRp90tPQl5TQAAnIzAEgAAAAASxNpfeXF1mVwuIyGvmZPp1oVTIsd7autbEvKaAAA4GYElAAAAACTIumiguDhB4+CWxdXl0ddnjyUAYPQjsAQAAACABPD1BfTOwQ5JRw/lJIq1x3LDrlYFQuGEvjYAAE5DYAkAAAAACbBhV6vCpjS1PE/jinIS+tpnn1Go0rxM9QyE9Nb+joS+NgAATkNgCQAAAAAJULszMq69JMHj4JLkchmxq+OMhQMARjsCSwAAAABIgNd2tkpSLFhMNGssfC2HdwAAoxyBJQAAAACM0MH2Xu1p6ZHbZWjBtNKkvMeSmsjhnXcOdqijdyAp7wEAgBMQWAIAAADACNVGux7Pm1ikwmxPUt6jyputmop8maa0fldrUt4DAAAnILAEAAAAgBFatzMSWCZrHNxidVmyxxIAMJoRWAIAAADACITDptZHA0trz2SyLJke3WNZ1yLTNJP6XgAA2IXAEgAAAABG4P0jnWrvDSgv063zJhYl9b3mTylRptulQx192tvam9T3AgDALgSWAAAAADAC66L7KxdMLZXHndw/YuVmZmjupOLo+zIWDgAYnQgsAQAAAGAEXouOgy9O8ji45dixcAAARiMCSwAAAAAYpv5ASJv3tklK/v5Ky9Lo4Z2Nu1sVCIVT8p4AAKQSgSUAAAAADNPre9s0EAyrsjBL08rzU/KeZ59RqJK8THX7g9p6oCMl7wkAQCoRWAIAAADAMNVa4+DV5TIMIyXv6XIZurg60s25ro49lgCA0YfAEgAAAACGqTZ6cCdV4+AW6/3W1rPHEgAw+hBYAgAAAMAwtHb79d7hTkmKdTymihVYvnOwQ77eQErfGwCAZCOwBAAAAIBhWL+rVZI0s6pA5QVZKX3vM7w5qq7IV9iU1u+iyxIAMLoQWAIAAADAMFjj4ItT3F1pYSwcADBaEVgCAAAAQJxM0zx6cCfF+ystS2vKJUnr6ptlmqYtNQAAkAwElgAAAAAQp72tvTrU0adMt0sXTSmxpYb5U0vkcRs62N6nfa29ttQAAEAyEFgCAAAASDsfHOlUZ799x2Zq65slSRdMKlJuZoYtNeRmZmjupGJJkS5LAABGCwJLAAAAAGnlgyOduur76/SRH6xTS7fflhrWRfdGLomOZdvFen/2WAIARhMCSwAAAABpZcu+dknSgbY+3fqzLeoPhFL6/sFQWBuiF8IvtungjsU6vLNxV6sCobCttQAAkCgElgAAAADSSl1DV+yf39zfobt+9bbC4dQdnXnnkE9d/qC8OR7NHu9N2fueyDnjvCrO9ajLH9TbBzpsrQUAgEQhsAQAAACQVuoaI4HldfMmyOM29Od3jujfX96RsvevjY5fL5pWKrfLSNn7nojbZcS6PBkLBwCMFgSWAAAAANKGaZqxwPJzCyfrvk+eK0n60au79MstB1JSgxVYLq6xdxzcsjS6x7KWwzsAgFGCwBIAAABA2mjpHlB7b0CGIU0rz9en507Q7X9XLUn6l9++q/U7k9tl2OMP6s39kR2ai23eX2mxgtOtBzrk67PvcjoAAIlCYAkAAAAgbdRHuyvPLMlVTqZbkvSVy6fro3PGKRg2ddvTb2hnU9epXmJENu1pVTBsamJJjiaV5iXtfeIxrihH08rzFDalDbsYCwcApD8CSwAAAABpwxoHn15ZEPuYYRj6t0+fq7mTitXVH9Q/PPm6Wrr9SXn/ddY4eHV5Ul5/uJZEx8LXsccSADAKEFgCAAAASBs7GrslSdMr8wd9PNvj1mMr5+rMklwdaOvTLT/bov5AKOHv/9pOK7B0xji4ZUl0LJzAEgAwGhBYAgAAAEgb9SfosLSU5mfpiX+4UN4cj97a36G7fvW2wmEzYe/d2NmvusZuGUbkQriTLJhaKo/b0P62Xu1r7bG7HAAARoTAEgAAAEBaME1TO04RWEqRQzw/+fu58rgN/fmdI/p/L+1I2Ptb18Fnj/eqOC8zYa+bCHlZGbrgzGJJ0lq6LAEAaY7AEgAAAEBaaOz0q6s/KLfL0NTykx+8WTitVPd98lxJ0o9X79IvXz+QkPe3xsEvdtg4uMUaC6+tb7a5EgAARobAEgAAAEBasA7uTC7NVVaG+5TPfnruBN3xd9WSpH954d1Y2DhcpmmqNvoaSxwbWEYO76zf2apgKGxzNQAADB+BJQAAAIC0cKIL4afyT5dP17VzxikYNvWFp9/QzqauEbx3t5q6/Mr2uHTBpOJhv04yzRrvVVGuR13+oN4+2GF3OQAADBuBJQAAAIC0YAWWNUMMLA3D0P2fPlfzJhWrqz+oG594Xc1d/mG9t9VdeeHkEmV7Tt3daRe3y4iNq3MtHACQzggsAQAAAKSFusZuSdKMIQaWkpTtceuxz83TpNJcHWzv0y0/26L+QCju97b2Qlp7Ip1qCYElAGAUILAEAAAA4Himaao+NhKeH9fXluRl6j9vvFDeHI+2HujQXb98W+GwOeSvHwiGtWlPmyRpcXV5XO+daoujgerWAx3y9QVsrgYAgOEhsAQAAADgeIc6+tQzEJLHbWhy2ckvhJ/MtPJ8PbpyrjxuQ39+94j+7aUdQ/7aN/e3q3cgpLL8TM2sGnp3px0mFOdqanmeQmFTG3a12l0OAADDQmAJAAAAwPHqo+PgU8vy5XEP748xC6aW6rufPFeS9MjqXfrF6/uH9HXWhfFF08rkchnDeu9UssbCa3c221wJAADDQ2AJAAAAwPF2xA7uxDcO/mGfmjtBd/xdtSTp6y9si4WRp2Ltg1zs8P2VliU1kbF19lgCANIVgSUAAAAAx6uL7a8c+Uj2P10+XR87b5yCYVNfePqN2G7ME/H1BvTOwQ5J0uLq9AgsF0wrVYbL0L7WXu1v7bW7HAAA4kZgCQAAAMDxrJHwRASWhmHoe586V/MmFaurP6h/ePJ1NXf5T/jsht2tCpvS1PI8jSvKGfF7p0J+VoYumFQsSVrHWDgAIA05JrC87777ZBiG7rzzztjHTNPUN7/5TY0bN045OTlavny53nvvvUFf5/f7dfvtt6usrEx5eXm69tprdfDgwUHPtLe3a+XKlfJ6vfJ6vVq5cqU6OjpS8F0BAAAAGKlw2FR90/AuhJ9Mtsetxz43T5NKc3WwvU+3/GyL+gOh456z9kAuSZPuSotV77o6xsIBAOnHEYHl66+/rscee0znnnvuoI/ff//9euCBB/Twww/r9ddfV1VVlS6//HJ1dR0d2bjzzjv1wgsv6Pnnn1dtba26u7t1zTXXKBQ6+sPG9ddfr61bt2rVqlVatWqVtm7dqpUrV6bs+wMAAAAwfAfae9UfCCszw6VJpfFfCD+ZkrxMPXHjhfLmeLT1QIe+8sutCofNQc/UxvZXlifsfVNhyfRIva/talEwFLa5GgAA4mN7YNnd3a0bbrhBjz/+uIqLi2MfN01TDz30kL7+9a/rk5/8pGbNmqWnnnpKvb29evbZZyVJPp9PP/3pT/Xv//7vuuyyy3T++efr6aef1rvvvqtXXnlFkvTBBx9o1apV+o//+A8tXLhQCxcu1OOPP64//elP2rFjhy3fMwAAAIChq4uOg1eX58ud4CvdU8vz9ejKufK4Df3Xuw36t5eO/hnhQFuv9rb2yu0yNH9qSULfN9lmj/fKm+NRV39Q7xzy2V0OAABxsT2w/NKXvqSPfOQjuuyyywZ9fM+ePWpoaNAVV1wR+1hWVpaWLVum9evXS5LeeOMNBQKBQc+MGzdOs2bNij2zYcMGeb1ezZ8/P/bMggUL5PV6Y88AAAAAcK6jB3cSMw7+YQumluq7n4xMez2yepee37xfkmIXxM+bWKTCbE9S3jtZ3C5DF1eXSmIsHACQfmwNLJ9//nm9+eabuu+++477XENDgySpsrJy0McrKytjn2toaFBmZuagzswTPVNRUXHc61dUVMSeORG/36/Ozs5BvwAAAACknhVY1iTg4M7JfGruBN1xaY0k6f/73TbV1rdoXTSwTJfr4B+2JDrGvq6ewzsAgPRiW2B54MABffnLX9bTTz+t7Ozskz5nGINHPkzTPO5jH/bhZ070/Ole57777osd6fF6vZo4ceIp3xMAAABAclgj4TOSGFhK0j9dVqOPnTdOwbCp//H0G1pbFz24U5OegaUVtL51oEOd/QGbqwEAYOhsCyzfeOMNNTU1ae7cucrIyFBGRobWrFmjH/zgB8rIyIh1Vn64C7KpqSn2uaqqKg0MDKi9vf2UzzQ2Nh73/s3Nzcd1bx7r7rvvls/ni/06cODAiL5fAAAAAPELhsLa1RQJLKcnObA0DEP3f/pcXTi5WF3+oLr6g8rPytCciUVJfd9kmViSq/FFOQqFTe1o6Dr9FwAA4BC2BZaXXnqp3n33XW3dujX2a968ebrhhhu0detWTZ06VVVVVXr55ZdjXzMwMKA1a9Zo0aJFkqS5c+fK4/EMeubIkSPatm1b7JmFCxfK5/Np8+bNsWc2bdokn88Xe+ZEsrKyVFhYOOgXAAAAgNTa19argVBYOR63JhTnJP39sjLcenTlPE0uzZUkLZhaIo/b9tX/w1ZZmCVJausZsLkSAACGLsOuNy4oKNCsWbMGfSwvL0+lpaWxj99555269957VVNTo5qaGt17773Kzc3V9ddfL0nyer26+eabddddd6m0tFQlJSX66le/qtmzZ8eO+Jx11llasWKFbrnlFj366KOSpFtvvVXXXHONZsyYkcLvGAAAAEC86mP7K/PlSvCF8JMpycvUz26ar0fW7NTKBZNT8p7JUpKXKYnAEgCQXmwLLIfia1/7mvr6+vTFL35R7e3tmj9/vl566SUVFBwdBXnwwQeVkZGh6667Tn19fbr00kv15JNPyu12x5555plndMcdd8SuiV977bV6+OGHU/79AAAAAIjPjobIOHhNRXLHwT/szNJc3Re9HJ7OinMJLAEA6ccwTdO0u4h00NnZKa/XK5/Px3g4AAAAkCJfevZN/fmdI7r7qpm6bdk0u8tJO/f95QM9uma3/nHxFP1/15xtdzkAgDFuqPla+i5jAQAAADDqWSPh06tS22E5WpRYHZa9dFgCANIHgSUAAAAARxoIhrW7uUdS8i+Ej1bF7LAEAKQhAksAAAAAjrS3tUfBsKn8rAyN82bbXU5asjos2wksAQBphMASAAAAgCPVHXMh3DBScyF8tCnJZyQcAJB+CCwBAAAAOFJdQ3R/ZYovhI8msR2W3QSWAID0QWAJAAAAwJHqGrslRTosMTzWDsuegZD6AyGbqwEAYGgILAEAAAA4Ul1TpMNyBhfCh60wO0MZrsg4fUdvwOZqAAAYGgJLAAAAAI7THwhpbwsXwkfKMAwuhQMA0g6BJQAAAADH2d3co7AZ6RCsKMiyu5y0FttjSWAJAEgTBJYAAAAAHKf+mHFwLoSPTHGeRxKXwgEA6YPAEgAAAIDj7IheCK9hHHzESqIj4e10WAIA0gSBJQAAAADHsS6ET6/gQvhIlbDDEgCQZggsAQAAADiONRI+nQvhI8YOSwBAuiGwBAAAAOAofQMh7W/rlcSF8ESIXQlnhyUAIE0QWAIAAABwlJ1N3TJNqTQvU2X5XAgfKXZYAgDSDYElAAAAAEepa7QO7rC/MhHYYQkASDcElgAAAAAcxQosGQdPjGJ2WAIA0gyBJQAAAABHIbBMrNhIeO+ATNO0uRoAAE6PwBIAAACAo9Q1dksisEwUK7AMhEx1+4M2VwMAwOkRWAIAAABwjG5/UIc6+iRJ09lhmRDZHrdyM92SpPaegM3VAABwegSWAAAAAByjPjoOXlGQpaLo7kWMnLXHsrXHb3MlAACcHoElAAAAAMeoZxw8KY7dYwkAgNMRWAIAAABwjB3RDssaxsETqjjPuhTOSDgAwPkILAEAAAA4hnUhfAYdlglVanVY9tBhCQBwPgJLAAAAAI5hjYTXEFgm1NEdlgSWAADnI7AEAAAA4Ai+voAaOvslMRKeaCV5Hkl0WAIA0gOBJQAAAABHsC6Ej/NmqzDbY3M1o0tshyVHdwAAaYDAEgAAAIAj1DEOnjTssAQApBMCSwAAAACOYB3cmc44eMJZOyzbCCwBAGmAwBIAAACAIxwNLOmwTLQSRsIBAGmEwBIAAACAIxBYJo+1w9LXF1AwFLa5GgAATo3AEgAAAIDt2noG1NId6f6rrmAkPNGKcjwyDMk0I6ElAABORmAJAAAAwHZWd+XEkhzlZWXYXM3ok+F2yZsTubzOHksAgNMRWAIAAACwXWwcvIJx8GQp4fAOACBNEFgCAAAAsJ0VWNawvzJprD2W7RzeAQA4HIElAAAAANvVNXZLkmZUsb8yWWKXwnvYYQkAcDYCSwAAAAC2Mk3zaIclI+FJc3Qk3G9zJQAAnBqBJQAAAABbNXf71dEbkMvgQngyFdNhCQBIEwSWAAAAAGxVHx0Hn1Sap2yP2+ZqRq+SvMiVcHZYAgCcjsASAAAAgK12NFjj4HRXJlNJXpYkroQDAJyPwBIAAACAreqbIoHldC6EJ5XVYUlgCQBwOgJLAAAAALayLoRPryKwTKbi2NEdAksAgLMRWAIAAACwjWmaqmuwOiwZCU+mkujRHXZYAgCcjsASAAAAgG0aOvvV5Q/K7TI0pSzP7nJGNSuw7B0IqT8QsrkaAABOjsASAAAAgG2scfApZXnKyuBCeDLlZ2XI4zYkMRYOAHA2AksAAAAAtmEcPHUMw2CPJQAgLRBYAgAAALBNXWMksKyp4OBOKrDHEgCQDggsAQAAANimrikyEj6DC+EpYQWWdFgCAJyMwBIAAACALcJhU/WNjISnUjGBJQAgDRBYAgAAALDFoY4+9Q6E5HEbmlTKhfBUKInusGwnsAQAOBiBJQAAAABb1DdFuiunlefL4+aPJqkQ67BkhyUAwMH4qQAAAACALXY0RPZX1lSyvzJVSq2jOz0BmysBAODkCCwBAAAA2MLaXzmD/ZUpY3VYtvb4ba4EAICTI7AEAAAAYIu66Eg4HZapc3SHJR2WAADnIrAEAAAAkHKhsKn6xshI+HQCy5QpzvNIYoclAMDZCCwBAAAApNyBtl75g2FlZbh0Zkmu3eWMGaV5WZIiV8JN07S5GgAATozAEgAAAEDK1UX3V1ZX5MvtMmyuZuwoyo10WAbDpjr7gzZXAwDAiRFYAgAAAEg5K7BkHDy1sj1u5WW6JUW6LAEAcCICSwAAAAApV8f+SttYl8LZYwkAcCoCSwAAAAApd7TDMt/mSsae0jzrUjiBJQDAmQgsAQAAAKRUMBTW7uYeSXRY2sHqsGwlsAQAOBSBJQAAAICU2tvaq4FQWLmZbo0vyrG7nDGnJJcOSwCAsxFYAgAAAEip+ug4eE1FvlxcCE85dlgCAJyOwBIAAABASu2wAkvGwW1Rwg5LAIDDEVgCAAAASKn66IXwGQSWtrACyzYCSwCAQxFYAgAAAEipuliHJRfC7VCcS2AJAHA2AksAAAAAKTMQDGtPCxfC7RQbCe8N2FwJAAAnRmAJAAAAIGX2tPQoGDZVkJWhM7zZdpczJjESDgBwOgJLAAAAAClz7Di4YXAh3A5WYOnrCygQCttcDQAAxyOwBAAAAJAyVmDJOLh9vDkeWVlxB2PhAAAHIrAEAAAAkDIElvZzuwwV5XgkSe29jIUDAJyHwBIAAABAytQ3dksisLQbeywBAE5GYAkAAAAgJfoDIe1ttS6E59tczdhGYAkAcDICSwAAAAApsau5W2FTKsr1qLwgy+5yxrTiXAJLAIBzEVgCAAAASInYOHhFARfCbWZ1WLYTWAIAHIjAEgAAAEBK7Ige3KlhHNx2sZFwju4AAByIwBIAAABAStRHA8sZVRzcsRs7LAEATkZgCQAAACAl6qIj4TUVBJZ2Y4clAMDJCCwBAAAAJF3vQFD723olcSHcCWI7LBkJBwA4EIElAAAAgKTb2RTprizLz1RpPhfC7Xb06E7A5koAADgegSUAAACApGMc3FmswLK1x29zJQAAHI/AEgAAAEDS1UUP7jAO7gzF0cCyPxBW30DI5moAABiMwBIAAABA0sUCSy6EO0JepluZ7sgfB9vYYwkAcBgCSwAAAABJVx8dCZ9eSWDpBIZhHLPHksASAOAsBJYAAAAAkqqrP6BDHX2SpOnssHSM4tgeSwJLAICzEFgCAAAASKr66IXwysIseXM9NlcDS0le5H8LOiwBAE5DYAkAAAAgqeoarIM7dFc6SXFupMOyjcASAOAwBJYAAAAAkqouur+yhnFwRym1dlhydAcA4DAElgAAAACSqr4p0mE5oyrf5kpwLHZYAgCcisASAAAAQFLtiI6E1zAS7ihcCQcAOBWBJQAAAICk8fUG1NTllyTVVNBh6STssAQAOBWBJQAAAICkqYuOg48vylFBNhfCnaSEHZYAAIcisAQAAACQNEfHwemudBorsKTDEgDgNASWAAAAAJKmvjF6cIf9lY5ztMMyoHDYtLkaAACOIrAEAAAAkDR1jd2SOLjjREW5kRH9UNhUV3/Q5moAADiKwBIAAABA0tRFOyynMxLuOFkZbuVnZUiS2thjCQBwEAJLAAAAAEnR2u1Xa8+ADEOq5kK4Ix3dY+m3uRIAAI4isAQAAACQFNY4+MTiXOVmZthcDU6kOBZYBmyuBACAowgsAQAAACQF4+DOVxLdY9nOpXAAgIMQWAIAAABIiqOBJQd3nCrWYckOSwCAgxBYAgAAAEiK+uhIOIGlc5XGRsIJLAEAzkFgCQAAACDhTNPUjmiHZQ0j4Y5VTGAJAHAgAksAAAAACdfc5ZevLyCXIU0rJ7B0qpLcSGDJDksAgJMQWAIAAABIOOtC+OTSPGV73DZXg5NhhyUAwIkILAEAAAAkHOPg6YEdlgAAJyKwBAAAAJBw9dHAcgYHdxyNHZYAACcisAQAAACQcHWxDksCSyezdlh29QcVCIVtrgYAgAgCSwAAAAAJZZqm6qM7LKcTWDqaN8cjlxH553b2WAIAHILAEgAAAEBCHfH1q8sfVIbL0JSyPLvLwSm4XIaKcxkLBwA4C4ElAAAAgISyxsGnlOUpM4M/cjgdeywBAE7DTw8AAAAAEsoKLBkHTw/WHsv2noDNlQAAEEFgCQAAACCh6thfmVaK8zySpDZ2WAIAHILAEgAAAEBC1cc6LPNtrgRDUZKXJUlq6yawBAA4A4ElAAAAgIQJh03VN0U6LGvosEwLJdEOS66EAwCcgsASAAAAQMJ80NCp3oGQcjxuTS7NtbscDAFXwgEATkNgCQAAACBhVu9oliRdXF2qDDd/3EgHJdEr4XRYAgCcgp8gAAAAACTMq9ubJEnLZ1TYXAmGygosW9lhCQBwCAJLAAAAAAnR0TugN/e3S5KWzyi3uRoMFR2WAACnIbAEAAAAkBBr61sUNiPXwScUs78yXRy7w9I0TZurAQCAwBIAAABAgqyOjoNfwjh4WrE6LP3BsPoCIZurAQCAwBIAAABAAoTDptbURQ7usL8yveRmupWVEfmjIXssAQBOQGAJAAAAYMTePeRTa8+A8rMyNG9ysd3lIA6GYbDHEgDgKASWAAAAAEbs1R2RcfAlNWXyuPljRro5do8lAAB24ycJAAAAACP26g5rHJzr4OmIDksAgJMQWAIAAAAYkZZuv9452CGJ/ZXpygos2WEJAHACWwPLRx55ROeee64KCwtVWFiohQsX6i9/+Uvs86Zp6pvf/KbGjRunnJwcLV++XO+9996g1/D7/br99ttVVlamvLw8XXvttTp48OCgZ9rb27Vy5Up5vV55vV6tXLlSHR0dqfgWAQAAgFFvbV2zTFM6+4xCVRZm210OhoEOSwCAk9gaWE6YMEHf/e53tWXLFm3ZskV/93d/p4997GOxUPL+++/XAw88oIcfflivv/66qqqqdPnll6urqyv2GnfeeadeeOEFPf/886qtrVV3d7euueYahUKh2DPXX3+9tm7dqlWrVmnVqlXaunWrVq5cmfLvFwAAABiNrHHwS2YyDp6uju6wDNhcCQAAUoadb/7Rj3500O+/853v6JFHHtHGjRt19tln66GHHtLXv/51ffKTn5QkPfXUU6qsrNSzzz6r2267TT6fTz/96U/185//XJdddpkk6emnn9bEiRP1yiuv6Morr9QHH3ygVatWaePGjZo/f74k6fHHH9fChQu1Y8cOzZgxI7XfNAAAADCKBENhra2LBpaMg6etkjyPJKmdozsAAAdwzA7LUCik559/Xj09PVq4cKH27NmjhoYGXXHFFbFnsrKytGzZMq1fv16S9MYbbygQCAx6Zty4cZo1a1bsmQ0bNsjr9cbCSklasGCBvF5v7JkT8fv96uzsHPQLAAAAwGBvH+yQry8gb45H500ssrscDFNJXpYkroQDAJzB9sDy3XffVX5+vrKysvSFL3xBL7zwgs4++2w1NDRIkiorKwc9X1lZGftcQ0ODMjMzVVxcfMpnKiqO/5veioqK2DMnct9998V2Xnq9Xk2cOHFE3ycAAAAwGr26PdJduXR6uTLctv/xAsNUHO2wbGOHJQDAAWz/iWLGjBnaunWrNm7cqP/xP/6HPv/5z+v999+Pfd4wjEHPm6Z53Mc+7MPPnOj5073O3XffLZ/PF/t14MCBoX5LAAAAwJjx6o4mSdIlM9hfmc5iR3fosAQAOIDtgWVmZqaqq6s1b9483XfffZozZ46+//3vq6qqSpKO64JsamqKdV1WVVVpYGBA7e3tp3ymsbHxuPdtbm4+rnvzWFlZWbHr5dYvAAAAAEc1dvbrvcOdMoxIhyXSV0nu0Svh4bBpczUAgLHO9sDyw0zTlN/v15QpU1RVVaWXX3459rmBgQGtWbNGixYtkiTNnTtXHo9n0DNHjhzRtm3bYs8sXLhQPp9Pmzdvjj2zadMm+Xy+2DMAAAAA4rcmeh383PFeleVn2VwNRqI42mEZNiVfH5fCAQD2svVK+L/8y7/oqquu0sSJE9XV1aXnn39eq1ev1qpVq2QYhu68807de++9qqmpUU1Nje69917l5ubq+uuvlyR5vV7dfPPNuuuuu1RaWqqSkhJ99atf1ezZs2NXw8866yytWLFCt9xyix599FFJ0q233qprrrmGC+EAAADACFjj4Mu5Dp72PG6XCrIz1NUfVFvvQCzABADADrYGlo2NjVq5cqWOHDkir9erc889V6tWrdLll18uSfra176mvr4+ffGLX1R7e7vmz5+vl156SQUFBbHXePDBB5WRkaHrrrtOfX19uvTSS/Xkk0/K7XbHnnnmmWd0xx13xK6JX3vttXr44YdT+80CAAAAo0ggFNa6+hZJ0iUzCSxHg5K8THX1ByN7LJnwBwDYyDBNkwUlQ9DZ2Smv1yufz8c+SwAAAIx5G3a16rOPb1RpXqZe//plcrlOfRgTzvfxH72mrQc69NjKubrinCq7ywEAjEJDzdcct8MSAAAAgPOtrouMgy+bXk5YOUqURsfA27gUDgCwGYElAAAAgLit3h45uLOccfBRw9pb2dZLYAkAsBeBJQAAAIC4HOro047GLrkMaWlNmd3lIEFKooFlOx2WAACbEVgCAAAAiMvq6HXwC84sVlEu16RHi+JcayQ8YHMlAICxjsASAAAAQFxetcbBZ3BKejQ5usPSb3MlAICxjsASAAAAwJD5gyG9trNFkrR8BvsrR5OjOyzpsAQA2IvAEgAAAMCQbd7Tpr5ASBUFWTpnXKHd5SCBSvI8kthhCQCwH4ElAAAAgCFbvePoOLhhGDZXg0SydlgSWAIA7EZgCQAAAGDIXo0e3LmEcfBRpzQvS5LU5Q/KHwzZXA0AYCwjsAQAAAAwJPtae7S7uUcZLkMX15TZXQ4SrCA7Q25XpGu2gz2WAAAbEVgCAAAAGBJrHHze5GIVZntsrgaJ5nIZKs6N/O/axlg4AMBGBJYAAAAAhoRx8NGPPZYAACcgsAQAAABwWn0DIW3Y1SpJWk5gOWqV5EUCy1YCSwCAjQgsAQAAAJzWxt2t8gfDGufN1vTKfLvLQZJYgWV7L4ElAMA+BJYAAAAATssaB18+s0KGYdhcDZKlOBpYssMSAGAnAksAAAAAp2SaZuzgDvsrR7cSdlgCAByAwBIAAADAKe1u6dH+tl5lul1aNK3U7nKQROywBAA4AYElAAAAgFN6dXtkHHz+1BLlZWXYXA2SiR2WAAAnILAEAAAAcErWODjXwUe/ozssAzZXAgAYywgsAQAAAJxUjz+oTXtaJUmXzCi3uRokGzssAQBOQGAJAAAA4KRe29miQMjUpNJcTSnLs7scJFlJ/tEr4aZp2lwNAGCsIrAEAAAAcFKvWuPg08tlGIbN1SDZrA7LgVBYPQMhm6sBAIxVBJYAAAAATsg0Ta3ZETm4s3wm+yvHgpxMt7I9kT8mMhYOALALgSUAAACAE6pr7NZhX7+yMlxaOLXU7nKQIlaXZRuBJQDAJgSWAAAAAE7o1Wh35aJppcr2uG2uBqly7B5LAADsQGAJAAAA4IRe3R4JLC9hHHxMKabDEgBgs7gDywMHDujgwYOx32/evFl33nmnHnvssYQWBgAAAMA+nf0BbdnXLklaPp3AciwpyYsElu29BJYAAHvEHVhef/31evXVVyVJDQ0Nuvzyy7V582b9y7/8i7797W8nvEAAAAAAqVdb36JQ2NS08jydWZprdzlIITosAQB2izuw3LZtmy666CJJ0i9/+UvNmjVL69ev17PPPqsnn3wy0fUBAAAAsEFsHHwG3ZVjTWkegSUAwF5xB5aBQEBZWVmSpFdeeUXXXnutJGnmzJk6cuRIYqsDAAAAkHLhsKnVdc2SpOUElmNOMYElAMBmcQeW55xzjn7yk59o3bp1evnll7VixQpJ0uHDh1VaWprwAgEAAACk1vtHOtXc5VduplsXTim2uxykGDssAQB2izuw/N73vqdHH31Uy5cv12c/+1nNmTNHkvSHP/whNioOAAAAIH2t3hEZB7+4ukxZGW6bq0GqscMSAGC3jHi/YPny5WppaVFnZ6eKi4/+beutt96q3FyWcQMAAADp7tUdkXFw9leOTaX5BJYAAHvFHVhKktvtVjAYVG1trQzD0PTp0zV58uQElwYAAAAg1dp7BvTW/nZJ0vIZ5TZXAztYHZYdfQGFwqbcLsPmigAAY03cI+E9PT266aabdMYZZ2jp0qVasmSJxo0bp5tvvlm9vb3JqBEAAABAiqytb1bYlGZWFWhcUY7d5cAGRbkeSZJpSr6+gM3VAADGorgDy6985Stas2aN/vjHP6qjo0MdHR36/e9/rzVr1uiuu+5KRo0AAAAAUmT1Dq6Dj3Uet0uF2ZFhPMbCAQB2iHsk/De/+Y1+/etfa/ny5bGPXX311crJydF1112nRx55JJH1AQAAAEiRcNjUmjprfyXj4GNZaX6WOvuDBJYAAFvE3WHZ29urysrK4z5eUVHBSDgAAACQxt455FNbz4AKsjJ0waTi038BRq3i6Fg4gSUAwA5xB5YLFy7UPffco/7+/tjH+vr69K1vfUsLFy5MaHEAAAAAUufV7U2SpCXTy+Rxx/1HBYwiJXmRwzvtvQSWAIDUi3sk/Pvf/75WrFihCRMmaM6cOTIMQ1u3blV2drZefPHFZNQIAAAAIAVW74gEluyvhHUpnA5LAIAd4g4sZ82apfr6ej399NPavn27TNPUZz7zGd1www3KyeGKIAAAAJCOmrv8evugT5K0fDr7K8e6knwCSwCAfeIOLCUpJydHt9xyS6JrAQAAAGCTtdFjO7PGF6qiMNvmamC3kmiHZTuBJQDABkMKLP/whz8M+QWvvfbaYRcDAAAAwB6vRsfBL2EcHJKKozss29hhCQCwwZACy49//ONDejHDMBQKhUZSDwAAAIAUC4bCsQ5L9ldCosMSAGCvIQWW4XA42XUAAAAAsMlbBzrU2R9UUa5H500ssrscOIC1w7KVwBIAYAOX3QUAAAAAsJd1HXxpTbncLsPmauAEdFgCAOwU99Gdb3/726f8/De+8Y1hFwMAAAAg9V7dHhkHv2Qm18ERYe2w7BkIqT8QUrbHbXNFAICxJO7A8oUXXhj0+0AgoD179igjI0PTpk0jsAQAAADSSIOvX+8f6ZRhRDosAUkqzM5QhstQMGyqozegKi+BJQAgdeIOLN96663jPtbZ2akbb7xRn/jEJxJSFAAAAIDUWFMXGQefM6FIpflZNlcDpzAMQ8V5mWru8qu1x68qb7bdJQEAxpCE7LAsLCzUt7/9bf2f//N/EvFyAAAAAFIkNg7OdXB8yNE9lgGbKwEAjDUJO7rT0dEhn8+XqJcDAAAAkGQDwbBqd7ZIYn8ljlec55EktfVyeAcAkFpxj4T/4Ac/GPR70zR15MgR/fznP9eKFSsSVhgAAACA5Nqyr03d/qDK8jM1a5zX7nLgMCV5XAoHANgj7sDywQcfHPR7l8ul8vJyff7zn9fdd9+dsMIAAAAAJNeaHZFx8GXTK+RyGTZXA6exAstWAksAQIrFHVju2bMnGXUAAAAASLFXd0QO7jAOjhM5usOSwBIAkFoJ22EJAAAAIH0cbO9VXWO3XIa0pJrAEscrjnZYssMSAJBqcXdY9vT06Lvf/a7++te/qqmpSeFweNDnd+/enbDiAAAAACTH6ug4+NxJxfLmemyuBk7EDksAgF3iDiz/8R//UWvWrNHKlSt1xhlnyDDYdQMAAACkm9XRcfDlMypsrgROZQWWbQSWAIAUizuw/Mtf/qI///nPuvjii5NRDwAAAIAk6w+E9NrOVknSJQSWOIniXAJLAIA94t5hWVxcrJKSkmTUAgAAACAFNu9pU18gpMrCLJ11RoHd5cChYiPhvQMyTdPmagAAY0ncgeW//uu/6hvf+IZ6e3uTUQ8AAACAJItdB59RwYonnJQVWAZCprr9QZurAQCMJXGPhP/7v/+7du3apcrKSk2ePFkez+AF3W+++WbCigMAAACQeGuiB3fYX4lTyfa4lZvpVu9ASG09AyrI5jgTACA14g4sP/7xjyehDAAAAACpsLelR7tbeuRxG7q4utTucuBwxbmZ6h3oU1vPgCaV5tldDgBgjIg7sLznnnuSUQcAAACAFLCug8+bVELHHE6rJC9Thzr61N7L4R0AQOrEvcMSAAAAQPp6NToOfsnMcpsrQToozrMuhQdsrgQAMJYQWAIAAABjRH8gpI27WyWxvxJDUxoLLP02VwIAGEsILAEAAIAx4o197fIHw6oszFJNRb7d5SANFOfSYQkASD0CSwAAAGCMWFsfGQdfXF0uwzBsrgbpoCQvsue0vYcdlgCA1Bl2YDkwMKAdO3YoGAwmsh4AAAAASVJb3yJJWlJTZnMlSBexHZYc3QEApFDcgWVvb69uvvlm5ebm6pxzztH+/fslSXfccYe++93vJrxAAAAAACPX2u3Xe4c7JUkXVxNYYmiO7rAksAQApE7cgeXdd9+tt99+W6tXr1Z2dnbs45dddpl+8YtfJLQ4AAAAAInx2q7IsZ2ZVQUqL8iyuRqkC2uHJSPhAIBUyoj3C373u9/pF7/4hRYsWDBo783ZZ5+tXbt2JbQ4AAAAAIlRG91fyTg44lHCSDgAwAZxd1g2NzeroqLiuI/39PSwuBsAAABwINM0Y/srF9eU21wN0om1w9LXF1AwFLa5GgDAWBF3YHnhhRfqz3/+c+z3Vkj5+OOPa+HChYmrDAAAAEBC7G7p0WFfvzLdLl00ucTucpBGinI8MgzJNKWOvoDd5QAAxoi4R8Lvu+8+rVixQu+//76CwaC+//3v67333tOGDRu0Zs2aZNQIAAAAYASs7sp5k4uVk+m2uRqkkwy3S94cjzp6A2rvGVBZPvtPAQDJF3eH5aJFi/Taa6+pt7dX06ZN00svvaTKykpt2LBBc+fOTUaNAAAAAEZgXWwcnP2ViF9JLpfCAQCpFXeHpSTNnj1bTz31VKJrAQAAAJBggVBYG3dHLoQvqWZ/JeJXnJcptfSoncM7AIAUGVJg2dnZOeQXLCwsHHYxAAAAABLr7QMd6vYHVZzr0Tnj+Fkd8bMuhbfSYQkASJEhBZZFRUVDvgAeCoVGVBAAAACAxFkbHQe/uLpMLtfQfqYHjmWNhLcTWAIAUmRIgeWrr74a++e9e/fqn//5n3XjjTfGroJv2LBBTz31lO67777kVAkAAABgWGrrmyVJS9hfiWEqzrN2WHIlHACQGkMKLJctWxb7529/+9t64IEH9NnPfjb2sWuvvVazZ8/WY489ps9//vOJrxIAAABA3Dr7A3r7oE+StLiG/ZUYnpI8jySxwxIAkDJxXwnfsGGD5s2bd9zH582bp82bNyekKAAAAAAjt2FXq0JhU1PL8jS+KMfucpCminPZYQkASK24A8uJEyfqJz/5yXEff/TRRzVx4sSEFAUAAABg5Gqj+ysXMw6OESjNZ4clACC1hjQSfqwHH3xQn/rUp/Tiiy9qwYIFkqSNGzdq165d+s1vfpPwAgEAAAAMT+3OaGBZTWCJ4bM6LNsILAEAKRJ3h+XVV1+t+vp6fexjH1NbW5taW1v1sY99THV1dbr66quTUSMAAACAOB1s79Welh65XYYWTCu1uxyksZI8AksAQGrF3WEpSRMmTNB3vvOdRNcCAAAAIEGscfDzJhapMNtjczVIZ9aV8L5ASH0DIeVkum2uCAAw2sXdYQkAAADA+dYxDo4EKcjKkMdtSOJSOAAgNQgsAQAAgFEmFDb1WjSwXMLBHYyQYRjssQQApBSBJQAAADDKvHfYp47egPKzMjRnYpHd5WAUYI8lACCVCCwBAACAUWZddH/lwmml8rj5kR8jZ3VYMhIOAEiFYR3dkaTm5mbt2LFDhmFo+vTpKi8vT2RdAAAAAIbJOrjDODgSpSSfDksAQOrE/detPT09uummmzRu3DgtXbpUS5Ys0bhx43TzzTert7c3GTUCAAAAGKK+gZDe2NcuiYM7SJwSq8OSwBIAkAJxB5Zf+cpXtGbNGv3hD39QR0eHOjo69Pvf/15r1qzRXXfdlYwaAQAAAAzRpj2tGgiFNb4oR1PK8uwuB6NEcXSHZSuBJQAgBeIeCf/Nb36jX//611q+fHnsY1dffbVycnJ03XXX6ZFHHklkfQAAAADiYI2DL64uk2EYNleD0aIk1yOJHZYAgNSIu8Oyt7dXlZWVx328oqKCkXAAAADAZrU7o4El+yuRQCX5WZLYYQkASI24A8uFCxfqnnvuUX9/f+xjfX19+ta3vqWFCxcmtDgAAAAAQ9fU1a/tDV0yDOli9lcigY7usAzYXAkAYCyIeyT8oYce0lVXXaUJEyZozpw5MgxDW7duVXZ2tl588cVk1AgAAABgCF6LdleeM65QJdGdg0AiFOdFRsLZYQkASIW4A8vZs2ervr5eTz/9tLZv3y7TNPWZz3xGN9xwg3JycpJRIwAAAIAhWBfbX1lucyUYbawAvL13QKZpsh8VAJBUcQWWgUBAM2bM0J/+9CfdcsstyaoJAAAAQJxM04wd3FnC/kokWHF0JDwUNtXZH5Q3x2NzRQCA0SyuHZYej0d+v5+/TQMAAAAcpq6xW01dfmVluDR3UrHd5WCUyfa4lZfpliS1MxYOAEiyuI/u3H777fre976nYDCYjHoAAAAADMO6+mZJ0vyppcr2uG2uBqNRcXQsnD2WAIBki3uH5aZNm/TXv/5VL730kmbPnq28vLxBn//tb3+bsOIAAAAADE1t9ODOEq6DI0lK8jJ1sL2PDksAQNLFHVgWFRXpU5/6VDJqAQAAADAM/mBIm3a3SZIWs78SSWId3mnrJbAEACRX3IHlE088kYw6AAAAAAzTm/s61BcIqSw/SzOrCuwuB6NUSfTwDh2WAIBki3uHpSQFg0G98sorevTRR9XV1SVJOnz4sLq7uxNaHAAAAIDTq90Z2V+5uLqUA5lIGmuHZRuBJQAgyeLusNy3b59WrFih/fv3y+/36/LLL1dBQYHuv/9+9ff36yc/+Uky6gQAAABwErX1kf2Vi2vKba4Eo1kJgSUAIEXi7rD88pe/rHnz5qm9vV05OTmxj3/iE5/QX//614QWBwAAAODUOnoH9M4hnyRpMQd3kERWYNnODksAQJLF3WFZW1ur1157TZmZmYM+PmnSJB06dChhhQEAAAA4vfW7WmWaUk1Fvqq82XaXg1GsOJcOSwBAasTdYRkOhxUKhY77+MGDB1VQwIJvAAAAIJXWxcbB6a5EcjESDgBIlbgDy8svv1wPPfRQ7PeGYai7u1v33HOPrr766kTWBgAAAOAUTNPUuvrIwZ0lBJZIspI8jyQCSwBA8sU9Ev7ggw/qkksu0dlnn63+/n5df/31qq+vV1lZmZ577rlk1AgAAADgBPa19upge588bkPzp5TaXQ5GuZK8LElSZ39QgVBYHnfc/S8AAAxJ3IHluHHjtHXrVj333HN68803FQ6HdfPNN+uGG24YdIQHAAAAQHKt2xkZBz//zGLlZcX9oz0QF2+OR4YhmabU0RtQeUGW3SUBAEapYf1Uk5OTo5tuukk33XRTousBAAAAMES10XHwpYyDIwXcLkNFOR619wbU1jNAYAkASJphBZaHDh3Sa6+9pqamJoXD4UGfu+OOOxJSGAAAAICTC4bCWr+rVZK0uKbc5mowVhTnZcYCSwAAkiXuwPKJJ57QF77wBWVmZqq0tFSGYcQ+ZxgGgSUAAACQAu8c8qmrPyhvjkezx3vtLgdjRGlepnY396i9l8ASAJA8cQeW3/jGN/SNb3xDd999t1wuliwDAAAAdqitj+yvXDStVG6XcZqngcQozs2UxKVwAEByxZ049vb26jOf+QxhJQAAAGAjK7BczP5KpFBJHoElACD54k4db775Zv3qV79KRi0AAAAAhqDbH9Sb+9slSUuq2V+J1CkmsAQApEDcI+H33XefrrnmGq1atUqzZ8+Wx+MZ9PkHHnggYcUBAAAAON6m3a0Khk2dWZKrM0tz7S4HY0hpNLBkhyUAIJniDizvvfdevfjii5oxY4YkHXd0BwAAAEByrWMcHDZhhyUAIBXiDiwfeOAB/ed//qduvPHGJJQDAAAA4HTW1TdLkpZUE1gitdhhCQBIhbh3WGZlZeniiy9ORi0AAAAATuOIr0+7mnvkMqRF0wgskVrWDst2AksAQBLFHVh++ctf1g9/+MNk1AIAAADgNKxx8NkTiuTN9ZzmaSCxrB2WbeywBAAkUdwj4Zs3b9bf/vY3/elPf9I555xz3NGd3/72twkrDgAAAMBgtdHAknFw2MHqsOwPhNU3EFJOptvmigAAo1HcgWVRUZE++clPJqMWAAAAAKcQDpt6bWc0sOTgDmyQl+lWptulgVBYrT1+TcjkSj0AIPHiDiyfeOKJhL35fffdp9/+9rfavn27cnJytGjRIn3ve9+LXSCXJNM09a1vfUuPPfaY2tvbNX/+fP3oRz/SOeecE3vG7/frq1/9qp577jn19fXp0ksv1Y9//GNNmDAh9kx7e7vuuOMO/eEPf5AkXXvttfrhD3+ooqKihH0/AAAAQDJ90NCp1p4B5Wa6df6ZxXaXgzHIMAwV53nU2OlXe09AE/jXEACQBHHvsEykNWvW6Etf+pI2btyol19+WcFgUFdccYV6enpiz9x///164IEH9PDDD+v1119XVVWVLr/8cnV1dcWeufPOO/XCCy/o+eefV21trbq7u3XNNdcoFArFnrn++uu1detWrVq1SqtWrdLWrVu1cuXKlH6/AAAAwEhY4+ALppYqM8PWH+UxhpXkZUlijyUAIHni7rCcMmWKDMM46ed379495NdatWrVoN8/8cQTqqio0BtvvKGlS5fKNE099NBD+vrXvx4bQ3/qqadUWVmpZ599Vrfddpt8Pp9++tOf6uc//7kuu+wySdLTTz+tiRMn6pVXXtGVV16pDz74QKtWrdLGjRs1f/58SdLjjz+uhQsXaseOHYM6OgEAAACnqo2Ogy9mfyVsVJIXuWPApXAAQLKcNrD87W9/q/nz52v8+PGSIt2MxwoEAnrrrbe0atUq/a//9b9GVIzP55MklZSUSJL27NmjhoYGXXHFFbFnsrKytGzZMq1fv1633Xab3njjDQUCgUHPjBs3TrNmzdL69et15ZVXasOGDfJ6vbGwUpIWLFggr9er9evXnzCw9Pv98vv9sd93dnaO6HsDAAAARqI/ENLmPW2S2F8JexXnRg7vtBJYAgCS5LSBpcvl0pIlS/TCCy9ozpw5+vKXv3zC5370ox9py5Ytwy7ENE195Stf0eLFizVr1ixJUkNDgySpsrJy0LOVlZXat29f7JnMzEwVFxcf94z19Q0NDaqoqDjuPSsqKmLPfNh9992nb33rW8P+fgAAAIBE2rK3Xf5gWJWFWaquyLe7HIxhJdFL4XRYAgCS5bSLbz7+8Y/r+eef1+c///lTPnfVVVfpN7/5zbAL+Z//83/qnXfe0XPPPXfc5z48gm6a5inH0k/0zImeP9Xr3H333fL5fLFfBw4cGMq3AQAAACTFuvpmSdLi6vLT/iwMJJMVWLLDEgCQLEPa1H3RRRdp7dq1p3zm17/+dWyUO1633367/vCHP+jVV18ddNm7qqpKko7rgmxqaop1XVZVVWlgYEDt7e2nfKaxsfG4921ubj6ue9OSlZWlwsLCQb8AAAAAu6yLHtxhHBx2o8MSAJBsQz66YwV2559//qC/0TVNUw0NDWpubtaPf/zjuN7cNE3dfvvteuGFF7R69WpNmTJl0OenTJmiqqoqvfzyyzr//PMlSQMDA1qzZo2+973vSZLmzp0rj8ejl19+Wdddd50k6ciRI9q2bZvuv/9+SdLChQvl8/m0efNmXXTRRZKkTZs2yefzadGiRXHVDAAAAKRaS7df7x+J7FS/mIM7sBk7LAEAyRb3lfCPf/zjg37vcrlUXl6u5cuXa+bMmXG91pe+9CU9++yz+v3vf6+CgoJYJ6XX61VOTo4Mw9Cdd96pe++9VzU1NaqpqdG9996r3NxcXX/99bFnb775Zt11110qLS1VSUmJvvrVr2r27Nmxq+FnnXWWVqxYoVtuuUWPPvqoJOnWW2/VNddcw4VwAAAAON5r0evgM6sKVF6QZXM1GOvosAQAJFvcgeU999yTsDd/5JFHJEnLly8f9PEnnnhCN954oyTpa1/7mvr6+vTFL35R7e3tmj9/vl566SUVFBTEnn/wwQeVkZGh6667Tn19fbr00kv15JNPyu12x5555plndMcdd8SuiV977bV6+OGHE/a9AAAAAMlSyzg4HCQWWLLDEgCQJIZpmqbdRaSDzs5Oeb1e+Xw+9lkCAAAgZUzT1KLv/k1HfP166qaLtGx6ud0lYYxr7OzX/Hv/KrfLUP3/vUouF0egAABDM9R8bcgdli6X67TXCA3DUDAYHHqVAAAAAE5pV3OPjvj6lZnh0kWTh3fkEkikolyPJCkUNtXZH1BRdKclAACJMuTA8oUXXjjp59avX68f/vCHolkTAAAASKza+mZJ0oWTi5WT6T7N00DyZWW4lZ+VoW5/UG09AwSWAICEG3Jg+bGPfey4j23fvl133323/vjHP+qGG27Qv/7rvya0OAAAAGCsq40e3FlczSg4nKMkL1Pd/iB7LAEASeEazhcdPnxYt9xyi84991wFg0Ft3bpVTz31lM4888xE1wcAAACMWYFQWBt3t0ni4A6cpTh6eKetJ2BzJQCA0SiuwNLn8+l//+//rerqar333nv661//qj/+8Y+aNWtWsuoDAAAAxqytBzrU7Q+qJC9TZ5/B4Uc4R0l0j2Vbj9/mSgAAo9GQR8Lvv/9+fe9731NVVZWee+65E46IAwAAAEicdXWR/ZWLppVyiRmOQoclACCZhhxY/vM//7NycnJUXV2tp556Sk899dQJn/vtb3+bsOIAAACAsWxddH8l4+BwmtJoYMkOSwBAMgw5sPzc5z4nw+BvdQEAAIBU8PUF9PaBDknS4hoO7sBZjnZYElgCABJvyIHlk08+mcQyAAAAABxrw65WhU1palmexhfl2F0OMEhJLoElACB5hnUlHAAAAEBy1e6M7K9czDg4HIgOSwBAMhFYAgAAAA5UWx/ZX7m4msASzsMOSwBAMhFYAgAAAA5zoK1Xe1t75XYZWjCt1O5ygOPQYQkASCYCSwAAAMBhaqPXwc+fWKTCbI/N1QDHs3ZYdvUHNRAM21wNAGC0IbAEAAAAHCY2Ds7+SjiUN8cjlxH55w7GwgEACUZgCQAAADhIKGzqtV2RwHIJgSUcyuUyVGxdCiewBAAkGIElAAAA4CDbDvnU0RtQQVaG5kwosrsc4KTYYwkASBYCSwAAAMBBrP2VC6aVKsPNj+twLmuPJYElACDR+AkIAAAAcJB19c2SGAeH8xXnRQ5CtRNYAgASjMASAAAAcIjegaDe2NcuSVpcTWAJZyvJy5IktfUEbK4EADDaEFgCAAAADrFpT5sCIVPji3I0pSzP7nKAUyqxOiw5ugMASDACSwAAAMAhXquP7K9cXF0mwzBsrgY4NetKeCsj4QCABCOwBAAAAByirqlbknTBpCJ7CwGGoCR6JZwdlgCARCOwBAAAAByi0dcvSTrDm2NzJcDpWYElV8IBAIlGYAkAAAA4RENnJLCs8mbbXAlwerEOS3ZYAgASjMASAAAAcID+QEi+vsi15cpCAks437E7LE3TtLkaAMBoQmAJAAAAOEBDdBw8x+NWYXaGzdUAp2d1WA4Ew+odCNlcDQBgNCGwBAAAABzAGgevLMziQjjSQm6mW1kZkT9SsscSAJBIBJYAAACAAzTGAkvGwZEeDMNgjyUAICkILAEAAAAHaOTgDtLQsXssAQBIFAJLAAAAwAEafH5JUhUdlkgjsQ5LAksAQAIRWAIAAAAOwEg40pEVWLLDEgCQSASWAAAAgAM0MBKONMQOSwBAMhBYAgAAAA7Q4KPDEunH2mFJhyUAIJEILAEAAACbhcOmmrrosET6KcnzSCKwBAAkFoElAAAAYLO23gEFQqYMQ6ooyLK7HGDISvIi/7629wRsrgQAMJoQWAIAAAA2s8bBS/Oy5HHzIzrSR7HVYckOSwBAAvHTEAAAAGCzxtjBHborkV64Eg4ASAYCSwAAAMBmsQvhHNxBminJPXolfCAYtrma4TNNU3WNXWn9PQDAaEJgCQAAANiskQvhSFNl+Vkqy8+UaUpv7m+3u5xhe+GtQ7riwbX68eqddpcCABCBJQAAAGA7OiyRrlwuQ4uryyRJa+uaba5m+H6/9bAkaf3OVpsrAQBIBJYAAACA7Ro6/ZKkSi+BJdLP0unlkqS19ekZWPYHQtq0JxJU1jV1yTRNmysCABBYAgAAADazRsLpsEQ6WlITCSy3HepUS7ff5mrit3lPm/oDkd2VHb0BtXRzQAgA7EZgCQAAANgsNhJOhyXSUHlBls4+o1CSVFvfYnM18VvzoVH2+sYumyoBAFgILAEAAAAb9QdC8vUFJEmVBQSWSE+xsfA03GNpBZY5Hrckqb6p285yAAAisAQAAABs1Rjtrsz2uFSYk2FzNcDwLJ0ePbxT36JwOH12QB7q6NPOpm65DOmTF4yXJNXRYQkAtiOwBAAAAGzUcMz+SsMwbK4GGJ55k0qUm+lWS7dfHzR02l3OkFkdoedNLNK8ycWS6LAEACcgsAQAAABsZO2vrOTgDtJYZoZLC6eWSpLW1qXPHss1OyKB5bLpFaqpKJAU2WHJpXAAsBeBJQAAAGCjRg7uYJRItz2WgVBYr+2MhKvLZpSruiJfhiG1cykcAGxHYAkAAADYqMHnlxQZCQfSmRVYbtnXph5/0OZqTm/rgQ51+YMqzvVo9nivsj1unVmSK0mqb2KPJQDYicASAAAAsFEjI+EYJSaX5mpiSY4CIVMbd7faXc5pWePgi2vK5XZF9sceHQtnjyUA2InAEgAAALBRAyPhGCUMw9CSmvQZC19TZ+2vLI99bHplviQuhQOA3QgsAQAAABtZV8LpsMRosNQKLOudfXinpduvdw/5JElLa8piH6+JBpZcCgcAexFYAgAAADYJh001ddFhidFjUXWp3C5De1p6dKCt1+5yTqo2GqiefUahKo75ywIuhQOAMxBYAgAAADZp6x1QIGTKMKSKgiy7ywFGrDDbowvOLJJ0dOTaiazalh4zDi5p0KXw1h4uhQOAXQgsAQAAAJtY4+CleVnyuPnRHKPDUofvsQyHzVhtyz4UWB57KZw9lgBgH34qAgAAAGzSGDu4Q3clRg+ra3H9rlYFQmGbqzne+0c61dozoLxMt+ZOKj7u81wKBwD7EVgCAAAANoldCOfgDkaRWeO9Ks71qNsf1Fv7O+wu5zjWOPjCaWXKzDj+j8RHD+/QYQkAdiGwBAAAAGzSyIVwjEJul6HFDh4LX7MjOg4+o/yEn58eDSzr6LAEANsQWAIAAAA2ocMSo9XSmjJJ0tp6ZwWWnf0BvbG/XZK0rObEgSWXwgHAfgSWAAAAgE0aOv2SpEovgSVGF2uP5buHfGpz0LXt9TtbFQqbmlqWpzNLc0/4zLRyLoUDgN0ILAEAAACbNHUyEo7RqbIwWzOrCmSa0joHdVla+yuXTj9xd6Uk5WRyKRwA7EZgCQAAANiEkXCMZlYouK6+xeZKIkzTjO3UXHaKwFLiUjgA2I3AEgAAALBBfyCkjt6AJAJLjE5La6zAstkRuyB3NXfrUEefMjNcmj+15JTPcikcAOxFYAkAAADYoDHaXZntcakwJ8PmaoDEmze5WNkelxo7/drhgNHqNXWRTs/5U0qUm3nq/+a4FA4A9iKwBAAAAGzQ4Ds6Dm4Yhs3VAImX7XFrwdRSSYqNYtsptr/yJNfBj2WNhO9sIrAEADsQWAIAAAA2aODgDsYAKxxcW2fvHsv+QEibdrdKkpbNOH1gaV0Kb+sZUEu3P9nlAQA+hMASAAAAsIE1El7lJbDE6GUd3tm8t019AyHb6ti4u1X+YFhneLNVU5F/2ue5FA4A9iKwBAAAAGzQ4It0bXFwB6PZtPI8jS/K0UAwrI17Wm2rw+rwXFpTPuQVDFawyVg4AKQegSUAAABgg0ZGwjEGGIahpdPLJNm7x3JNXZOkoY2DW2oqI3ss6bAEgNQjsAQAAABs0MBIOMaIo3ss7QksD7b3aldzj9wuQxdXlw3567gUDgD2IbAEAAAAbGBdCafDEqPdouoyuV2GdjX36FBHX8rf3xoHP39ikbw5niF/HZfCAcA+BJYAAABAioXDppq66LDE2ODN8ei8iUWS7OmytMbBrQNAQ8WlcACwD4ElAAAAkGJtvQMKhEwZhlRRkGV3OUDS2TUWHgiF9drOyLGfZXEGlsdeCq9nLBwAUorAEgAAAEgxaxy8NC9LHjc/kmP0sw7v1O5sUTAUTtn7vrmvXd3+oEryMjV7vDfur7cuhdc3cXgHAFKJn44AAACAFGuMHdyhuxJjw7kTilSU61FXf1BvH+xI2fuurY90dC6pKZPLZcT99VwKBwB7EFgCAAAAKRa7EM7BHYwRx17oXhM9gpMKa6Ij6NZIerxiHZaMhANAShFYAgAAACnW2Bk54FFBYIkxZFmK91g2d/m17VCnJGlJdCQ9XtOjHZb1XAoHgJQisAQAAABSrNFHhyXGHis0fOdghzp6B5L+fuui4+DnjCtURcHw/lvjUjgA2IPAEgAAAEgxRsIxFp3hzdH0ynyFzcjxnWSzOjmXxnkd/Fg5mW5NLOZSOACkGoElAAAAkGLW0Z1KL4ElxpalKRoLD4dNra2PhKLLRhBYStL0Si6FA0CqEVgCAAAAKUaHJcYqq9txbV2LTNNM2vtsO+xTW8+A8rMydMGZxSN6LS6FA0DqEVgCAAAAKdQfCKmjNyCJwBJjz0VTSpSV4VJDZ39SD9lYHZyLppUqM2Nkf+zlUjgApB6BJQAAAJBC1jh4tselwpwMm6sBUivb49b8qaWSkjsWviYB+ystXAoHgNQjsAQAAABSqOGYC+GGYdhcDZB6S2si18LXJCmw9PUF9Ob+Dkkj318pDb4U3sqlcABICQJLAAAAIIWs/ZWVjINjjLK6HjfvaVN/IJTw11+/s0WhsKmp5XmaWJI74tc79lJ4HWPhAJASBJYAAABAClkj4VVcCMcYVVORr6rCbPmDYW3a05bw119bH+ncTER3pYVL4QCQWgSWAAAAQAo1+CIjpRzcwVhlGIaWTo+MhSd6j6VpmlqzI3H7Ky3VFdE9lnRYAkBKEFgCAAAAKdTISDgQCxMTHVjubOrWYV+/MjNcWjClNGGva3VY1jXSYQkAqUBgCQAAAKRQAyPhgBZXl8llRC5vH+7oS9jrWod85k8pUU6mO2Gvy6VwAEgtAksAAAAghawr4XRYYiwrys3UuROKJEnr6hPXZWkFloncXylxKRwAUo3AEgAAAEiRcNhUUxcdloB07Fh4S0Jer28gFDvik+jAkkvhAJBaBJYAAABAirT1DigQMmUYUkVBlt3lALZaFj28U7uzRaGwOeLX27inVQPBsMZ5s1VdkT/i1/swa4/lTi6FA0DSEVgCAAAAKWId3CnNy5LHzY/iGNvmTChSQXaGfH0BvXOwY8SvZ10HXzajXIZhjPj1Psy6FE6HJQAkHz8lAQAAACly9EI43ZVAhtulxdWRLstEjIWvje7CXFqT2HFwC5fCASB1CCwBAACAFGnwRY51VHFwB5B0zB7LER7eOdDWq93NPXK7DC2KhqCJVhPtsNzJpXAASDoCSwAAACBFGqwOSw7uAJKOBpZbD3TI1xcY9utY18EvOLNI3hxPQmr7sOqKyKXwVi6FA0DSEVgCAAAAKdLoi14Ip8MSkCSNL8rRtPI8hcKm1u8c/lj42mhgmejr4MfiUjgAjEx/IKSvv/DukJ4lsAQAAABSxOqwJLAEjhrpWPhAMKz1u1oHvVay1FRwKRwAhmtdfYt+v/XwkJ4lsAQAAABSpJGRcOA4scCyrkWmacb99W/ub1e3P6iSvEzNGudNdHmD1FRyKRwAhuvF9xqG/CyBJQAAAJAidFgCx1swpVSZGS4d6ujTruaeuL/e2l+5tKZMLpeR6PIG4VI4AAxPIBTWKx80Dvl5AksAAAAgBfoDIXX0Ro6KEFgCR+VkunXR5BJJR3dRxiO2v3JGcsfBJS6FA8Bwbd7Tpo7egIpzh3YYjcASAAAASAFrHDzb41JhTobN1QDOsnR6maT491g2dfXrvcOdkqQlNckPLLkUDgDDs2pbZBz8khkVQ3qewBIAAABIgYZjLoQbRnLHVoF0Y+2x3Li7Vf2B0JC/bl1d5LL4rPGFKsvPSkptx8rJdGtCcY4kqZ4uSwAYknDY1EvvRwLLy86uHNLXEFgCAAAAKWDtr6xkHBw4zozKAlUWZqk/ENaWve1D/jprf+WyJF8HP9b06Fh4PXssAWBIth7sUGOnX/lZGZo/tWRIX0NgCQAAAKSANRJexYVw4DiGYcRGuoc6Fh4Km1pXbx3cSV1gyaVwAIjPi9Y4+MwKZWW4h/Q1BJYAAABACjT4IvvuOLgDnJg1Fj7UwzvbDvnU3htQflaGLphUnMzSBqmpiFwKr2+iwxIATsc0Tb34XiSwXHFO1ZC/jsASAAAASIFGRsKBU1pSXSbDkLY3dMX+ezkVaxz84upSedyp+6Pt9EprJJwOSwA4nR2NXdrb2qvMDJeWzxh6NzyBJQAAAJACDYyEA6dUnJepc8d7JQ2ty3JtbH/l0C7OJgqXwjGW/PWDxthfDgDDYV0HX1pTrrysjCF/HYElAAAAkALWlXA6LIGTi42F17ec8jlfb0Bv7m+Pfk1Z0us6FpfCMVYc8fXplp9t0T8+9bp8vQG7y0GasgLLK88Z2nVwC4ElAAAAkGSmaaqpiw5L4HSswLK2vlmhsHnS517b1aKwKU0rz9OE4txUlRfDpXCMBat3NCtsSoGQqc172+wuB2loX2uPtjd0ye0ydNlZBJYAAACAo7T1DCgQioQv5flZNlcDONd5E4tUkJWh9t6Ath3ynfS5NTvsGQe3WJfC6bDEaLZ6R1PsnzfubrWxEqQr69jOgqklKs7LjOtrCSwBAACAJLP2V5blZyozgx/BgZPxuF1aVF0q6eR7LE3T1Nr6aGAZxwGHRLIuhdfRYYlRaiAY1ms7j4aUBJYYjqPj4EO/Dm7hpyUAAAAgybgQDgzd0T2WJw4s65u6dcTXr6wMl+ZPKUllaTFcCsdo98a+dnX7gyqIHkl5/0inOnoHbK4K6aSps19v7u+QJF1xNoElAAAA4DgNvsgl4SoCS+C0ltZEAss393eos//4Qx/WOPj8qaXK9rhTWptlWkWeJC6FY/RaXRcZB7/87EpNK8+TaUqb97DHEkP34vuNkqTzzywa1v5uAksAAAAgyayR8EoO7gCnNbEkV1PL8hQKm1q/8/gx1DV11v5Ke8bBJSk3M0MTS7gUjtErtid2RrkWTI2sadi4m8ASQ/didBx8xTDGwSUCSwAAACDpGn3RC+F0WAJDsqSmTNLxY+G9A8FYl9ey6WUpr+tYXArHaHXE16ftDV1yGZGO56OBJXssMTQdvQPaEP33ZTj7KyUCSwAAACDprA5LAktgaGJ7LOuaZZpm7OObdrdpIBTW+KIcTSvPt6s8SVJ1ZeT96bDEaGN1V86ZWKTivEzNnxrZFftBA3ssMTR//aBJobCpmVUFmlyWN6zXILAEAAAAkqyRkXAgLgumlsrjNnSwvU97WnpiH7fGwZdOL5dhGHaVJ+lohyWXwjHarI4GlsunV0iSKgqyVV2RL9OUNrHHEkOw6r3hXwe3EFgCAAAASUaHJRCfvKwMzZsU6epaW3d0LNwJ+ystXArHaBQIhfXazhZJ0vIZR/87WxDtsmQsHKfTOxCM/d9tAksAAADAofoDIXX0Ri4dE1gCQxcbC6+PhCf7W3u1p6VHbpehRdWldpYmiUvhGJ3e2NeuLn9QJXmZmj3eG/s4h3cwVGt2NMsfDOvMklyddUbBsF+HwBIAAABIImscPNvjUmFOhs3VAOljafSozoZdrfIHQ1oTPcAz98xiFWZ77CxNEpfCMTpZ4+BLa8rkch1duzB/SiSw/OBIp9p72GOJk7PGwVfMqhrR6g4CSwAAACCJGo65EG73zj0gnZxVVaiy/Cz1BUJ6Y2977BDIshn2j4NbaqxL4QSWGCVW72iSJC2fUTHo4+UFWaqpiByaYo8lTsYfDOlvH0T+HbrynMoRvRaBJQAAAJBE1v7KSsbBgbi4XIaW1kS6LP+6vUkbdkVGw52wv9JSY10K5/AORoEGX7+2N3TJMI6uZDjW0bFw9ljixNbvalWXP6jygiydP7F4RK9FYAkAAAAkkTUSXsWFcCBuVmjy7Kb96hkIqTQvU2efUWhzVUdxKRyjyZq6SGfcuROKVJKXedznCSxxOi/FroNXDlopMBy2BpZr167VRz/6UY0bN06GYeh3v/vdoM+bpqlvfvObGjdunHJycrR8+XK99957g57x+/26/fbbVVZWpry8PF177bU6ePDgoGfa29u1cuVKeb1eeb1erVy5Uh0dHUn+7gAAAACpwRc5xsHBHSB+i6Mdln2BkKRIgDnSPwQnktVhuZORcIwC1v7K5SfpYp4fvRS+vaFLbeyxxIeEwqZeeq9RkrTinDNG/Hq2BpY9PT2aM2eOHn744RN+/v7779cDDzyghx9+WK+//rqqqqp0+eWXq6vr6N9e3XnnnXrhhRf0/PPPq7a2Vt3d3brmmmsUCoViz1x//fXaunWrVq1apVWrVmnr1q1auXJl0r8/AAAAoLGLkXBguMryszRr/NGOSieNg0tSdXSnX0v3AAEO0logFFZtfWTtwvKT7Ikty8/S9GhIv3kPXZYYbMveNrX2DMib44mF2yNh65nCq666SlddddUJP2eaph566CF9/etf1yc/+UlJ0lNPPaXKyko9++yzuu222+Tz+fTTn/5UP//5z3XZZZdJkp5++mlNnDhRr7zyiq688kp98MEHWrVqlTZu3Kj58+dLkh5//HEtXLhQO3bs0IwZM1LzzQIAAGBMavQxEg6MxNKacm071CnpaMelU1iXwg+09amusSs2Mgukm7f2d6jLH1RxrkfnTig66XMLppaqrrFbG3e3acWskXfRYfR4MdpdeelZFfK4R94f6dgdlnv27FFDQ4OuuOKK2MeysrK0bNkyrV+/XpL0xhtvKBAIDHpm3LhxmjVrVuyZDRs2yOv1xsJKSVqwYIG8Xm/smRPx+/3q7Owc9AsAAACI19GjO1k2VwKkp6tmnSGXIS2cWqqyfOf9d8SlcIwG1nXwpdPL5T7F2gUrlN+wiw5LHGWapl6M7q9ccU5VQl7TsYFlQ0PkG62sHHwGvbKyMva5hoYGZWZmqri4+JTPVFRUHPf6FRUVsWdO5L777ovtvPR6vZo4ceKIvh8AAACMPaZpqqkzssOSkXBgeGZP8OovX16qR/7+ArtLOSEuhWM0iO2vPMk4uGX+lMio747GLrV2+5NeF9LDtkOdOtTRpxyP+4QX5ofDsYGlxTAGJ/umaR73sQ/78DMnev50r3P33XfL5/PFfh04cCDOygEAADDWtfUMaCAUliRVFBBYAsM1o6pARbnHXy12AutSeH0jHZZIT02d/Xr/SKcMI7KC4VRK87M0ozLy7/zmPW2pKA9pYNV7RyRFAu9sjzshr+nYwLKqKtJC+uEuyKampljXZVVVlQYGBtTe3n7KZxobG497/ebm5uO6N4+VlZWlwsLCQb8AAACAeFjj4GX5mcrMcOyP3gBGINZh2USHJdLT6rpId+W5470qHcLahQXRgyobdzMWjghrf+WKWYkZB5ccHFhOmTJFVVVVevnll2MfGxgY0Jo1a7Ro0SJJ0ty5c+XxeAY9c+TIEW3bti32zMKFC+Xz+bR58+bYM5s2bZLP54s9AwAAACRDYycXwoHRjkvhSHdrouPgy2Ycv07vRGJ7LAksIWlnU5d2NnXL4zZ0ycyh/Ts0FLZeCe/u7tbOnTtjv9+zZ4+2bt2qkpISnXnmmbrzzjt17733qqamRjU1Nbr33nuVm5ur66+/XpLk9Xp1880366677lJpaalKSkr01a9+VbNnz45dDT/rrLO0YsUK3XLLLXr00UclSbfeequuueYaLoQDAAAgqRp8kf1eVQSWwKiVm5mhCcU5Otjep/rGLs3nUjjSSDAU1rr6oe2vtFj/jtc1dqul2+/IY1hIHau7ctG0MhVmexL2urYGllu2bNEll1wS+/1XvvIVSdLnP/95Pfnkk/ra176mvr4+ffGLX1R7e7vmz5+vl156SQUFBbGvefDBB5WRkaHrrrtOfX19uvTSS/Xkk0/K7T46M//MM8/ojjvuiF0Tv/baa/Xwww+n6LsEAADAWBW7EO4lsARGs+mVBTrY3qe6pm4CS6SVtw50qLM/qOJcj+ZMKBrS15TkZWpmVYG2N3Rp8542XT37jOQWCUeLXQdP4Di4ZHNguXz5cpmmedLPG4ahb37zm/rmN7950meys7P1wx/+UD/84Q9P+kxJSYmefvrpkZQKAAAAxK3RFwks6bAERreaynz9bXsTl8KRdlbvaJIkLakpl9t16gPHx1owtVTbG7q0cXcrgeUYdqijT+8c9MkwpMvPPvmdmOFw7A5LAAAAIN1ZHZYElsDoVsOlcKSp1TviGwe3WId3Nuxij+VY9uK2SHflhZNKEr4agMASAAAASJJGRsKBMWE6l8KRhpq6+vXe4U5J0tLp8QWW86dEVh/UN0X2WGJsssbBr0zwOLhEYAkAAAAkDR2WwNjApXCkI+s6+LkTvHF3xxVH91hK0qbdbQmvDc7X0u3X63sj/9tfeU5ix8ElAksAAAAgKfoDIXX0BiQRWAKjnXUpXBJ7LJE2VtdFx8Hj7K60LIgemNqwuyVhNSF9vPJ+o8KmNHu8VxOKcxP++gSWAAAAQBJY4+DZHpcKc2y9dQkgBaZXRrrN6prYYwnnC4bCWhcNLJfNqBjWa1iB5UY6LMekVdY4eBK6KyUCSwAAACApGo65EG4YQ7+8CiA91UTHwnfSYYk0sPVAhzr7g/LmeHTexKJhvcaCqSUyDGlnU7eau9hjOZZ09ge0fmfk4NKKJOyvlAgsAQAAgKSw9ldWMg4OjAk1Vocll8KRBqzr4EtqyuR2De8v1YpyMzWzqlCStGkP18LHkle3N2kgFNa08jxVVxQk5T0ILAEAAIAkaOqMdJtUcSEcGBO4FI50srquSZK0fJjj4JYFU0skSRt2EViOJbHr4Ockp7tSIrAEAAAAkoIOS2BsmVbOpXCkh6aufm071ClJWjbMgzuWhbE9lgSWY0V/IBTr0E3WOLhEYAkAAAAkBYElMLbkZXEpHOlhbV3kqves8YUqL8ga0WtdNCWyx3JXc4+auvoTUR4cbl19i3oHQhrnzdbs8d6kvQ+BJQAAAJAEjccc3QEwNnApHOlgTfQ6+PLpIxsHlyJ7LM+y9lg69Fp4fWOXrnt0g9bVN9tdyqiwalt0HHxWVVKPChJYAgAAAElgdVhWeUfWvQIgfXApHE4XCpux4G75jJGNg1sWRMfCNzh0LPyBl+u0eU+bvvH79xQKm3aXk9YCobBe+aBRUnL3V0oElgAAAEDCmaYZO7rDSDgwdnApHE639UCHOnoDKszO0HkTixLymgunOXePZWNnv156PxKw7Wnp0cvRf8bwbN7TJl9fQKV5mbpwcklS34vAEgAAAEiwtp4BDYTCkqSKAgJLYKw4eimcwBLOtGZH5Dr4kunlynAnJhK6aHJkj+Xu5h41dTprj+UvXz+gUNiUKzq5/NjaXfYWlOascfDLz66U25W8cXCJwBIAAABIOGscvCw/U5kZ/MgNjBVHL4X71c6lcDjQ6tj+ysSMg0uSN9ejs8+I7LHcuMc5eyxDYVPPbd4vSfrnq2Yq0+3Sm/s7tGWvc2pMJ+GwqRffi+6vTPI4uERgCQAAACRcIxfCgTHp2EvhdeyxhMO0dPv1zkGfJGlZgvZXWmJ7LHc5Zyx89Y4mHfb1qyjXo88tnKxPnD9ekvTo2t02V5aeth7sUFOXX/lZGVpUXZr09yOwBAAAABKswRfZX8mFcGDssQ7vMBYOp1kb7a48Z1xhwteVLIwGlpsctMfymU2R7sr/NneCsj1u3bJ0iiTplQ8atauZ/z7j9WJ0HPzvZlYoK8Od9PcjsAQAAAASzBoJr/QSWAJjzfTo4Z16OizhMKt3JPY6+LEunBLdY9nSE5sysNPB9l69Gt3X+dmLzpQkVVcU6LKzKmSa0n+s22NneWnHNE2tio6Dr5iV/HFwicASAAAASLhGX+QPa3RYAmMPl8LhRKGwqbX1VmBZkfDX9+Z4dM646B5LB3RZ/uL1AzJN6eLqUk2N7paVpFuXTpMk/ebNg2ru8ttVXtrZ3tClfa29ysxwaVkC95+eCoElAAAAkGBWhyWBJTD2MBIOJ3r7YIc6egMqzM7Q+ROLkvIeC6ZExsLtDiwDobCef/2AJOmG+ZMGfe7CycU6b2KRBoJh/WzDXhuqS0/WsZ2lNeXKy8pIyXsSWAIAAAAJ1shIODBmVVdwKRzOY42DL6kpV4Y7OVHQwmlWYGnvFe5X3m9Uc5df5QVZuvzsykGfMwxDty2dKkn6+cZ96h0I2lFi2lm1LbXj4BKBJQAAAJBwdFgCYxeXwuFEa6L7HBN9HfxY8yaXyGVIe1p61OCzb4+ldWznv8+bKM8JwtkrzqnS5NJcdfQG9MtoJyZObl9rj7Y3dMntMnTZWYlfJ3AyBJYAAABAAvUHQuroDUgisATGKsbC4SSt3X69c8gnSVqexP2DkT2WXkn2jYXvaelR7c4WGYb0mYsmnvAZt8vQzUsiXZb/UbtHwVA4lSWmHWscfMHUEhXlZqbsfQksAQAAgARq6ows8c/2uFSYk5o9TwCchUvhw9cfCOnhv9Vr8x57x4pHk7X1zTJN6ewzClWR5L9IWzC1RJJ9geVzmyPdlZfMqNCE4tyTPvfpCyaoJC9TB9v79JfouDNOLDYOfk7qxsElAksAAAAgoY4dBzcMw+ZqANihmg7LYfEHQ7rt52/o/71Upzuee0vhsGl3SaOCtb9yeRLHwS1H91imPrDsD4T0qy3WsZ0zT/lsTqZbKxdEDvI8tna3TJN/106ksbNfb+7vkBQZpU8lAksAAAAggazAMtldLACcy+qwrGsksByqgWBYX3rmLa2pi4RrDZ39enN/u81Vpb9Q2NTaOiuwTP7+QWuP5d7WXh3x9SX9/Y61aluD2nsDGufNHtL3+rmFk5SV4dK7h3y2Hwpyqpei4+Dnn1mkyhT/XENgCQAAACRQo4+DO8BYx6Xw+ARDYd35i7f0ygeNyspw6dwJkT2If3rniM2Vpb93DnaovTeggqwMnX9mUdLfrzDbo1nj7dlj+cymfZKkz150ptyu0084lOZn6b/NmyBJemztrqTWlq5efK9RUurHwSUCSwAAACChYiPhXgJLYKzKy8rQ+KLIpXDGwk8tFDZ116/e1n+926BMt0uPrpyrO/6uRpL0l21HGAsfIWscfHFN2QkvZifDwqnRsfBdqeta3NHQpdf3tsvtMvTfLzzxsZ0T+cfFU2UY0qs7mlXHztlBOnoHtCEaOl9JYAkAAACkNyuwTPXoFABnmV4Z6bIkBDm5cNjUP//mHf1+62FluAz96IYLtHxGhZZML1NBVoYaO/16g7HwEVldl7r9lZYFVmC5J3Udls9GuyuvOLsyrpUsk8vyYt2Dj63dnZTa0tUrHzQpFDY1s6pAk8vyUv7+BJYAAABAAjESDkDiUvjpmKap//P7bfrVGwfldhn6wWfP1+VnV0qSsjLcuvycyD//mbHwYWvt9uudgx2SpGXTk7+/0jJvcrFchrSvtVeHO5K/x7J3IKjfvnlIknT9aY7tnMitS6dKkn6/9ZAaov9/OKQXo/sr7eiulAgsAQAAgIQ6OhKeZXMlAOzEpfCTM01T3/7T+3pm034ZhvTAdXN09ewzBj1zzbmR3//Xu4yFD1ftzhaZpjSzqiCla0oKsj2ancI9ln98+7C6/EFNKs3VxdPK4v76888s1kWTSxQImXpi/Z4kVJh+evzB2LGmFbMILAEAAIC0Zpqmmjr9khgJB8Y6LoWfmGma+u6q7Xritb2SpO996lx97Lzxxz23uLpcBdkZaurya8s+xsKHw9pfmYrr4B+2YFp0LDwFgeUzm/ZLkq6/6Ey5hnBs50RuiXZZPrtxv7r6AwmrLV2tqWuWPxjWpNJczawqsKUGAksAAAAgQdp6BjQQCkuSKgoILIGxjEvhJ/bgy3V6dE1kV+B3PjFL18078YGUzAyXrjg70tn153cOp6y+0SIcNmMdcqncX2mJ7bHcndzDO+8c7NA7B33KdLv06bkThv06l86s0LTyPHX5g/rF6wcSWGF6WrXt6Di4YQwvBB4pAksAAAAgQaxx8LL8TGVm8KM2MJZxKfx4D/+tXj/4205J0j0fPVs3zJ90yudjY+HbGhRiLDwu7x7yqbVnQAVZGZo7qTjl7z9vUrHcLkP723p1KIl7LJ+NdldeNbtKpfnDX8Xichm6ZUmky/I/a/coEP3Lx7HIHwzp1e1NkuzbXykRWAIAAAAJ08iFcADH4FL4UY+v3a3/91KdJOnuq2bqHy6ectqvubi6TIXZGWru8mvL3uR26o021jj4xdVl8rhTH/0UZHs0y9pjuSs5Y+Gd/QH9fmuk+/Z04fdQfPz88SrLz9JhX7/+NIa7etfvalWXP6iKgiydP7HItjoILAEAAIAEafBF9ldyIRyAJNVE91juHOMdlk+t36vv/NcHkqSvXD5dty2bNqSvy8xw6Ypoh9ef3+VaeDxW10U65OwYB7csnJrcPZa/e+uQ+gIh1VTk68LJI+8izfa49Q8XT5YkPbpmt0xzbHb1vhgdB7/inMph7wRNBAJLAAAAIEGskfDKFF5jBeBcNRV0WD67ab/u+cN7kqT/eUm17ri0Jq6v/0jsWjhj4UPV3jOgrQc6JEnLbAwsF0wtkSRt3JP4wNI0TT2zMTIOfsP8MxO2Z/Hv509SbqZb2xu6tK6+JSGvmU5CYVMvv98oSVpxzhm21kJgCQAAACRIoy8SWNJhCUDiUviv3zior//uXUnSrUun6q4rpsf9GhdPK5M3x6OWbr8272EsfCjW1jfLNKWZVQU6w5tjWx3zJpfI7TJ0oK1PB9t7E/rab+5v147GLmV7XPrEBcM/tvNh3lyP/vuFkUNQj63dnbDXTRdb9raptWdA3hyP5kcDZ7sQWAIAAAAJYnVYElgCkMb2pfA/vH1YX/v12zJN6cZFk3X3VTOH1QWXmeHSledUSpL+i7HwIVkT3V9pZ3elJOVnZWi2tccywdfCre7Ka+eMkzfHk9DXvuniKXK7DNXubNG2Q76EvrbTrXovMg5+2VmVtuw+PRaBJQAAAJAgjYyEAzjGWL0U/pd3j+iffrFVYVP67EVn6p6Pnj2ikd2rZ0dGU/+y7Qhj4acRDptaUxcJLJdPr7C5GmnhtMTvsWzvGdCfouF1Io7tfNjEktzYv3OPrxs7XZaBUDi2v9L6SwI7EVgCAAAACdJIhyWADxlrl8Jfeb9Rtz/3lkJhU5+6YIK+8/FZI94veHG1NRY+oE1J2Ic4mmw77FNrz4DyszI0LwGHaEZqQRIO7/zmzYMaCIY1a3yhzp3gTdjrHuu2pVMlSX9654gOdfQl5T2c5jt//kCHff0qzM7Q0un2dudKBJYAAABAQvQHQmrvDUgisARw1Fi6FL6mrllffOZNBcOmPjpnnO7/9LkJuTLscbu0wroW/g5j4aeyOjoOfnF1qe0jvZI0b1Kx3C5DB9v7dKBt5HssTdPUM5usYzuTEnZs58Nmjfdq0bRShcKm/rN2T1Lew0l+teWAnly/V5L0b/9tjrI9bnsLEoElAAAAkBBNnX5JUlaGS4U5GTZXA8Apxsql8PW7WnTrz7ZoIBTWinOq9MB1c+ROQFhpuTp6LfzF9xoUDIUT9rqjzeodTZKk5TPsHweXImsRrC7IRHRZbtjVqj0tPcrPytC1c8aN+PVO5dZol+Xzm/fL1xdI6nvZaeuBDn39d9skSV++tEZXRv9ywG4ElgAAAEACxA7ueLOT1vEBIP1Yl8JH8w7L1/e26eYnt8gfDOvSmRX6wWfPT3h336JppSrKjYyFcy38xDp6B7T1QIckabnNB3eOtTA2Fj7y/92s7spPnD9eeVnJ/cvBZdPLNbOqQD0DIT2zaV9S38suTV39+sLP39BAMKzLzqrUly+tsbukGAJLAAAAIAGswLKScXAAx7AuhTd3+dXRO/ouhb+1v13/8MTr6guEtKSmTD+64QJlZiQ+ajh2LPxPXAs/obX1LQqb0ozKAp3hzbG7nJhj91ia5vCPJjV19evF6BXr6+efmZDaTsUwDN2yJNJl+cRre+UPhpL+nqk0EAzri0+/qYbOfk0rz9OD/31OQlY4JAqBJQAAAJAAjT4O7gA43rGXwusaR1eX5bZDPn3uPzer2x/Uwqmlevxz85K6++4j0bHwVdsYCz+Ro+PgzumulKS5k4qV4TJ0qKNPB9uHf8DmV1sOKhg2NXdSsc46ozCBFZ7cR+eMU1Vhtpq7/Pr9W4dT8p6p8q0/vqct+9pVkJWhxz83TwXZHrtLGoTAEgAAAEiAY0fCAeBYNdFL4fVNo2eP5QdHOvX3P92krv6gLpxcrJ/emNywUoqMFhfnetTWM6BNjIUPEg6bWlsXObizzAEXno+Vl5WhOROLJEkbhrnHMhQ29Wx0HPz6i5LfXWnJzHDppsWTJUmPrdutcHj4HaJO8uym/Xpm034ZhvT9z56nqeX5dpd0HAJLAAAAIAEYCQdwMrE9lqOkw3JnU5f+/j82qaM3oPMmFuk/b7xQuZnJPzaW4XZpxazoWDjXwgd573CnWroHlJfp1rzJJXaXc5wFUyM1Dffwztq6Zh3q6JM3xxPrtE2Vz1x0pvKzMrSzqVur65pS+t7J8Ma+Nt3zh8iRna9eMUN/N7PS5opOjMASAAAASABGwgGczGi6FL6npUfXP75JrT0DmjW+UE/ddFFKR0k/MjtyGXrVtiOMhR/DGgdfVF2WlB2iIxXbY7lreHssraM3n547IemdvB9WmO2J7cx8dM3ulL53ojX4+vWFp99UIGTq6tlV+uLyaXaXdFLO+7cYAAAASENHR8KzbK4EgNPUjJJL4at3NOm6RzeoqcuvmVUF+vlN8+XNSe3euwVTS1SSl6n23sCwx4tHo9XRcXCn7a+0zJ1ULI/b0GFfvw60xbfH8lBHn/62PRLIpuLYzon8w8WTleEytGlPm96OXmJPN/2BkG57+g01d/k1o7JA//bpOTIM5xzZ+TACSwAAAGCETNNUU6dfEiPhAI5Xk+aXwrv9Qd3923d14xOvx8KOp/9xvorzMlNeS4bbpSuj18L/i2vhkiRfb0Bv7W+XJC2fUWFzNSeWm5mhOROKJMU/Fv6LzfsVNiM7TKfZtGvxDG+Orj0v0t372Nr067I0TVPf+P02vX2gQ94cjx773FzlZSV/jcNIEFgCAAAAI9TWM6CB6GhiRQGBJYDB0vlS+MbdrVrx0Fo9tzly8OSmi6fod1+6WGX59nWTX3PMtfAAY+Fat7NZYTMSjFv/njlRbCw8jsAyEArr+dcPSJJuWGBPd6Xl1qVTJUl/2XZE+1t7ba0lXj/fuE+/3HJQLkP64WfP16TSPLtLOi0CSwAAAGCErHHwsvxMR+4OA2C/GVWRsfCv/fptvfx+47D2+KVSfyCkf/3T+/rs4xt1sL1P44ty9NwtC/SNj56tnMzU7hD8sPlTSlRqjYXvYix89Q5nj4NbrMByw+6h77H86weNauryqyw/U1ecXZXM8k5rZlWhlk0vV9iU/qM2fbosN+5u1bf/+L4k6Z+vmqmlDrsifzL8NAUAAACMUCMXwgGcxpcvrVF5QZb2tvbqlp9t0cqfbtb2hk67yzqhtw906CM/WKef1u6RaUqfuXCiVt25RAunldpdmqTB18L/PMavhYfDptbE9lc6cxzcYu2xPOLr1/62oXUoPrMp0tl73byJjvgLwduiXZa/3HJAbT3OX+9wqKNPX3rmTQXDpq6dM063LJlqd0lDZv//2gAAAECaa/BF9ldyIRzAycyZWKRXv7pcX1w+TZkZLtXubNHV31+nr7/wrlq7/XaXJ0kaCIb17y/t0CcfWa9dzT2qKMjSEzdeqO9+6tyUXgIfio/MjoyFv/j+2B4Lf/9Ip5q7/MrNdGve5GK7yzmlnEy3zptYJGloY+F7W3q0rr5FhiF99iJ7x8EtC6eVatb4QvUHwvr5hn12l3NK/YGQbvv5FrX2DOjsMwr1vU+d6+gjOx9GYAkAAACMkDUSXuklsARwcvlZGfraipn661eW6erZVQqbkQ6y5f9vtf5j3W4NBO0L3rY3dOrjP3pNP/zbToWi3Vgv/dNSXTLTmV17F00pUVl+pjp6A1o/hsfCre7KRdPKlJVh76j+UMTGwofwv5m1N3XZ9HJNLMlNal1DZRhGrEvxZxv2qj8QsrmiEzNNU3f/9l1tO9SpkrxMPfa5ubavcogXgSUAAAAwQk3RwJIOSwBDMbEkVz++Ya6ev3WBzj6jUF39Qf3fP3+gKx9aq1dSvN8yFDb149U79dEf1ur9I50qzvXoR9dfoB989nwV5ab+CvhQDR4LP2xzNfZZvaNJkvP3V1qOHt5pO+W/5/5gSL9646Ak6Yb5k1JS21B9ZPYZGl+Uo9aeAf3mzYN2l3NCP63doxfeOiS3y9DD15+vCcXOCHzjQWAJAAAAjFADgSWAYVgwtVR/vH2xvvep2SrLz9Kelh7948+26HP/uVk7GrqS/v67m7v16Z+s1/2rdigQMnXZWRV68Z+W6iPRK9xOd7U1Fv5e45gcC/f1BfTm/g5J6RNYXnBmZI9lQ2e/9p3i0vaqbQ1q6xnQGd5sXeKw7y3D7dLNi6dIkv7j/2/vzsOqrPP/j78O22EHNzYFFHFLwVQKMbeyXCudlmkx07HN0ty+VmPbUFo6mqPWjEx1zaitNr9MZ9Iml1zTbEQlzRVXzAREQTY5wOH+/YGcxBVQPOfg83FdXJfnvu9z7jccP4IvPu/PZ/0hWcscawOt71Oz9PY3uyVJrw5ooy7NG9q5opohsAQAAACuUvppWsIB1Iyri0kP3RKh1RN6aESP5vJwddH61Cz1m71Ory3+uVY29igrMzRvwyH1f3e9tqXlyM/spukPxOrDx+MU5Oc8/47FN2ughr4eOn2mRBv2Z9m7nOvu+9QsWcsMRQf5Os0MOi8PV3UIL19r83LrWFZstvPwLRFyc3W86OqhW8IV4OWuQ1kFWrErw97l2Bw9VahRn29VmSE90KmJhnVpau+Saszx3nUAAADAyfy2S7jZzpUAcFZ+nu76Y7/WWjm+h/q1K1/f8uNNR9Rz+mr94/tD12x9y1+yC/XYP35U4te7VFRSptuiG+jbcd31YFy4U23IIZWHvf3alc+yvBF3C/9uT3lQ1rOlY81AvJLOUfUlST9cIrBMzcjT/w6dOhvmh1/P0qrMx+ymxzqXbwT0wboDdq6mXGFxqZ76KFk5hSVq3yRAkwe1c7oxfS4CSwAAAOAqFJVYlV1YIomWcABXL6KBt5Ie66TPn+qsNqH+yi0q1aQlu9R31jqt2lPz9S0Nw9C/ko+q76z12njgpLzcXfXmwLb6eHi8Ggd6XePP4vqpaF9ftjPdrpsWXU8n8y0as2Cbvtp6TJJ0h4NujHQpv61jefKif58rZlfe2SZIIQ7cuTC0S1N5uLpoa1qOkg+fsmsthmHohS+3a096nhr6mvX3IZ3k6e5cm+ycj8ASAAAAuAqZuRZJktnNRQFe7nauBkBdkdC8gZY831VT74tRQ18PHcwq0PB55etb7suo3vqWmblFenJ+sl78crvyLaXqFFlP34zppscTmsrFxXlnYEnSLU3rq6GvWblFpdpwoG63hRuGoUXbftGdf1mrf6f8KheT9GzP5kpo3sDepVVLx8h68nB1UUauRYfPW8fyTLHVtpGNo222c74gP0/d17GxJOn9dQftWsvf1x7U0u3H5e5q0t8f66jQAOf9JUQFAksAAADgKtg23AnwdOrWKwCOx9XFpIdvjdDqCT31TI+oc9a3XK/X//2zsquwvuWS7b+q96x1+m5PpjxcXTSxX2v965kENWvocx0+g9rn6mJS/5iK3cLrblv4L9mFGjZ3s8Z98ZOyC0vUOsRPi567TS/1be1033s83V11c0SgpAvXsfx6+6/KKypVRH1vdY12/M1inuxWvvnOyt0ZOnAi3y41rNmbqWnL9kiSEu9tq7im9e1Sx7VGYAkAAABchXTb+pWO27YGwLn5ebprYr82WjG+u/q2DZG1zNBHPxxRj+mr9c/vD110h+zsgmKN+myrRn22TTmFJWob5q+vn++qZ3o0l6uTz6o834CYutsWbi0zNHfDIfWeuU5r952Qh5uLXujTSl8/31XtwwPtXV6NVbSF/3CgcmBZ0Q7+aHyEU8z+jQ7y051tgmQY5TuGX2+Hsgo0+vNtMgzpkVsjHH5WanUQWAIAAABXIePsDuGsXwmgtkU28NHfh3TSZ0/F29a3fHPJLvWZtU6r92Tarlu1J0O9Z63Tku3H5epi0uheLbToudvUKsTPjtXXnrim9RXkZ1ZeUam+33/C3uVcM/sy8vTA3zfqja93qbDYqlub1td/x3TTyNuj5e6AO2dXR8XGO+euY/nzsdP66WiO3F1NeqBTE3uWVy1Pd28uSVq49Rct35muU1WY+Xwt5FtK9fRHycotKl/mIfHem67Lfa8XN3sXAAAAADizc1vCAeB66NK8oZY831X/Sj6qd5bt1cETBfrDvM3q0bKRGvmZ9eWW8jUAo4N89Zfft1dsk0D7FlzLyncLD9H8H45o6fZ03dE62N4lXRVLqVVzVh/QnDX7VWI15Gt20x/7tdajtzrHrMOq6BhRTx5uLsrMs+hQVoGiGvnaZlf2bReqhr5mO1dYdbc0racOEYHalpajpz/eIklq1tBHHSIC1TGinjpG1FOrEL9rOrO5rMzQ+C9SlJqZr2B/s5IGd5TZzbk32TkfgSUAAABwFWgJB2APri4mPXJrhAbEhupvq/brnxsOae2+8tmFJpP0ZNdm+r/erZx+p+CqGhAbpvk/HNHyXemylLZz2vBmy5Fs/XHhdqVmlq+HeGebIE0a1K5ObKJyLk93V3UID9SPh05p08FTauRn1r9Tync9HxwfYefqqsdkMmnO4I6avTJVyUeytT8zX4eyCnQoq8C2k7uPR/m6nRUBZoeIQAV6e9T4nu+t2q/luzLk4eqivz/WSUF18GcQAksAAADgKtASDsCe/D3dNbF/Gz1ya4SmL9urtFOFenVAG8VHOdfO0VcrLrKegvzMysyz6PvULPVq41yzLPMtpXpn2V7N/+GwDENq6OuhxHvbakBMqNNtqlNVnaMa6MdDp/TDwZOyGoYKi61q3shH8c2cb9OY0AAvTb0/VpJ0urBEW49ma9uRbG1Ny1HK0RzlW0q1Yf9Jbdj/25qdUY181CminjpGloeYLYJ8qzSDdsWuDM1cuU+SNPl37dQhol7tfFJ2RmAJAAAAXIXfWsKdp30NQN3TtKGP/ja4o73LsBsXF5P6x4Rq3sbDWrr9uFMFlqv3ZurVRT/rWM4ZSdIDnZro1QFtrmoGnjPoHNVAs79L1aaDJ5WakSdJGhwf6fQBbYC3u25vFaTbWwVJKt84KTUzT1uOZGvrkRxtS8vWwawCHTxR/vH/zi7h4Gd2+20WZmQ93RweqAAv90qvvT8zT+O+SJEkDU2I1O/jwq/r53Y9EVgCAAAANWQYhjJzLZJoCQcAexsQWx5YrtiVIUup1eHbwk/mW/Tmkl36d8qvkqTw+l6a8rtYdW3R0M6VXR8dIgLl4eaiE3kWncizyOzmovs7Os9mO1Xl6mJS6xB/tQ7xt+3ifaqgWNvSsrU1rTzE/OmXHOVZSrU+NUvrU7MklS/tEN3I92yAGaibQgM0ZsE25VtKdWuz+nr17rq1yc75CCwBAACAGjpVUKxia5kkKciPwBIA7KlTRD0F+5uVkWvR+n1ZuvMmx5xlaRiGFqcc05tf71J2YYlcTNLw25ppfO+W8va4cWIaT3dXdYwI1KaDpyRJ97QPU4C3+xWeVTfU9/FQrzbBtpnApdYy7c3I09azbeRb07J15GShUjPzlZqZry+Sj9qeGxbgqTmDOzr9TvFXcuOMBAAAAOAaq2gHb+jrIQ+3uv0fBwBwdBVt4XM3HNbSHccdMrD8JbtQry7+WWv2lm+Q1DrET3++P1btwwPtW5iddI5qYAssnW2znWvJzdVFbcMC1DYsQEMSyo9l5VsqBZjbf8mRu4uL3h8S51S7qNcUgSUAAABQQxnsEA4ADmXA2cByxa4MFZVYHWaXdGuZoY9+OKzpy/aqsNgqDzcXjenVQk93j6rzM+Uup1frYM3+LlU3hwfq5hs0tL2Uhr5m9W4bot5tQyRJJdYyWcsMh/k7XdsILAEAAIAayji7fiU7hAOAY+gYUU8h/p5Kzy3S+tQs3eUAsyz3ZeTppYXbtS0tR5J0a9P6mnJ/jJo38rVvYQ4gpkmAvh3TXSH+nk6/2U5tc3d10Q2SVUqSbtwYHwAAALhK6afPzrAMILAEAEdQ0RYuSUu3/2rXWiylVs1csU8D3l2vbWk58jW7afKgdlrwdGfCynO0CvG7YdauRNUxwxIAAACooYqWcGZYAoDjGBAbqn9uOKSVuzPt1hb+87HTGvdFilIz8yVJd7YJ0qRB7RQa4HXdawGcEYElAAAAUEPptjUs6/7i9wDgLDqEByo0wFPHTxdp3b4TtjUAr5c1ezP13KdbVVhsVUNfDyXe21YDYkJpeQaqgZZwAAAAoIZsLeHMsAQAh1GpLXzH8et67/+XfFRPzE9WYbFV3Vo01IpxPXR3bBhhJVBNBJYAAABADdlawlnDEgAcyoDY8sBy5dndwmubYRh677tUvfDldlnLDN3XobH+MfQW1fPxqPV7A3URgSUAAABQA0UlVmUXlkhiDUsAcDQdwgPVONBLBcVWrdl7olbvVWot08uLftaMFfskSc/1bK4Zv28vDzciF6CmGD0AAABADWTmWiRJZjcXBXixuykAOBKTyaR+7crXrvymFtvCC4tLNeKTLfr8f2lyMUmTBrXTi31b0wIOXCUCSwAAAKAG0s9pB+c/pgDgeGxt4btrpy38ZL5Fj374o1buzpTZzUVJj3XSkM6R1/w+wI2IwBIAAACogd92CKcdHAAc0c1n28ILi61aszfzmr72kZMFuj9po1KO5ijQ212fPRWvPtd5N3KgLiOwBAAAAGog4+wO4axfCQCOyWQy2WZZLtl+7drCt/+So/uTNurwyUI1qeelhc92UafI+tfs9QEQWAIAAAA1ks4O4QDg8PrHlAeWq/Zk6kzx1beFr96bqYc/2KSs/GK1DfPXV891UfNGvlf9ugAqI7AEAAAAaoCWcABwfO2bBFyztvB/bT6qJ+cnq7DYqm4tGuqLZxIU5Mf3AKA2EFgCAAAANUBLOAA4PpPJpLsr2sJruFu4YRiavTJVLy7cLmuZofs6NtY/h90iX7PbtSwVwDkILAEAAIAa+K0l3GznSgAAl2NrC99d/bbwUmuZXl60QzNX7pMkjbo9WjMebC93V+IUoDYxwgAAAIBqMgxDmbkWSbSEA4Cji20SoCb1vHSmxKrV1WgLLywu1TMfb9Hn/zsqF5M0eVA7TejTSiaTqRarBSARWAIAAADVll1YomJrmSSxfhkAOLhzdwtfWsXdwrPyLXrkwx/13Z5Mebq76P0hcXqsc2RtlgngHASWAAAAQDWln12/sqGvhzzc+JEaABzd3TFhkqTv9mSosLj0stcezirQ/Ukb9dPRHNXzdtenT3bWXTcFX48yAZzFT1cAAABANWWwQzgAOJV2jf0VXt9LRSVlWrXn0m3hPx3N0f1JG3XkZKHC63tp4bNd1Cmy3nWsFIBEYAkAAABUWzqBJQA4FZPJpAFnZ1l+c4ndwlftydDDH2zSyYJitWvsr4XPdlFUI9/rWSaAswgsAQAAgGqqaAknsAQA53H32XUsV+3JVIGlclv4F5vT9NRHW3SmxKruLRtpwdMJrFEM2BGBJQAAAFBNFS3hIQSWAOA02ob5K7KBd6W2cMMwNHPFPr20cIesZYYe6NRE/xgaJ1+zm52rBW5sBJYAAABANVW0hIcEmO1cCQCgqkwmk/rHlM+y/GbHcZVay/THhTs0+7tUSdLoO6I1/YFYubsSlQD2xigEAAAAqomWcABwTgNifmsLf2J+sr5IPioXk/TW79ppfO9WMplMdq4QgERgCQAAAFSbrSU8gMASAJxJ2zB/NW3gLUtpmdbuOyFPdxd9MCROg+Mj7V0agHMQWAIAAADVUFRiVXZhiSTWsAQAZ2MymXR3bPlu4fV9PPT5U511503Bdq4KwPlYRRYAAACohsxciyTJ7OaiAC93O1cDAKiuET2by9/LTf3ahSq8vre9ywFwEQSWAAAAQDWkn9MOzlpnAOB8fM1uerp7c3uXAeAyaAkHAAAAqqEisGTDHQAAgNpBYAkAAABUQ8bZHcJZvxIAAKB2EFgCAAAA1ZDODuEAAAC1isASAAAAqAZawgEAAGoXgSUAAABQDbSEAwAA1C4CSwAAAKAaMvIqWsLNdq4EAACgbiKwBAAAAKrIMAxl5Fok0RIOAABQWwgsAQAAgCrKLixRcWmZJCnIj8ASAACgNhBYAgAAAFWUfnb9ygY+HvJw40dpAACA2sBPWQAAAEAVZbBDOAAAQK0jsAQAAACqKD23YsMdAksAAIDaQmAJAAAAVFFFSzgzLAEAAGoPgSUAAABQRRUt4SEElgAAALWGwBIAAACoot9aws12rgQAAKDuIrAEAAAAqoiWcAAAgNpHYAkAAABUUQab7gAAANQ6AksAAACgCopKrMouLJHEGpYAAAC1icASAAAAdUJxaZlKrGW19vqZuRZJktnNRQFe7rV2HwAAgBudm70LAAAAAGoqt6hEK3dl6Jsdx7VuX5aKrWXyNbspwMtd9XzcFejloQBvdwV6uSvQ+/zHHqrn7a4Ab3cFeLnL7OZ62Xuln9MObjKZrsenBwAAcEMisAQAAIBTOX3mt5ByfWp5SHmufEup8i2lOpZzplqv6+3hqkAvdwV4e/wWcHqXB5uBXu46cqpQEhvuAAAA1DYCSwAAADi802dKtMIWUp5QidWwnWsR5Kv+MaHqHxOqYH+zcgpLlF1YrJwzJTpdWKKcs3/OKSzR6TNnz539c05hsU6fKVGZIRUWW1VYbNWvZ3cCv5RQNtwBAACoVQSWAAAAcEinC0u0fFe6vtlxXN/vz6oUUrYMLg8pB8SEqkWwX6XnBXp7qKl8qnyfsjJDeZbS8nDzTLGyC38LMnMKz36cKdbpwhIVW8s0tEvTa/UpAgAA4CIILAEAAOAwTheWaNnZkHLDeSFlq2C/8pAyNkTRQX6XeZXqcXExKcCrfB3LCHlfs9cFAABAzdxQgeWcOXM0ffp0HT9+XG3bttWsWbPUrVs3e5cFAABwQ8spLNbys+3e36dmqbSs9kNKAAAAOK4bJrD84osvNHbsWM2ZM0e33Xab3n//ffXr10+7du1SRESEvcsDAAC4oeQUFmv5zgwtPTuT8tyQsnWIn21NyuggXztWCQAAAHswGYZhXPky5xcfH6+OHTsqKSnJdqxNmzYaNGiQpkyZcsXn5+bmKiAgQKdPn5a/v39tlgoAAFAnZRcUa/mudC3dka6NFwkpB8SEqn9sqJo3IqQEAACoi6qar90QMyyLi4u1ZcsW/fGPf6x0vHfv3tq4ceNFn2OxWGSxWGyPc3NzJUn3zdkgN8+qL+IOAAAAqcwwdPBEASElAAAAruiGCCyzsrJktVoVHBxc6XhwcLDS09Mv+pwpU6bojTfeuOD4vox8uZjLaqVOAACAuq5NqL8GxISof0yooggpAQAAcBE3RGBZwWQyVXpsGMYFxypMnDhR48ePtz3Ozc1VeHi4Pnw8Tj6+LPgOAABQXU3qealpQzpVAAAAcHk3RGDZsGFDubq6XjCbMjMz84JZlxXMZrPMZvMFxxOaN2ANSwAAAAAAAKCWuNi7gOvBw8NDnTp10ooVKyodX7Fihbp06WKnqgAAAAAAAACc74aYYSlJ48eP15AhQxQXF6eEhAR98MEHSktL04gRI+xdGgAAAAAAAICzbpjA8qGHHtLJkyf15ptv6vjx42rXrp2++eYbRUZG2rs0AAAAAAAAAGeZDMMw7F2EM8jNzVVAQIBOnz7NGpYAAAAAAABANVU1X7sh1rAEAAAAAAAA4BwILAEAAAAAAAA4DAJLAAAAAAAAAA6DwBIAAAAAAACAwyCwBAAAAAAAAOAwCCwBAAAAAAAAOAwCSwAAAAAAAAAOg8ASAAAAAAAAgMMgsAQAAAAAAADgMAgsAQAAAAAAADgMAksAAAAAAAAADoPAEgAAAAAAAIDDILAEAAAAAAAA4DAILAEAAAAAAAA4DAJLAAAAAAAAAA6DwBIAAAAAAACAwyCwBAAAAAAAAOAwCCwBAAAAAAAAOAwCSwAAAAAAAAAOg8ASAAAAAAAAgMMgsAQAAAAAAADgMAgsAQAAAAAAADgMAksAAAAAAAAADoPAEgAAAAAAAIDDcLN3Ac7CMAxJUm5urp0rAQAAAAAAAJxPRa5WkbNdCoFlFeXl5UmSwsPD7VwJAAAAAAAA4Lzy8vIUEBBwyfMm40qRJiRJZWVl+vXXX+Xn5yeTyWTvcoAbRm5ursLDw3X06FH5+/vbuxwA1xDjG6i7GN9A3cX4Buqu6zG+DcNQXl6ewsLC5OJy6ZUqmWFZRS4uLmrSpIm9ywBuWP7+/vxABNRRjG+g7mJ8A3UX4xuou2p7fF9uZmUFNt0BAAAAAAAA4DAILAEAAAAAAAA4DAJLAA7NbDbrT3/6k8xms71LAXCNMb6BuovxDdRdjG+g7nKk8c2mOwAAAAAAAAAcBjMsAQAAAAAAADgMAksAAAAAAAAADoPAEgAAAAAAAIDDILAEAAAAAAAA4DAILAE4hHXr1umee+5RWFiYTCaTFi9eXOm8YRhKTExUWFiYvLy81LNnT+3cudM+xQKosilTpuiWW26Rn5+fgoKCNGjQIO3du7fSNYxvwDklJSUpNjZW/v7+8vf3V0JCgv773//azjO2gbpjypQpMplMGjt2rO0YYxxwTomJiTKZTJU+QkJCbOcdZWwTWAJwCAUFBWrfvr3++te/XvT8tGnT9Je//EV//etftXnzZoWEhOiuu+5SXl7eda4UQHWsXbtWI0eO1KZNm7RixQqVlpaqd+/eKigosF3D+AacU5MmTTR16lQlJycrOTlZd9xxhwYOHGj7Tw1jG6gbNm/erA8++ECxsbGVjjPGAefVtm1bHT9+3PaxY8cO2zlHGdsmwzCM63pHALgCk8mkRYsWadCgQZLKf8MTFhamsWPH6qWXXpIkWSwWBQcH689//rOeeeYZO1YLoDpOnDihoKAgrV27Vt27d2d8A3VM/fr1NX36dA0fPpyxDdQB+fn56tixo+bMmaPJkyfr5ptv1qxZs/j+DTixxMRELV68WCkpKRecc6SxzQxLAA7v0KFDSk9PV+/evW3HzGazevTooY0bN9qxMgDVdfr0aUnloYbE+AbqCqvVqgULFqigoEAJCQmMbaCOGDlypAYMGKA777yz0nHGOODcUlNTFRYWpmbNmunhhx/WwYMHJTnW2Ha7rncDgBpIT0+XJAUHB1c6HhwcrCNHjtijJAA1YBiGxo8fr65du6pdu3aSGN+As9uxY4cSEhJUVFQkX19fLVq0SDfddJPtPzWMbcB5LViwQFu3btXmzZsvOMf3b8B5xcfH66OPPlLLli2VkZGhyZMnq0uXLtq5c6dDjW0CSwBOw2QyVXpsGMYFxwA4rlGjRmn79u36/vvvLzjH+AacU6tWrZSSkqKcnBwtXLhQQ4cO1dq1a23nGduAczp69KjGjBmj5cuXy9PT85LXMcYB59OvXz/bn2NiYpSQkKDmzZtr/vz56ty5syTHGNu0hANweBU7llX8tqdCZmbmBb/5AeCYnn/+ef3nP//R6tWr1aRJE9txxjfg3Dw8PBQdHa24uDhNmTJF7du31+zZsxnbgJPbsmWLMjMz1alTJ7m5ucnNzU1r167Vu+++Kzc3N9s4ZowDzs/Hx0cxMTFKTU11qO/fBJYAHF6zZs0UEhKiFStW2I4VFxdr7dq16tKlix0rA3AlhmFo1KhR+uqrr7Rq1So1a9as0nnGN1C3GIYhi8XC2AacXK9evbRjxw6lpKTYPuLi4jR48GClpKQoKiqKMQ7UERaLRbt371ZoaKhDff+mJRyAQ8jPz9f+/fttjw8dOqSUlBTVr19fERERGjt2rN5++221aNFCLVq00Ntvvy1vb289+uijdqwawJWMHDlSn332mf7973/Lz8/P9tvagIAAeXl5yWQyMb4BJ/Xyyy+rX79+Cg8PV15enhYsWKA1a9bo22+/ZWwDTs7Pz8+23nQFHx8fNWjQwHacMQ44pwkTJuiee+5RRESEMjMzNXnyZOXm5mro0KEO9f2bwBKAQ0hOTtbtt99uezx+/HhJ0tChQzVv3jy9+OKLOnPmjJ577jllZ2crPj5ey5cvl5+fn71KBlAFSUlJkqSePXtWOj537lwNGzZMkhjfgJPKyMjQkCFDdPz4cQUEBCg2Nlbffvut7rrrLkmMbaCuY4wDzumXX37RI488oqysLDVq1EidO3fWpk2bFBkZKclxxrbJMAzjut4RAAAAAAAAAC6BNSwBAAAAAAAAOAwCSwAAAAAAAAAOg8ASAAAAAAAAgMMgsAQAAAAAAADgMAgsAQAAAAAAADgMAksAAAAAAAAADoPAEgAAAAAAAIDDILAEAABwYocPH5bJZFJKSkqVrh82bJgGDRpUqzXVhsTERN18881X9Ro9e/bU2LFjr0k9dZnJZNLixYvtXQYAALiBEVgCAADUsmHDhslkMslkMsnd3V1RUVGaMGGCCgoKrvq1w8PDdfz4cbVr165K18+ePVvz5s276vui7jp+/Lj69etn7zIAAMANzM3eBQAAANwI+vbtq7lz56qkpETr16/Xk08+qYKCAiUlJV3V67q6uiokJKTK1wcEBFzV/VD3VefvEwAAQG1ghiUAAMB1YDabFRISovDwcD366KMaPHiwre3WYrFo9OjRCgoKkqenp7p27arNmzfbnpudna3BgwerUaNG8vLyUosWLTR37lxJF28J37lzpwYMGCB/f3/5+fmpW7duOnDggKQLW8KvdO81a9bIZDLpu+++U1xcnLy9vdWlSxft3bu30uf39ddfq1OnTvL09FRUVJTeeOMNlZaW2s4nJiYqIiJCZrNZYWFhGj169GW/XlOnTlVwcLD8/Pz0xBNPqKio6IJr5s6dqzZt2sjT01OtW7fWnDlzLv8mnOeTTz5RXFyc/Pz8FBISokcffVSZmZmXfc6cOXPUokULeXp6Kjg4WA888IDtnGEYmjZtmqKiouTl5aX27dvryy+/rPT8y703F2tZHzRokIYNG2Z7XFxcrBdffFGNGzeWj4+P4uPjtWbNGtv5efPmKTAwUMuWLVObNm3k6+urvn376vjx45Ve95///Kfatm0rs9ms0NBQjRo1ynbu/Jbwl156SS1btpS3t7eioqL02muvqaSk5LJfJwAAgKtBYAkAAGAHXl5ettDnxRdf1MKFCzV//nxt3bpV0dHR6tOnj06dOiVJeu2117Rr1y7997//1e7du5WUlKSGDRte9HWPHTum7t27y9PTU6tWrdKWLVs0fPjwSuHhua507wqvvPKKZsyYoeTkZLm5uWn48OG2c8uWLdNjjz2m0aNHa9euXXr//fc1b948vfXWW5KkL7/8UjNnztT777+v1NRULV68WDExMZf82vzrX//Sn/70J7311ltKTk5WaGjoBWHkhx9+qFdeeUVvvfWWdu/erbfffluvvfaa5s+ff4Wv/G+Ki4s1adIk/fTTT1q8eLEOHTpUKRw8X3JyskaPHq0333xTe/fu1bfffqvu3bvbzr/66quaO3eukpKStHPnTo0bN06PPfaY1q5dK6n6783F/OEPf9CGDRu0YMECbd++XQ8++KD69u2r1NRU2zWFhYV655139PHHH2vdunVKS0vThAkTbOeTkpI0cuRIPf3009qxY4f+85//KDo6+pL39PPz07x587Rr1y7Nnj1bH374oWbOnFnlmgEAAKrNAAAAQK0aOnSoMXDgQNvjH3/80WjQoIHx+9//3sjPzzfc3d2NTz/91Ha+uLjYCAsLM6ZNm2YYhmHcc889xh/+8IeLvvahQ4cMSca2bdsMwzCMiRMnGs2aNTOKi4uvWEtV7r169WpDkrFy5UrbNUuXLjUkGWfOnDEMwzC6detmvP3225Xu8/HHHxuhoaGGYRjGjBkzjJYtW16ypvMlJCQYI0aMqHQsPj7eaN++ve1xeHi48dlnn1W6ZtKkSUZCQsIlX7dHjx7GmDFjLnn+f//7nyHJyMvLu+j5hQsXGv7+/kZubu4F5/Lz8w1PT09j48aNlY4/8cQTxiOPPGIYxpXfm4vVN3DgQGPo0KGGYRjG/v37DZPJZBw7dqzSNb169TImTpxoGIZhzJ0715Bk7N+/33b+b3/7mxEcHGx7HBYWZrzyyisXrcEwDEOSsWjRokuenzZtmtGpU6dLngcAALharGEJAABwHSxZskS+vr4qLS1VSUmJBg4cqPfee08HDhxQSUmJbrvtNtu17u7uuvXWW7V7925J0rPPPqv7779fW7duVe/evTVo0CB16dLlovdJSUlRt27d5O7ufsWaqnLvCrGxsbY/h4aGSpIyMzMVERGhLVu2aPPmzbYZlZJktVpVVFSkwsJCPfjgg5o1a5aioqLUt29f9e/fX/fcc4/c3C7+o+ju3bs1YsSISscSEhK0evVqSdKJEyd09OhRPfHEE3rqqads15SWllZrjc5t27YpMTFRKSkpOnXqlMrKyiRJaWlpuummmy64/q677lJkZKTt8+jbt69+97vfydvbW7t27VJRUZHuuuuuSs8pLi5Whw4dJFXvvbmYrVu3yjAMtWzZstJxi8WiBg0a2B57e3urefPmtsehoaG2VvfMzEz9+uuv6tWrV5Xv++WXX2rWrFnav3+/8vPzVVpaKn9//xp9DgAAAFVBYAkAAHAd3H777UpKSpK7u7vCwsJsoVXF2oImk6nS9YZh2I7169dPR44c0dKlS7Vy5Ur16tVLI0eO1DvvvHPBfby8vKpck2EYV7x3hXNDtopzFQFfWVmZ3njjDd13330X3MPT01Ph4eHau3evVqxYoZUrV+q5557T9OnTtXbt2hqFdxX3/fDDDxUfH1/pnKura5Veo6CgQL1791bv3r31ySefqFGjRkpLS1OfPn1UXFx80ef4+flp69atWrNmjZYvX67XX39diYmJ2rx5s62mpUuXqnHjxpWeZzabJV35vXFxcbG9JxXOXSuyrKxMrq6u2rJlywWfp6+vr+3P539NTSaT7XWr8/dDkjZt2qSHH35Yb7zxhvr06aOAgAAtWLBAM2bMqNbrAAAAVAdrWAIAAFwHPj4+io6OVmRkZKVAKTo6Wh4eHvr+++9tx0pKSpScnKw2bdrYjjVq1EjDhg3TJ598olmzZumDDz646H1iY2O1fv36Km2KUtV7X0nHjh21d+9eRUdHX/Dh4lL+46aXl5fuvfdevfvuu1qzZo1++OEH7dix46Kv16ZNG23atKnSsXMfBwcHq3Hjxjp48OAF92vWrFmVat6zZ4+ysrI0depUdevWTa1bt77ihjuS5ObmpjvvvFPTpk3T9u3bdfjwYa1atUo33XSTzGaz0tLSLqgpPDxc0pXfm0aNGlXaHMdqternn3+2Pe7QoYOsVqsyMzMvuEdVd/b28/NT06ZN9d1331Xp+g0bNigyMlKvvPKK4uLi1KJFCx05cqRKzwUAAKgpZlgCAADYkY+Pj5599lm98MILql+/viIiIjRt2jQVFhbqiSeekCS9/vrr6tSpk9q2bSuLxaIlS5ZcMlAcNWqU3nvvPT388MOaOHGiAgICtGnTJt16661q1apVte9dFa+//rruvvtuhYeH68EHH5SLi4u2b9+uHTt2aPLkyZo3b56sVqvi4+Pl7e2tjz/+WF5eXoqMjLzo640ZM0ZDhw5VXFycunbtqk8//VQ7d+5UVFSU7ZrExESNHj1a/v7+6tevnywWi5KTk5Wdna3x48dfseaIiAh5eHjovffe04gRI/Tzzz9r0qRJl33OkiVLdPDgQXXv3l316tXTN998o7KyMrVq1Up+fn6aMGGCxo0bp7KyMnXt2lW5ubnauHGjfH19NXTo0Cu+N3fccYfGjx+vpUuXqnnz5po5c6ZycnJs92/ZsqUGDx6sxx9/XDNmzFCHDh2UlZWlVatWKSYmRv3796/S+5WYmKgRI0YoKChI/fr1U15enjZs2KDnn3/+gmujo6OVlpamBQsW6JZbbtHSpUu1aNGiKt0HAACgpggsAQAA7Gzq1KkqKyvTkCFDlJeXp7i4OC1btkz16tWTJHl4eGjixIk6fPiwvLy81K1bNy1YsOCir9WgQQOtWrVKL7zwgnr06CFXV1fdfPPNldaprM69q6JPnz5asmSJ3nzzTU2bNk3u7u5q3bq1nnzySUlSYGCgpk6dqvHjx8tqtSomJkZff/11pXUXz/XQQw/pwIEDeumll1RUVKT7779fzz77rJYtW2a75sknn5S3t7emT5+uF198UT4+PoqJidHYsWOrVHOjRo00b948vfzyy3r33XfVsWNHvfPOO7r33nsv+ZzAwEB99dVXSkxMVFFRkVq0aKHPP/9cbdu2lSRNmjRJQUFBmjJlig4ePKjAwEB17NhRL7/8sqQrvzfDhw/XTz/9pMcff1xubm4aN26cbr/99ko1zJ07V5MnT9b//d//6dixY2rQoIESEhKqHFZK0tChQ1VUVKSZM2dqwoQJatiwoR544IGLXjtw4ECNGzdOo0aNksVi0YABA/Taa68pMTGxyvcDAACoLpNx/kI5AAAAAAAAAGAnrGEJAAAAAAAAwGEQWAIAAAAAAABwGASWAAAAAAAAABwGgSUAAAAAAAAAh0FgCQAAAAAAAMBhEFgCAAAAAAAAcBgElgAAAAAAAAAcBoElAAAAAAAAAIdBYAkAAAAAAADAYRBYAgAAAAAAAHAYBJYAAAAAAAAAHAaBJQAAAAAAAACH8f8BMWf3oRe3WcMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1600x900 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "contador_N = defaultdict(int)\n",
    "print(contador_N)\n",
    "for rec in recs:\n",
    "    for i,letter in enumerate(rec.seq): \n",
    "        pos = i+1\n",
    "        if letter == 'N':#es decir, si es deconocida\n",
    "            contador_N[pos] += 1\n",
    "longitud_secuencia= max(contador_N.keys())\n",
    "posiciones = range(1, longitud_secuencia + 1)\n",
    "fig, ax = plt.subplots(figsize=(16,9))\n",
    "ax.plot(posiciones, [contador_N[x] for x in posiciones])\n",
    "ax.set_xlim(1, longitud_secuencia)\n",
    "ax.set_xlabel('Posiciones de la secuencia')\n",
    "ax.set_ylabel('Número de nulos')\n",
    "pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5340325d",
   "metadata": {},
   "source": [
    "Para estos dos siguientes casos ignoramos las 25 primeras posiciones de cada secuencia, ya que son de alta calidad y no presentan variación alguna:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ecea662a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40: 22.76 3007221\n",
      "16: 2.05 270431\n",
      "32: 1.03 136763\n",
      "35: 0.91 120704\n",
      "31: 1.08 142464\n",
      "39: 0.78 102568\n",
      "22: 1.59 209905\n",
      "24: 1.46 193259\n",
      "20: 1.73 228923\n",
      "28: 1.23 162226\n",
      "12: 2.35 309918\n",
      "10: 4.95 654154\n",
      "13: 2.28 301033\n",
      "26: 1.33 176263\n",
      "23: 1.52 201164\n",
      "37: 0.84 111179\n",
      "34: 0.95 125624\n",
      "21: 1.66 219602\n",
      "30: 1.13 149449\n",
      "33: 0.99 131291\n",
      "25: 1.40 184846\n",
      "17: 1.97 259779\n",
      "38: 0.80 106290\n",
      "4: 4.77 630923\n",
      "36: 0.88 115701\n",
      "18: 1.88 248982\n",
      "11: 2.41 318303\n",
      "19: 1.81 239621\n",
      "5: 4.88 645266\n",
      "9: 2.51 332259\n",
      "3: 4.04 533458\n",
      "27: 1.28 168902\n",
      "8: 2.53 334410\n",
      "29: 1.17 154892\n",
      "15: 2.12 280719\n",
      "7: 2.51 331743\n",
      "2: 3.77 498679\n",
      "14: 2.20 291341\n",
      "6: 2.50 330834\n",
      "1: 1.52 200558\n",
      "0: 0.40 52229\n",
      "defaultdict(<class 'int'>, {40: 3007221, 16: 270431, 32: 136763, 35: 120704, 31: 142464, 39: 102568, 22: 209905, 24: 193259, 20: 228923, 28: 162226, 12: 309918, 10: 654154, 13: 301033, 26: 176263, 23: 201164, 37: 111179, 34: 125624, 21: 219602, 30: 149449, 33: 131291, 25: 184846, 17: 259779, 38: 106290, 4: 630923, 36: 115701, 18: 248982, 11: 318303, 19: 239621, 5: 645266, 9: 332259, 3: 533458, 27: 168902, 8: 334410, 29: 154892, 15: 280719, 7: 331743, 2: 498679, 14: 291341, 6: 330834, 1: 200558, 0: 52229})\n"
     ]
    }
   ],
   "source": [
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "contador_calidad = defaultdict(int)\n",
    "for rec in recs:\n",
    "    for i, calidad in enumerate(rec.letter_annotations['phred_quality']):\n",
    "        if i < 25:\n",
    "            continue\n",
    "        contador_calidad[calidad] += 1\n",
    "tot = sum(contador_calidad.values())\n",
    "for calidad, contador in contador_calidad.items():\n",
    "    print('%d: %.2f %d'% (calidad, 100. * contador / tot, contador))\n",
    "print(contador_calidad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "585d3b12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABQcAAALgCAYAAAAp28GCAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAABQY0lEQVR4nO3dfZhkZX0n/F9Nv1T14MxEovMmkwmuoCsIT1ZcXtbIizI4SbAV13TUIBjjqryshBgNEh8n5pIhZiUYiezqGgK6LBNXpIkgMD5hBl3EAIJOWBdRh0iEkZUFZoCpmu6e+/lj7Ka6p9+qq7rrFPfnc119XVXnnLrrO4czfU9/uU91KaWUAgAAAADIzqJ2BwAAAAAA2kM5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmutsdYKK9e/fGww8/HEuWLIlSqdTuOAAAAADQUVJKsWvXrli9enUsWjT92sDClYMPP/xwrFmzpt0xAAAAAKCjPfTQQ3HQQQdNe0zhysElS5ZExL7wS5cubXMaAAAAAOgsO3fujDVr1oz1bNMpXDk4eivx0qVLlYMAAAAAMEez+cg+v5AEAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMtVUObhx48YolUpx3nnnjW1LKcWGDRti9erV0dfXFyeccELcd999zeYEAAAAAFpszuXgnXfeGZ/97GfjiCOOGLf9E5/4RFxyySVx2WWXxZ133hkrV66Mk08+OXbt2tV0WAAAAACgdeZUDj711FPx9re/PT73uc/F85///LHtKaW49NJL48ILL4zTTjstDj/88LjyyivjmWeeiauvvrploQEAAACA5s2pHDz77LPjN3/zN+N1r3vduO3bt2+PHTt2xLp168a2lcvlOP744+P222+fdKxarRY7d+4c9wUAAAAAzL/uRl9wzTXXxHe+8524884799u3Y8eOiIhYsWLFuO0rVqyIf/7nf550vI0bN8af/umfNhoDAAAAAGhSQysHH3rooXj/+98fX/ziF6NSqUx5XKlUGvc8pbTftlEXXHBBPPnkk2NfDz30UCORAAAAAIA5amjl4N133x2PPvpovPKVrxzbNjIyErfddltcdtllcf/990fEvhWEq1atGjvm0Ucf3W814ahyuRzlcnku2QEAAACAJjS0cvC1r31tbNu2Le69996xr6OOOire/va3x7333hsvfvGLY+XKlbF58+ax1+zZsye2bt0axx13XMvDAwAAAABz19DKwSVLlsThhx8+btsBBxwQv/zLvzy2/bzzzouLLrooDjnkkDjkkEPioosuisWLF8fb3va21qUGAAAAAJrW8C8kmckHP/jB2L17d5x11lnx+OOPx9FHHx233HJLLFmypNVvBQAAAAA0oZRSSu0OUW/nzp2xbNmyePLJJ2Pp0qXtjgMAAAAAHaWRfq2hzxwEAAAAAJ47lIMAAAAAkCnlIAAAAABkSjkIAAAAAJlq+W8rBgAAAJitdevWjT2+5ZZb2phkbjo5fydnj5C/VawcBAAAANqivhyZ7HnRdXL+Ts4eIX8rKQcBAAAAIFOllFJqd4h6O3fujGXLlsWTTz4ZS5cunfXrUkpRrVZnPG70mEqlMqtxK5VKlEqlWeeYi9lmj8gr/0Jkj+jsayeis/O7diYn/8xyuXYiOju/a6f1csmf07Uzelyn5nftzE4n53ftTE7+1phupVQn3CLayfk7OXuE/LPRSL/2nCkHd+/eHf39/S3PMzg4GH19fS0ft958ZY/o7PwLkT1C/qm4dmYm/+RcOzOTf3KunZnJP7lOvnYiOju/a2d2Ojl/J2ePkH8qC5V/OrO5hbLIJU8n5+/k7BHyz1Yj/ZrbigEAAAAgU8+ZlYOzWW5drVZjYGAgIiI2bdpUmOXujSwVzyl/kZbqF/XcR3R2ftfO5OSfWQ7XTkRn53ftzI8c8ud27UR0dn7Xzux0cn7XzuTkb57VX+3Tydkj5J+tRvq17qbfrSBKpVJDy6IrlUrbl1GPajR7hPyt1MnXTkRn53fttFcn58/t2ono7PydnD1C/lZy7bRPbvmLlD2is/O7dtqr0/NP55Zbbunoz43r5PydnD1C/vngtmIAAABgwU1VghS93BnVyfk7OXuE/K2mHAQAAACATCkHAQAAgLaYuFKqU1Z+jerk/J2cPUL+VnrOfOYgAAAA0Hk6rdSZqJPzd3L2CPlbxcpBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFN+IQkAAADQNuvWrRt7XJRf0NCITs7fydkj5G8VKwcBAACAtqgvRyZ7XnSdnL+Ts0fI30rKQQAAAADIlHIQAAAAWHBTrZTqlBVgnZy/k7NHyN9qykEAAABgQc1UghS95Onk/J2cPUL++aAcBAAAAIBMKQcBAAAAIFPKQQAAAGBB3XLLLU3tb7dOzt/J2SPknw/KQQAAAGDBTVWCFL3cGdXJ+Ts5e4T8raYcBAAAAIBMKQcBAACAtpi4UqpTVn6N6uT8nZw9Qv5W6m7bOwMAAADZ67RSZ6JOzt/J2SPkbxUrBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBM+YUkAAAAAHO0bt26scdF+QUTs9XJ2SPkbxUrBwEAAADmoL7cmex5kXVy9gj5W0k5CAAAAACZUg4CAAAALZVSit27d8/q6/HHH4/HH398VsemlNr9Rxsz1UqvTljB1snZI+RvNZ85CAAAALRUtVqN/v7+lo87ODgYfX19LR+3UTOVOOvWrSvsZ+B1cvYI+eeDlYMAAAAAkCkrBwEAAICWqlQqMTg4OONx1Wo1BgYGIiJi06ZNUalUZhwXaC0rBwEAAICWKpVK0dfXN+NXfdlXqVRmPL5UKrXxT/WsmW77LPJtrZ2cPUL++aAcBAAAAGjQVCVO0cupiM7OHiF/qykHAQAAACBTykEAAACAOZi40qtTVq5FdHb2CPlbyS8kAQAAAJijTiul6nVy9gj5W8XKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTfiEJAAAAwBytW7du7HFRfsFELjr93Bclv5WDAAAAAHNQX+5M9pz50+nnvkj5lYMAAAAAkCnlIAAAAECDplrpVYQVbCml2L1794xfjz/+eDz++OOzOnb37t2RUmr3Hy0iin3uZ6No+X3mIAAAAEADZipx1q1b19bPkKtWq9Hf39/ycQcHB6Ovr6/l4zai6Od+JkXMb+UgAAAAAGTKykEAAACA55BKpRKDg4PTHlOtVmNgYCAiIjZt2hSVSmVW4/LcY+UgAAAAQANmuu2z3be1lkql6Ovrm/arvuirVCozHt/X1xelUqmNf6p9in7uZ1LE/MpBAAAAgAZNVeIUvZx6Luj0c1+0/MpBAAAAAMiUchAAAABgDiau9OqUlWvPBZ1+7ouU3y8kAQAAAJijTiulnks6/dwXJb+VgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECm/EISAAAAADrOunXrxh4X5Zd7NKIo+a0cBAAAAKCj1Bdrkz0vuiLlb6gcvPzyy+OII46IpUuXxtKlS+PYY4+Nr33ta2P7zzzzzCiVSuO+jjnmmJaHBgAAAACa19BtxQcddFBcfPHF8ZKXvCQiIq688sro7++Pe+65Jw477LCIiHj9618fV1xxxdhrent7WxgXAAAAYH6llKJarc7q2NHjKpXKjMdWKpUolUpNZWPqVXbr1q3riNuLi5a/oXLw1FNPHff84x//eFx++eVxxx13jJWD5XI5Vq5cOesxa7Va1Gq1sec7d+5sJBIAAABAS1Wr1ejv72/5uIODg9HX19fycXMy0+23RS8Ii5h/zp85ODIyEtdcc008/fTTceyxx45t37JlSyxfvjwOPfTQePe73x2PPvrotONs3Lgxli1bNva1Zs2auUYCAAAAABrQ8G8r3rZtWxx77LFRrVbjec97XnzlK1+Jl7/85RERsX79+njLW94Sa9euje3bt8dHPvKROOmkk+Luu++Ocrk86XgXXHBBnH/++WPPd+7cqSAEAAAA2qZSqcTg4OCMx1Wr1RgYGIiIiE2bNs14a/Fsbj2GhdZwOfjSl7407r333njiiSfiy1/+cpxxxhmxdevWePnLXz72FyIi4vDDD4+jjjoq1q5dGzfccEOcdtppk45XLpenLA4BAAAAFlqpVGr49t9KpeKW4QVwyy23THtrbpFvKY4oZv6Gbyvu7e2Nl7zkJXHUUUfFxo0b48gjj4xPfepTkx67atWqWLt2bTzwwANNBwUAAACAqQq0oheDo4qWf86fOTgqpTTuF4rUe+yxx+Khhx6KVatWNfs2AAAAAECLNVQOfvjDH45vfOMb8eCDD8a2bdviwgsvjC1btsTb3/72eOqpp+IDH/hAfOtb34oHH3wwtmzZEqeeemq84AUviDe96U3zlR8AAACAzExcZdcpqwZHFSl/Q585+LOf/SxOP/30eOSRR2LZsmVxxBFHxE033RQnn3xy7N69O7Zt2xZXXXVVPPHEE7Fq1ao48cQTY9OmTbFkyZL5yg8AAABAhjqtEJyoKPkbKgc///nPT7mvr68vbr755qYDAQAAAAALo+nPHAQAAAAAOpNyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMdbc7AAAAAABERKSUolqtzurY0eMqlcqMx1YqlSiVSk1lm41OzK8cBAAAAKAQqtVq9Pf3t3zcwcHB6Ovra/m4E3VifrcVAwAAAECmrBwEAAAAoBAqlUoMDg7OeFy1Wo2BgYGIiNi0adOMt+bO5tbdVujE/MpBAAAAAAqhVCo1fPtspVJZkFuGZ6MT87utGAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADLVUDl4+eWXxxFHHBFLly6NpUuXxrHHHhtf+9rXxvanlGLDhg2xevXq6OvrixNOOCHuu+++locGAAAAAJrXUDl40EEHxcUXXxx33XVX3HXXXXHSSSdFf3//WAH4iU98Ii655JK47LLL4s4774yVK1fGySefHLt27ZqX8AAAAADA3DVUDp566qnxG7/xG3HooYfGoYceGh//+Mfjec97Xtxxxx2RUopLL700LrzwwjjttNPi8MMPjyuvvDKeeeaZuPrqq+crPwAAAAAwR3P+zMGRkZG45ppr4umnn45jjz02tm/fHjt27Ih169aNHVMul+P444+P22+/fcpxarVa7Ny5c9wXAAAAADD/Gi4Ht23bFs973vOiXC7He9/73vjKV74SL3/5y2PHjh0REbFixYpxx69YsWJs32Q2btwYy5YtG/tas2ZNo5EAAAAAgDlouBx86UtfGvfee2/ccccd8b73vS/OOOOM+F//63+N7S+VSuOOTyntt63eBRdcEE8++eTY10MPPdRoJAAAAABgDrobfUFvb2+85CUviYiIo446Ku6888741Kc+FR/60IciImLHjh2xatWqseMfffTR/VYT1iuXy1EulxuNAQAAAAA0ac6fOTgqpRS1Wi0OPvjgWLlyZWzevHls3549e2Lr1q1x3HHHNfs2AAAAAECLNbRy8MMf/nCsX78+1qxZE7t27YprrrkmtmzZEjfddFOUSqU477zz4qKLLopDDjkkDjnkkLjoooti8eLF8ba3vW2+8gMAAAAAc9RQOfizn/0sTj/99HjkkUdi2bJlccQRR8RNN90UJ598ckREfPCDH4zdu3fHWWedFY8//ngcffTRccstt8SSJUvmJTwAAAAAMHcNlYOf//znp91fKpViw4YNsWHDhmYyAQAAAAALoOnPHAQAAAAAOpNyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU93tDjCTlFJUq9WWjFU/TqvGjIioVCpRKpUm3Sf/sxY6fydnBwAAAFgIhS8Hq9Vq9Pf3t3zcgYGBlo01ODgYfX19k+6Tf3ILkb+TswMAAAAsBLcVAwAAAECmCr9ysN5nfuMtUe6ae+SUUuwZGYmIiN6urqZu56yNDMdZN36podf89W+cGOXurjm/5778eyMiordrUXP5h0fi7Btvbeg1l56yKspdc3/PfflTRET0dpWaPP8pzrv5kVkf///+Rm/0zv3UR0ophvZdOtHTFU1l3zMS8bEb98w9DAAAAECLdFQ5WO7qjkp3T1Nj9DX38qaUu7ui0t3cKW9r/q5SlLubW2xaaVn+vQ0d3dsV0dvdzGf7laLcsuypVQMBAAAANMVtxQAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZ6m53AGByKaWoVqstGat+nFaNGRFRqVSiVCpNuq+T83dC9ojpzz8AAADMhnIQCqparUZ/f3/Lxx0YGGjZWIODg9HX1zfpvk7O3wnZI6Y//wAAADAbbisGAAAAgExZOQgd4D++oTt6mvjbmlKK4ZF9j7u7oqlbUYeGI/7q+uGGXnPGGxcVKv+V1+2d9fGnvTmiu6nsESO/yN7VFdHsXcDDwxHXfrm5MQAAAGCUchA6QE93RG93M61SKco9rUqTGn5FT3dET5P5e9uUv7u7uXIwIqKnZdkBAACgtdxWDAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlqqBzcuHFjvOpVr4olS5bE8uXL441vfGPcf//9444588wzo1Qqjfs65phjWhoaAAAAAGheQ+Xg1q1b4+yzz4477rgjNm/eHMPDw7Fu3bp4+umnxx33+te/Ph555JGxrxtvvLGloQEAAACA5nU3cvBNN9007vkVV1wRy5cvj7vvvjte85rXjG0vl8uxcuXK1iQEoCEppahWqy0Zq36cVo0ZEVGpVKJUKk26r5PztzJ7RHvOPwAAkJeGysGJnnzyyYiIOPDAA8dt37JlSyxfvjx+6Zd+KY4//vj4+Mc/HsuXL590jFqtFrVabez5zp07m4kEkL1qtRr9/f0tH3dgYKBlYw0ODkZfX9+k+zo5/3xlj1i48w8AAORlzr+QJKUU559/frz61a+Oww8/fGz7+vXr47/9t/8W//AP/xCf/OQn484774yTTjppXAFYb+PGjbFs2bKxrzVr1sw1EgAAAADQgDmvHDznnHPie9/7Xnzzm98ct71+ZcPhhx8eRx11VKxduzZuuOGGOO200/Yb54ILLojzzz9/7PnOnTsVhAAt8rqBiK4m1oinFDEysu9xV1dEM3eijgxHfH1TY6/5f363FIuayp9i7/C+x4u6o6lbafcOR9z7xTTr40u/1xXRM+e3i4h9+eMX+aPJ/DEUkf5mpLlAAADAc86cfuQ699xz4/rrr4/bbrstDjrooGmPXbVqVaxduzYeeOCBSfeXy+Uol8tziQHADLq6I7p7mvtsuSb7rTqzL9ZGLeqO6Goqfymit4mXj9Ng/p6IUpPnvtTC/GkO5x8AAHjua6gcTCnFueeeG1/5yldiy5YtcfDBB8/4msceeyweeuihWLVq1ZxDAgAAAACt19BnDp599tnxxS9+Ma6++upYsmRJ7NixI3bs2BG7d++OiIinnnoqPvCBD8S3vvWtePDBB2PLli1x6qmnxgte8IJ405veNC9/AAAAAABgbhpaOXj55ZdHRMQJJ5wwbvsVV1wRZ555ZnR1dcW2bdviqquuiieeeCJWrVoVJ554YmzatCmWLFnSstAAAAAAQPMavq14On19fXHzzTc3FQgAAAAAWBgN3VYMAAAAADx3KAcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyFR3uwMAAMWQUopqtdqSserHadWYERGVSiVKpdKk+zo9PwAAtINyEACIiH0lWH9/f8vHHRgYaNlYg4OD0dfXN+m+Ts8PAADt4LZiAAAAAMiUlYMAwH66zjwyomfu/w8xpRQxvHffk+5Fzd1KO7Q3Rv72uw29pPuMEyO6u+b8li3NPzwSw1feOvfXAwDAPFIOAgD761kUpZ65l2uliIje1kRJc3lRd1eUeub+z5y25wcAgAXitmIAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATHW3OwAAABEppahWqy0Zq36cVo0ZEVGpVKJUKu23vZOzAwDkTjkIAFAA1Wo1+vv7Wz7uwMBAy8YaHByMvr6+/bZ3cnYAgNy5rRgAAAAAMmXlIABAwXS/49SI7rn/My2lFDE88ovBupq7nXZ4OIav+vtZH95z+m8XKvvQF/5u7q8HAMiAchAAoGi6u6PUM/d/ppUiInp7WhIlNfqC7u4o9cz9vfdln/PLx2k4OwBAhtxWDAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJCp7nYHAACAdkopRbVabclY9eO0asyIiEqlEqVSadJ9nZ4fAGgv5SAAAFmrVqvR39/f8nEHBgZaNtbg4GD09fVNuq/T8wMA7eW2YgAAAADIlJWDAADwC72nnxHR3TPn16eUIoaH9z3p7m7uVtrhodjzhSsbeknf6WcVKv/uL3xm7q8HABaEchAAAEZ190SpZ+7lWikiore3JVHSXF7U3ROlnrm//7785Tm/vt6c8gMAC85txQAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKa62x0AAADIU0opqtVqS8aqH6dVY0ZEVCqVKJVKk+7r9PwAEKEcBAAA2qRarUZ/f3/Lxx0YGGjZWIODg9HX1zfpvk7PDwARbisGAAAAgGxZOQgAALTdAe/44yh198759SmliOGhfU+6e5q6lTYN74mnr7q4odf80hl/EaXu8tzfM6WI4T37nnT3Npm/Fk9c+Udzfj0AeVEOAgAAbVfq7o1Sz9zLwVJERO/cy7lmlbrLUeqZ+/vvy19pWR4AmC23FQMAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmGioHN27cGK961atiyZIlsXz58njjG98Y999//7hjUkqxYcOGWL16dfT19cUJJ5wQ9913X0tDAwAAAADNa6gc3Lp1a5x99tlxxx13xObNm2N4eDjWrVsXTz/99Ngxn/jEJ+KSSy6Jyy67LO68885YuXJlnHzyybFr166WhwcAAAAA5q67kYNvuummcc+vuOKKWL58edx9993xmte8JlJKcemll8aFF14Yp512WkREXHnllbFixYq4+uqr4z3veU/rkgMAADBnKaWoVqstGat+nFaNGRFRqVSiVCq1bDwA9tdQOTjRk08+GRERBx54YEREbN++PXbs2BHr1q0bO6ZcLsfxxx8ft99++6TlYK1Wi1qtNvZ8586dzUQCAABgFqrVavT397d83IGBgZaNNTg4GH19fS0bD4D9zfkXkqSU4vzzz49Xv/rVcfjhh0dExI4dOyIiYsWKFeOOXbFixdi+iTZu3BjLli0b+1qzZs1cIwEAAAAADZjzysFzzjknvve978U3v/nN/fZNXPadUppyKfgFF1wQ559//tjznTt3KggBAAAW0IlnXBZdPeU5vz6lFHuH90RExKLu3qZuBR4ZqsWtV54z59cD0Jg5lYPnnntuXH/99XHbbbfFQQcdNLZ95cqVEbFvBeGqVavGtj/66KP7rSYcVS6Xo1ye+yQEAABAc7p6ytHdRDkYERG9ldaEAWBBNXRbcUopzjnnnLj22mvjH/7hH+Lggw8et//ggw+OlStXxubNm8e27dmzJ7Zu3RrHHXdcaxIDAAAAAC3R0MrBs88+O66++uoYHByMJUuWjH2O4LJly6Kvry9KpVKcd955cdFFF8UhhxwShxxySFx00UWxePHieNvb3jYvfwAAAAAAYG4aKgcvv/zyiIg44YQTxm2/4oor4swzz4yIiA9+8IOxe/fuOOuss+Lxxx+Po48+Om655ZZYsmRJSwIDAAAAAK3RUDmYUprxmFKpFBs2bIgNGzbMNRMAAAAAsAAa+sxBAAAAAOC5QzkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKa62x0AAAAAGpFSimq12pKx6sdp1ZijKpVKlEql/bZ3Qv6psgPPPcpBAAAAOkq1Wo3+/v6WjzswMNDS8QYHB6Ovr2+/7Z2Qf6rswHOP24oBAAAAIFNWDgIAANCx3vGOy6O7uzzn16eUYnh4T0REdHf3Nn0r7fBwLa666n2zPv79b/lM9DSZf2hkX/6erubyDw3X4lNfOmvOrwc6k3IQAACAjtXdXY6enkpTY/T2tu/22Z7ucvQ2mb8cbv8F5s5txQAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZ6m53AAAAAKBzpJSiWq22ZKz6cVo1ZkREpVKJUqnUsvHguUw5CAAAAMxatVqN/v7+lo87MDDQsrEGBwejr6+vZePBc5nbigEAAAAgU1YOAgAAAHOy4bf+Onq7y3N+fUophkb2RERET1dvU7cC7xmuxYavnj3n10OulIMAAADAnPR2l6PcXWlqjEqP23+hndxWDAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJCp7nYHmElKaexxbXi4jUnGq89Sn3Gi8flH5jVTI+qzzD7/3nnN1Ij6LNPlBwAAgFEppahWqy0br36sVo5bqVSiVCrtt72V+Rc6O8VV+HKwVquNPT7ra19qY5Kp1Wq1WLx48ZT7Rp19460LFakhs81/3i07FipSQ6bLDwAAAKOq1Wr09/fPy9gDAwMtG2twcDD6+vr22z5f+RciO8XltmIAAAAAyFThVw6Wy+Wxx59Z/5Yodxcjcm14eGwlY33Gier3/fVvnBjl7q55zzYbteGRsZWMs81/6bqVUe4uRp9cG947tpJxuvwAAAAwmU+/5tNR7mru58mUUuzZuyciInoX9TZ1O21tpBbn3nburI+/7MQ/jXJX75zfb1/2oYiI6F3U02T2PXHOrR+d8+tpr2I0bdOovzjL3d1R6e5pY5rJTfcXaHz+rqgUpNysN/v8iwpTDtbzWQYAAAA0qtxVbrocjIioRKUFaRpX7uqNcndz+duVnWIpXtMDAAAAACwI5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJnqbncAmC8ppbHHe4bTNEcurPos9Rknqt83VKD8Q5nkBwAAoLVSSlGtVlsyVv04rRozIqJSqUSpVJp0X6fnn4pykOesWq029vhjXxtqY5Kp1Wq1WLx48ZT7Rv3V9SMLFakhs81/5XUpIopXxE2XHwAAgNaqVqvR39/f8nEHBgZaNtbg4GD09fVNuq/T80/FbcUAAAAAkCkrB3nOKpfLY4//3/U90dvd2LLa+bJnOI2tZKzPOFH9vv/4hq7oKUj+oeE0tpJxtvnPeGOpUPn3rWScPj8AAADz57KT/zDKXT1zfn1KKfaM7PvZurerp+FbaevVRobinM2fbOg1f33Ku1uQfzgiInq7upvOf/bNn5vz65WDPGfV/8Xq7S4VphysN91f/vp9Pc+B/EUpB/fZVw42880XAACAuSt39USlu7epMfp62rfgY1/+uZeDERF9Pc39+VvFbcUAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGSq4XLwtttui1NPPTVWr14dpVIprrvuunH7zzzzzCiVSuO+jjnmmFblBQAAAABapLvRFzz99NNx5JFHxjvf+c5485vfPOkxr3/96+OKK64Ye97b2zv3hAALLKU09nh4uI1BJlGfpz5nvf3zT37cQptN9on7RoaKkT1ifJbZnPtUoOwR4/NMd/4BAIC8NFwOrl+/PtavXz/tMeVyOVauXDmr8Wq1WtRqtbHnO3fubDQSQEvVf0+69sttDDKDWq0WixcvnnT7qP9v00Immr2pso/uG/Xd/xZRlHKz3mzOffzN3gIm32e68w8AAORlXj5zcMuWLbF8+fI49NBD493vfnc8+uijUx67cePGWLZs2djXmjVr5iMSAAAAADBBwysHZ7J+/fp4y1veEmvXro3t27fHRz7ykTjppJPi7rvvjnK5vN/xF1xwQZx//vljz3fu3KkgBNqq/nvVaW+O6G75d8q5Gx5+djXjZN9TJ25/7UBx8g8PP7uScarsE/cd+faIrp7SfEeblZGh9IuVjLM79/F7i6JUkOwRv7it+G/2RsT05x8AAMhLy39kHBgYGHt8+OGHx1FHHRVr166NG264IU477bT9ji+Xy35IAQqlVHq20OnuLk65NlF9zqm2d3dHdBemoHr2Jtupsk/c19VTKkw5uM++P8Nszn2pp1SocjDi2f8C051/AAAgL/NyW3G9VatWxdq1a+OBBx6Y77cCAAAAABow7+XgY489Fg899FCsWrVqvt8KAAAAAGhAwzfLPfXUU/HDH/5w7Pn27dvj3nvvjQMPPDAOPPDA2LBhQ7z5zW+OVatWxYMPPhgf/vCH4wUveEG86U1vamlwAAAAAKA5DZeDd911V5x44oljz0d/mcgZZ5wRl19+eWzbti2uuuqqeOKJJ2LVqlVx4oknxqZNm2LJkiWtSw0AAAAANK3hcvCEE06IlNKU+2+++eamAgEAAAAAC2PeP3MQAAAAACgm5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmepudwAAoBhSSs8+HhppY5Lx6rPUZ9zvuA7PDwAA7aAcBAAiIqJWq4093vu332tjkqnVarVYvHjxlPtGjVx160JFash0+QEAoB3cVgwAAAAAmbJyEACIiIhyuTz2eNGZR0Spp6uNaZ6VhkbGVjLWZ5yofl/XO04sVP7RlYzT5QcAgHZQDgIAERFRKpWefdzTVZhyrV59xun27ctfvH/mTJcfAADawW3FAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJnqbncAAAAiUkrPPh4abmOS8eqz1Gccd0wHZ5+4Lw0NzWumRtRnmS4/AEAzlIMAAAVQq9XGHo984e/bmGRqtVotFi9ePOn2UcNf/LuFjDRrU2Uf3Tdq6ItXLlSkhkyXHwCgGW4rBgAAAIBMWTkIAFAA5XJ57HHX6adGqacY/0xLQ8NjKxnrM9ar3979u79dqOyjKxmnyj5xX8/vnhGlnp55zzYbaWhobCXjdPkBAJpRjH+5AQBkrlQqPfu4p7swBVu9+oxTbd+XvRjlWr2psk/cV+rp6bj8AADNcFsxAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKa62x0AAACYu5TSs4+HhtqYZLz6LPUZxx0zLvueec80W/VZpsoOAM8VykEAAOhgtVpt7HH1i59pY5Kp1Wq1WLx48aTbRz39hYsXMtKsTZUdAJ4r3FYMAAAAAJmychAAADpYuVwee1z53bOi1NPTxjTPSkNDYysZ6zPWq99+wOl/HKWe3gXJNpM0tGdsJeNU2QHguUI5CAAAHaxUKj37uKenMAVbvfqMU20v9fR2VHYAeK5wWzEAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAprrbHQAAAKATpZSefTxUa2OS8eqz1Gfc77i6fSMFyj8yy/wAtIZyEAAAYA5qtWdLrCeu+qM2JplarVaLxYsXT7lv1K1XnrNQkRoyXX4AWsNtxQAAAACQKSsHAQAA5qBcLo89/qV3/EWUesrTHL1w0lBtbCVjfcaJ6vedeMZl0VWQ/CNDtbGVjNPlB6A1lIMAAABzUCqVnn3cUy5MOVivPuN0+7p6ytHdYfkBaA23FQMAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmGi4Hb7vttjj11FNj9erVUSqV4rrrrhu3P6UUGzZsiNWrV0dfX1+ccMIJcd9997UqLwAAAADQIg2Xg08//XQceeSRcdlll026/xOf+ERccsklcdlll8Wdd94ZK1eujJNPPjl27drVdFgAAAAAoHW6G33B+vXrY/369ZPuSynFpZdeGhdeeGGcdtppERFx5ZVXxooVK+Lqq6+O97znPc2lBQAAIHsppbHHQ0O1NibZX32e+pz16rfvGS5O/vosU2WfuK/T8tdvr40UJ3vE+Dyzy79n3jPNVn2W6a4diqnhcnA627dvjx07dsS6devGtpXL5Tj++OPj9ttvn7QcrNVqUas9+xdg586drYwEAADAc0z9z5Bf+ML72phkerVaLRYvXjzp9lF/9aWzFjLSrE2VfXTfqA1fPXuhIjVkNuf+3NvOXchIDZlN/nNu/ehCRpq16a4diqmlv5Bkx44dERGxYsWKcdtXrFgxtm+ijRs3xrJly8a+1qxZ08pIAAAAAMAUWrpycFSpVBr3PKW037ZRF1xwQZx//vljz3fu3KkgBAAAYErlcnns8emnXx49PeVpjl5YQ0O1sdWM9Tnr1W//j2/5TPR2FyP/nuHa2ErGqbJP3Lfht/66UPlHVzLO5tx/+jWfjnJXMbJH7LuteHQ142zyX3bin0a5q3dBss2kNrJnbCXjdNcOxdTScnDlypURsW8F4apVq8a2P/roo/utJhxVLpddOAAAAMxa/eKTnp5y9PRU2phmalMtkqnf3ttdjt4C5p8q+8R9vd3lKHd3Tv767eWucqHKwXqzy98b5YIUs/Wmu3YoppbeVnzwwQfHypUrY/PmzWPb9uzZE1u3bo3jjjuulW8FAAAAADSp4ZWDTz31VPzwhz8ce759+/a4995748ADD4xf+ZVfifPOOy8uuuiiOOSQQ+KQQw6Jiy66KBYvXhxve9vbWhocAAAAAGhOw+XgXXfdFSeeeOLY89HPCzzjjDPib//2b+ODH/xg7N69O84666x4/PHH4+ijj45bbrkllixZ0rrUAAAAAEDTGi4HTzjhhEgpTbm/VCrFhg0bYsOGDc3kAgAAAADmWUs/cxAAAAAA6BzKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAy1d3uAI2ojQw39fqUUuwZGYmIiN6uriiVSm3LAgAAAJCTlNLY49rwnjYmGa8+S33GicbnH5rXTI2ozzJd/ql0VDl41o1fancEAAAAAOagVquNPT7n65e0McnUarVaLF68eMp9o86+5XMLFakh0+WfituKAQAAACBThV85WKlUYnBwsCVjVavVGBgYiIiITZs2RaVSacm4rRoHAAAA4LmqXC6PPb7sdedHubu3jWmeVRveM7aSsT7jRPX7/nrdu6Pc3TPv2WajNjw0tpJxuvxTKXw5WCqVoq+vr+XjViqVeRkXAAAAgP3V/+6HcndvVApSDtab7vdTjM/fE5WClIP15vL7NdxWDAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZKq73QFyUhseaer1KaXYM7I3IiJ6uxZFqVRa0Cy1kRQRe+f8nvvyp4iI6O0qNZf/F+PM1p6RiIjGXlMvpRRDvzhlPV3RVPY9c7gMhoYjms0/+p+8u8n8+7LM5TWdmx8AAACeq5SDC+jsG29td4SmnHfzI+2OMGcfu3FPuyM05a+u7+xG68rr5l4qT27uRSMAAADwLLcVAwAAAECmrBycZ5VKJQYHB1syVrVajYGBgYiI2LRpU1QqlZaMO904nZy/k7OP7pN/n3bkBwAAgBwoB+dZqVSKvr6+lo9bqVTmZdyJOjl/J2ePkH8qC5UfAAAAcuC2YgAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMdbc7AECRDQ839/qUIkZG9j3u6ooolRY2z8hwRESa8/u1Mv/IHM7l3qbzp1+MEbGoO6LUxB9gb6P5hyJSE9kj9uWP0fdtMn8MNXr83qbS78u+d9+T7kVNZt/b+GuGR4qTf3ikiSQAADC/lIMA07j2y+1O0Jyvb2p3gubc+8XmyrX9tXq8ad7pb1pfCC1c+oiRv/3uAr5b6w1feWu7IwAAQEdwWzEAAAAAZMrKQYAJKpVKDA4OtmSsarUaAwMDERGxadOmqFQqLRk3IqYcqxPyTzdOJ+dvZfaIzs7v2plcK78HAABAKygHASYolUrR19fX8nErlcq8jDuR/JNbiPzzlT2is/O7dgAAoLjcVgwAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQqe52BwAAYILh4UhNvDylFDE8su9Jd1eUSqWmsjR6fMdmj4gYHmpB/l+8b3d3k/mH5vSajs4PACw45SAAQMEMX/X37Y4wZ0Nf+Lt2R2jKni9c2e4ITdn9hc+0OwIA0GHcVgwAAAAAmbJyEACgACqVSgwODrZkrGq1GgMDAxERsWnTpqhUKi0Zd6pxOjn76D7592lHfgCgvZSDAAAFUCqVoq+vr+XjViqVeRm3Xidnj5B/KguVHwBoL7cVAwAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKZaXg5u2LAhSqXSuK+VK1e2+m0AAAAAgCZ1z8eghx12WHz9618fe97V1TUfbwMAAAAANGFeysHu7m6rBQEAgFlLw3uae31KEcND+55090SpVFrQLGm4Nuf3ixjN/4v37e5tMn/jWUaGms+/9xf5FzWZv9Eswy0498O/yN7dZPa55BlqQf6hkX35e7qayz+XLHsKlL/RLLWR5rJH7Mu/Z+++/L2LmsvfijwwF/NSDj7wwAOxevXqKJfLcfTRR8dFF10UL37xiyc9tlarRa327F+AnTt3zkckAACgwJ6+6uJ2R2jKE1f+UbsjNOXWK89pd4Q5u+qq97U7QlM+9aWz2h2hKRu+ena7I8zZubed2+4IUAgt/8zBo48+Oq666qq4+eab43Of+1zs2LEjjjvuuHjssccmPX7jxo2xbNmysa81a9a0OhIAAAAAMImWrxxcv3792ONXvOIVceyxx8a/+lf/Kq688so4//zz9zv+ggsuGLd9586dCkIAAMhApVKJwcHBloxVrVZjYGAgIiI2bdoUlUqlJeNON478z1ro/J2QPaKz87t2Zqcd5x9abV5uK653wAEHxCte8Yp44IEHJt1fLpejXC7PdwwAAKBgSqVS9PX1tXzcSqUyL+NOJP/kFiJ/J2ePkH8qnXztRCzc+YdWa/ltxRPVarX4/ve/H6tWrZrvtwIAAAAAGtDycvADH/hAbN26NbZv3x7f/va349//+38fO3fujDPOOKPVbwUAAAAANKHltxX/y7/8S7z1rW+Nn//85/HCF74wjjnmmLjjjjti7dq1rX4rAAAAAKAJLS8Hr7nmmlYPCQAAAADMg3n/zEEAAAAAoJiUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkqrvdAQAAAABoTG1kT1OvTynFnr1DERHRu6gnSqXSgmapjQzN+f0ifpH/F2P0djWbv/Esrck/HBERvV3dC56/nnIQAAAAoMOcc+tH2x2hKeds/mS7IzTl7Js/1+4ILeO2YgAAAADIlJWDAAAAAB2gUqnE4OBgS8aqVqsxMDAQERGbNm2KSqXSknGnG0f+Z7Uj/1SUgwAAAAAdoFQqRV9fX8vHrVQq8zLuRPJPbqHyT8VtxQAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZEo5CAAAAACZUg4CAAAAQKaUgwAAAACQKeUgAAAAAGRKOQgAAAAAmVIOAgAAAECmlIMAAAAAkCnlIAAAAABkSjkIAAAAAJlSDgIAAABAppSDAAAAAJAp5SAAAAAAZGreysHPfOYzcfDBB0elUolXvvKV8Y1vfGO+3goAAAAAmIN5KQc3bdoU5513Xlx44YVxzz33xK//+q/H+vXr4yc/+cl8vB0AAAAAMAfd8zHoJZdcEu9617vi93//9yMi4tJLL42bb745Lr/88ti4ceN8vGWklKJarU57TP3+mY4dValUolQqNZVtJrPJHpFf/oXIHtHZ105EZ+d37UxO/pnlcO1EdHZ+1878yCF/btdORGfnd+3MTifnd+1MTv6Z5XDtRHR2ftfO/OjE/KWUUmrlgHv27InFixfHl770pXjTm940tv39739/3HvvvbF169Zxx9dqtajVamPPd+7cGWvWrIknn3wyli5dOuv33b17d/T39zf/B5hgcHAw+vr6Wj5uvfnKHtHZ+Rcie4T8U3HtzEz+ybl2Zib/5Fw7M5N/cp187UR0dn7Xzux0cv5Ozh4h/1RcOzOTf3KunZkVJf/OnTtj2bJls+rXWn5b8c9//vMYGRmJFStWjNu+YsWK2LFjx37Hb9y4MZYtWzb2tWbNmlZHAgAAAAAm0fKVgw8//HC86EUvittvvz2OPfbYse0f//jH4wtf+EL87//9v8cd36qVg40u26xUKrMat0hLZiPyyl/UJb9FOvcRnZ3ftTM5+WeWy7UT0dn5XTutl0v+nK6d0eM6Nb9rZ3Y6Ob9rZ3LyzyyXayeis/O7dlqvKPkbWTnY9tuKJ2okPAAAAAAwXltvK+7t7Y1XvvKVsXnz5nHbN2/eHMcdd1yr3w4AAAAAmKN5+W3F559/fpx++ulx1FFHxbHHHhuf/exn4yc/+Um8973vnY+3AwAAAADmYF7KwYGBgXjsscfiYx/7WDzyyCNx+OGHx4033hhr166dj7cDAAAAAOag5Z852CyfOQgAAAAAc9fWzxwEAAAAADqDchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU8pBAAAAAMiUchAAAAAAMqUcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEx1tzvARCmliIjYuXNnm5MAAAAAQOcZ7dVGe7bpFK4c3LVrV0RErFmzps1JAAAAAKBz7dq1K5YtWzbtMaU0mwpxAe3duzcefvjhWLJkSZRKpZaPv3PnzlizZk089NBDsXTp0paPP9/kb59Ozh4hfzt1cvYI+dutk/N3cvYI+dupk7NHyN9OnZw9Qv526uTsEfK3Wyfn7+TsEfJPJ6UUu3btitWrV8eiRdN/qmDhVg4uWrQoDjrooHl/n6VLl3bkhTNK/vbp5OwR8rdTJ2ePkL/dOjl/J2ePkL+dOjl7hPzt1MnZI+Rvp07OHiF/u3Vy/k7OHiH/VGZaMTjKLyQBAAAAgEwpBwEAAAAgU9mVg+VyOT760Y9GuVxud5Q5kb99Ojl7hPzt1MnZI+Rvt07O38nZI+Rvp07OHiF/O3Vy9gj526mTs0fI326dnL+Ts0fI3yqF+4UkAAAAAMDCyG7lIAAAAACwj3IQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyNRzshzcuHFjvOpVr4olS5bE8uXL441vfGPcf//9+x33/e9/P97whjfEsmXLYsmSJXHMMcfET37ykzYkHm82+Uul0qRff/EXf9Gm1M+aTf6nnnoqzjnnnDjooIOir68v/vW//tdx+eWXtynxeLPJ/7Of/SzOPPPMWL16dSxevDhe//rXxwMPPNCmxM+6/PLL44gjjoilS5fG0qVL49hjj42vfe1rY/tTSrFhw4ZYvXp19PX1xQknnBD33XdfGxOPN1P+a6+9Nk455ZR4wQteEKVSKe699972hZ3EdPmHhobiQx/6ULziFa+IAw44IFavXh3veMc74uGHH25z6n1mOvcbNmyIl73sZXHAAQfE85///Hjd614X3/72t9uYeLyZ8td7z3veE6VSKS699NKFDTmNmfKfeeaZ+32/P+aYY9qY+FmzOfdFnW8jZs5f5Pk2Yub8RZ5vI2bOX9T5djIbN26MUqkU55133ti2os+79SbLX/R5d9TE7EWfcyea7NwXfd6tN1n+ekWcd0dNlr3Ic+5EU537Is+79SbLX/R5d9Rk2Ys+59abLH+R59wNGzbsd02sXLlybH/R59uZ8hdhvn1OloNbt26Ns88+O+64447YvHlzDA8Px7p16+Lpp58eO+ZHP/pRvPrVr46XvexlsWXLlvjud78bH/nIR6JSqbQx+T6zyf/II4+M+/qbv/mbKJVK8eY3v7mNyfeZTf4/+IM/iJtuuim++MUvxve///34gz/4gzj33HNjcHCwjcn3mSl/Sine+MY3xo9//OMYHByMe+65J9auXRuve93rxv0Z2+Gggw6Kiy++OO66666466674qSTTor+/v6xb4yf+MQn4pJLLonLLrss7rzzzli5cmWcfPLJsWvXrrbmHjVT/qeffjr+3b/7d3HxxRe3Oenkpsv/zDPPxHe+8534yEc+Et/5znfi2muvjR/84Afxhje8od2xI2Lmc3/ooYfGZZddFtu2bYtvfvOb8au/+quxbt26+D//5/+0Ofk+M+Ufdd1118W3v/3tWL16dZuSTm42+V//+teP+75/4403tjHxs2bKXuT5NmLm/EWebyNmzl/k+TZi+vxFnm8nuvPOO+Ozn/1sHHHEEeO2F33eHTVV/qLPuxGTZy/6nFtvqnNf9Hl31FT5RxV13o2YPntR59x6U+Uv+rw7aqr8RZ93I6bOXvQ5d9Rk+Tthzj3ssMPGXRvbtm0b29cJ8+10+Qsx36YMPProoyki0tatW8e2DQwMpN/93d9tY6rZmyz/RP39/emkk05awFSzN1n+ww47LH3sYx8bd9y/+Tf/Jv3Jn/zJQseb0cT8999/f4qI9E//9E9jxwwPD6cDDzwwfe5zn2tXzCk9//nPT//1v/7XtHfv3rRy5cp08cUXj+2rVqtp2bJl6T//5//cxoTTG81fb/v27Ski0j333NOeUA2YLP+of/zHf0wRkf75n/95gVPNznTZn3zyyRQR6etf//oCp5q9ifn/5V/+Jb3oRS9K//RP/5TWrl2b/vIv/7J94WahPv8ZZ5yR+vv72xuoAfXZO2m+HTXdtV/k+XZUff5Omm9HjebvlPl2165d6ZBDDkmbN29Oxx9/fHr/+9+fUkodM+9Olb9eUefd2WQfVcQ5t5H8RZx3Z8pf5Hl3uuydMOdOl78T5t1Grv2izbvTZe+EOXeq/EWfcz/60Y+mI488ctJ9nTDfTpe/Xjvn2+fkysGJnnzyyYiIOPDAAyMiYu/evXHDDTfEoYceGqecckosX748jj766LjuuuvamHJqE/NP9LOf/SxuuOGGeNe73rWQsWZtsvyvfvWr4/rrr4+f/vSnkVKKW2+9NX7wgx/EKaec0q6YU5qYv1arRUSM+79vXV1d0dvbG9/85jcXPuAURkZG4pprromnn346jj322Ni+fXvs2LEj1q1bN3ZMuVyO448/Pm6//fY2Jp3cxPydZjb5n3zyySiVSvFLv/RLCxtuBjNl37NnT3z2s5+NZcuWxZFHHtmGhNObLP/evXvj9NNPjz/6oz+Kww47rM0JpzfV+d+yZUssX748Dj300Hj3u98djz76aBtTTm5i9k6bb2e69os+306Wv5Pm24n5O2W+Pfvss+M3f/M343Wve9247Z0y706VvxM0kr2Ic+5s8xd13p0uf9Hn3ZnOfdHn3Knyd8q8O9trv4jz7nTZO2HOnSp/J8y5DzzwQKxevToOPvjg+J3f+Z348Y9/HBGdM99Olb8wFryOXGB79+5Np556anr1q189tu2RRx5JEZEWL16cLrnkknTPPfekjRs3plKplLZs2dLGtPubLP9Ef/7nf56e//znp927dy9gstmZKn+tVkvveMc7UkSk7u7u1Nvbm6666qo2pZzaZPn37NmT1q5dm97ylrek//t//2+q1Wpp48aNKSLSunXr2ph2n+9973vpgAMOSF1dXWnZsmXphhtuSCml9D//5/9MEZF++tOfjjv+3e9+dyFyj5oqf72irmBIaXb5U0pp9+7d6ZWvfGV6+9vfvsAJpzZT9r//+79PBxxwQCqVSmn16tXpH//xH9uUdHLT5b/ooovSySefnPbu3ZtSSoVbwZDS9Pmvueaa9NWvfjVt27YtXX/99enII49Mhx12WKpWq21M/KypsnfKfDvbv7dFnW+ny98J8+1U+Ys+36aU0n//7/89HX744WPXRP0qjE6Yd6fLX6+I8+5ss6dUzDl3NvmLPO/OlL/I8+5M2Ys+506XvxPm3Ub+7hZt3p0pe9Hn3OnyF33OvfHGG9P/+B//I33ve98bW/W4YsWK9POf/7wj5tvp8tdr53z7nC8HzzrrrLR27dr00EMPjW376U9/miIivfWtbx137Kmnnpp+53d+Z6EjTmuy/BO99KUvTeecc84Cppq9qfL/xV/8RTr00EPT9ddfn7773e+mT3/60+l5z3te2rx5c5uSTm6q/HfddVc68sgjU0Skrq6udMopp6T169en9evXtynps2q1WnrggQfSnXfemf74j/84veAFL0j33Xff2DfNhx9+eNzxv//7v59OOeWUNqXd31T56xXxh5RRs8m/Z8+e1N/fn37t134tPfnkk21Kur+Zsj/11FPpgQceSN/61rfS7/3e76Vf/dVfTT/72c/amHi8qfLfddddacWKFeP+wVCkH1JGzebaGfXwww+nnp6e9OUvf3mBU05uquydMt/O9twXdb6dLn8nzLfT5S/yfPuTn/wkLV++PN17771j2yYrB4s6786Uv17R5t1Gshdxzp1t/qLOuzPlL/K828i1M6pIc+5M+Ys+7zZ6/os0784me5Hn3NnkL/KcO9FTTz2VVqxYkT75yU8Wfr6dTH3+esrBeXLOOeekgw46KP34xz8et71Wq6Xu7u70Z3/2Z+O2f/CDH0zHHXfcQkac1lT56912220pIsb9JS+KqfI/88wzqaenJ331q18dt/1d73pXof7yzub8P/HEE+nRRx9NKaX0b//tv01nnXXWQsWbtde+9rXpP/yH/5B+9KMfpYhI3/nOd8btf8Mb3pDe8Y53tCndzEbz1yvaDynTmZh/z5496Y1vfGM64ogj9vs/RUUz2bmv95KXvCRddNFFC5ioMaP5//Iv/zKVSqXU1dU19hURadGiRWnt2rXtjjml2Zz/+s9WKZLR7J0y30402bkv8nw70Wj+TplvJ5rs/Bdxvv3KV74y9gNU/feW0e83P/zhDws9786Uf3h4eOzYos27s81e1Dm3kXNfryjz7kz5/9N/+k+FnXebOfdFmHNnyl+tVgs97zZy/os2786U/amnnir0nNvIuS/inDuZ173udem9731vx/6cO5q/Xjvn2+5W3Z5cJCmlOPfcc+MrX/lKbNmyJQ4++OBx+3t7e+NVr3pV3H///eO2/+AHP4i1a9cuZNRJzZS/3uc///l45StfWajPH5kp/9DQUAwNDcWiReM/8rKrqyv27t27kFEn1cj5X7ZsWUTs+/yAu+66K/7sz/5soWLOWkoparVaHHzwwbFy5crYvHlz/Nqv/VpE7PsMm61bt8af//mftznl1Ebzd6r6/ENDQ/Hbv/3b8cADD8Stt94av/zLv9zmdNOb6dwX/b/NaL7TTz99v89VOeWUU+L000+Pd77znW1KN7Ppzu9jjz0WDz30UKxatWqBU83OaPaiz7dTmezcF3G+ncpo/qLPt1OZ7PwXcb597WtfO+43DUZEvPOd74yXvexl8aEPfShe/OIXF3renSl/V1dXm5LNbDbZizznzvXcF2XenSn/qlWr9vuMtaLMu3M590Wac2fKXy6XCz3vNnL+izbvzpR9ZGSk0HNuI+e+iHPuRLVaLb7//e/Hr//6r3fkz7n1+QtjwevIBfC+970vLVu2LG3ZsiU98sgjY1/PPPPM2DHXXntt6unpSZ/97GfTAw88kD796U+nrq6u9I1vfKONyfeZTf6U9v3WssWLF6fLL7+8TUknN5v8xx9/fDrssMPSrbfemn784x+nK664IlUqlfSZz3ymjcn3mU3+v/u7v0u33npr+tGPfpSuu+66tHbt2nTaaae1MfU+F1xwQbrtttvS9u3b0/e+97304Q9/OC1atCjdcsstKaWULr744rRs2bJ07bXXpm3btqW3vvWtadWqVWnnzp1tTr7PTPkfe+yxdM8996QbbrghRUS65ppr0j333JMeeeSRNiffZ7r8Q0ND6Q1veEM66KCD0r333jvu2qrVau2OPm32p556Kl1wwQXpW9/6VnrwwQfT3Xffnd71rnelcrk87jeatdNM185ERbm9adR0+Xft2pX+8A//MN1+++1p+/bt6dZbb03HHntsetGLXlSIv7sznfsiz7cpze7aKep8m9LM+Ys836Y0c/6izrdTmXiLVtHn3Ykm5i/6vFuvPnvR59zJ1OfvhHl3opluzS3avFuvPnvR59zJTDz3RZ93J5rs2inyvFtvYvaiz7kTTcxf5Dn3D//wD9OWLVvSj3/843THHXek3/qt30pLlixJDz74YEqp+PPtTPmLMN8+J8vBiJj064orrhh33Oc///n0kpe8JFUqlXTkkUem6667rj2BJ5ht/v/yX/5L6uvrS0888UR7gk5hNvkfeeSRdOaZZ6bVq1enSqWSXvrSl6ZPfvKTYx9a3E6zyf+pT30qHXTQQamnpyf9yq/8SvqTP/mTQvxj8/d+7/fS2rVrU29vb3rhC1+YXvva1477AXfv3r3pox/9aFq5cmUql8vpNa95Tdq2bVsbE483U/4rrrhi0v82H/3oR9sXus50+UeXiE/2deutt7Y3eJo+++7du9Ob3vSmtHr16tTb25tWrVqV3vCGNxTqg9FnunYmKtoPKdPlf+aZZ9K6devSC1/4wrHvOWeccUb6yU9+0ubU+8zm3Bd1vk1pdvmLOt+mNHP+Is+3Kc2cv6jz7VQm/qBV9Hl3oon5iz7v1qvPXvQ5dzL1+Tth3p3ouVIOFn3Oncxk577I8+5Ek+Uv8rxbb2L2os+5E03MX+Q5d2BgIK1atSr19PSk1atXp9NOO23c50MXfb6dKX8R5ttSSim1ciUiAAAAANAZFs18CAAAAADwXKQcBAAAAIBMKQcBAAAAIFPKQQAAAADIlHIQAAAAADKlHAQAAACATCkHAQAAACBTykEAAAAAyJRyEAAAAAAypRwEAAAAgEwpBwEAAAAgU/8/dUuoU/f95K0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1600x900 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "posicion_calidad = defaultdict(list)\n",
    "for rec in recs:\n",
    "    for i, calidad in enumerate(rec.letter_annotations['phred_quality']):\n",
    "        if i < 25 or calidad == 40:\n",
    "            continue\n",
    "        posicion= i + 1\n",
    "        posicion_calidad[posicion].append(calidad)\n",
    "vps = []\n",
    "poses = list(posicion_calidad.keys())\n",
    "poses.sort()\n",
    "for pos in poses:\n",
    "    vps.append(posicion_calidad[pos])\n",
    "fig, ax = plt.subplots(figsize=(16,9))\n",
    "sns.boxplot(data=vps,ax=ax)\n",
    "ax.set_xticklabels([str(x) for x in range(26, max(posicion_calidad.keys()) + 1)])\n",
    "pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b677252",
   "metadata": {},
   "source": [
    "## NCBI: Accesibilidad con entrez"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfa00925",
   "metadata": {},
   "source": [
    "OJO: Reglas a cumplir:\n",
    "\n",
    "\n",
    "        - especificar una dirección de correo con la solicitud\n",
    "        \n",
    "        - evitar alto número de solicitudes(100 o más) durante los tiempos de mayor concurrencia(de 9 de la mañana a 5 de la tarde respecto la zona horaria del este americano, entre las 3 de la mañana y las 11 de la mañana)\n",
    "        \n",
    "        - no publicar más de tres solicitudes por segundo(Biopython lo cumple automáticamente)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e56aee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4dbd0890",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email=\"gcs1005@alu.ubu.es\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32794898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['DbList'])\n"
     ]
    }
   ],
   "source": [
    "# el comando Einfo nos proporciona una lista de bases de datos disponibles para trabajar\n",
    "handle = Entrez.einfo()\n",
    "rec= Entrez.read(handle)\n",
    "handle.close()\n",
    "print(rec.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "289e678b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']\n"
     ]
    }
   ],
   "source": [
    "print(rec['DbList'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0478c6b0",
   "metadata": {},
   "source": [
    "## Acceso base de datos de nucleotidos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4c5b4b7",
   "metadata": {},
   "source": [
    "Las bases de datos más típicas son: Genbank, RefSeq, TPD y PDB.\n",
    "\n",
    "La secuencia que vamos a utilizar se corresponde con el gen del transportador de resistencia a la cloroquina(CRT) (código:**KM288867**) en el organismo de **plasmodium falciparum**. Con ello seleccionaremos que nos proporcione un total de 40 registros (retmax='40)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "75d5e2ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3081'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handle= Entrez.esearch(db='nucleotide', term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]', retmax='40')\n",
    "rec_list = Entrez.read(handle)\n",
    "handle.close()\n",
    "rec_list['Count'] #con ello sacamos el número total de registros que se logran con nuestra búsqueda, pero a nosotros solo \n",
    "#nos interesan un total de 40."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "af8e2878",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(rec_list['IdList']) # efectivamente nos saca 40 resgitros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "19cc317b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2587918588', '2507817686', '2507817684', '2507817682', '2507817680', '2507817678', '2507817676', '2507817674', '2507817672', '2507817670', '2507817668', '2507817666', '2507817664', '2507817662', '2507817660', '2507817658', '2507817656', '2507817654', '2507817652', '2507817650', '2507817648', '2507817646', '2507817644', '2507817642', '2507817640', '2507817638', '2507817636', '2507817634', '2507817632', '2507817630', '2507817628', '2507817626', '2507817624', '2507817622', '2507817620', '2507817618', '2507817616', '2507817614', '2507817612', '2507817610']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec_list['IdList']# id de los 40 registros aletorios solicitados"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c36b4b0e",
   "metadata": {},
   "source": [
    "Ya tenemos los ids de todos los registros, pero necesitamos que nos devuelva los registros de forma correcta."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43a94c05",
   "metadata": {},
   "source": [
    "Para ello utilizamos el comando EFetch: que descarga los resgitros al completo desde Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "362ef4fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#con este nuevo comando deberemos especificar ciertos carácteres nuevos, rettype= tipo de datos que se devuelven,gb=GenBank(aqui) \n",
    "#db= base de datos\n",
    "#OJO HAY QUE TENER CUIDADO CON ESTO , YA QUE PODEMOS LLEGAR A DESACARGAR GRANDES CANTIDADES DE DATOS\n",
    "id_list=rec_list['IdList']\n",
    "handle= Entrez.efetch(db='nucleotide', id=id_list, rettype='gb') # formato de genbank y necesitamos leerlo con parse de SeqIO\n",
    "recs= list(SeqIO.parse(handle,'gb'))\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b08a94d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('GGTGGAGGTTCTTGTCTTGGTAAATGTGCTCATGTGTTTAAACTTATTTTTAAA...AAA'), id='OR483864.1', name='OR483864', description='Plasmodium falciparum isolate PE-26 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672451.1', name='OQ672451', description='Plasmodium falciparum isolate ML_14 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672450.1', name='OQ672450', description='Plasmodium falciparum isolate ML_13 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672449.1', name='OQ672449', description='Plasmodium falciparum isolate ML_12 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672448.1', name='OQ672448', description='Plasmodium falciparum isolate ML_11 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672447.1', name='OQ672447', description='Plasmodium falciparum isolate ML_10 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672446.1', name='OQ672446', description='Plasmodium falciparum isolate ML_09 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672445.1', name='OQ672445', description='Plasmodium falciparum isolate ML_08 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672444.1', name='OQ672444', description='Plasmodium falciparum isolate ML_07 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672443.1', name='OQ672443', description='Plasmodium falciparum isolate ML_06 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672442.1', name='OQ672442', description='Plasmodium falciparum isolate ML_05 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672441.1', name='OQ672441', description='Plasmodium falciparum isolate ML_04 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672440.1', name='OQ672440', description='Plasmodium falciparum isolate ML_03 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672439.1', name='OQ672439', description='Plasmodium falciparum isolate ML_02 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672438.1', name='OQ672438', description='Plasmodium falciparum isolate ML_01 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672437.1', name='OQ672437', description='Plasmodium falciparum isolate CZS_20 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672436.1', name='OQ672436', description='Plasmodium falciparum isolate CZS_19 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672435.1', name='OQ672435', description='Plasmodium falciparum isolate CZS_18 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672434.1', name='OQ672434', description='Plasmodium falciparum isolate CZS_17 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672433.1', name='OQ672433', description='Plasmodium falciparum isolate CZS_16 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672432.1', name='OQ672432', description='Plasmodium falciparum isolate CZS_15 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672431.1', name='OQ672431', description='Plasmodium falciparum isolate CZS_14 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672430.1', name='OQ672430', description='Plasmodium falciparum isolate CZS_13 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672429.1', name='OQ672429', description='Plasmodium falciparum isolate CZS_12 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672428.1', name='OQ672428', description='Plasmodium falciparum isolate CZS_11 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672427.1', name='OQ672427', description='Plasmodium falciparum isolate CZS_10 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672426.1', name='OQ672426', description='Plasmodium falciparum isolate CZS_09 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672425.1', name='OQ672425', description='Plasmodium falciparum isolate CZS_08 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672424.1', name='OQ672424', description='Plasmodium falciparum isolate CZS_07 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672423.1', name='OQ672423', description='Plasmodium falciparum isolate CZS_06 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672422.1', name='OQ672422', description='Plasmodium falciparum isolate CZS_05 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672421.1', name='OQ672421', description='Plasmodium falciparum isolate CZS_04 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672420.1', name='OQ672420', description='Plasmodium falciparum isolate CZS_03 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672419.1', name='OQ672419', description='Plasmodium falciparum isolate CZS_02 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672418.1', name='OQ672418', description='Plasmodium falciparum isolate CZS_01 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672417.1', name='OQ672417', description='Plasmodium falciparum isolate GJ_03 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672416.1', name='OQ672416', description='Plasmodium falciparum isolate GJ_02 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672415.1', name='OQ672415', description='Plasmodium falciparum isolate GJ_01 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672414.1', name='OQ672414', description='Plasmodium falciparum isolate MAO_28 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATT...TTG'), id='OQ672413.1', name='OQ672413', description='Plasmodium falciparum isolate MAO_27 chloroquine resistance transporter (crt) gene, partial cds', dbxrefs=[])]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8a8b2b8",
   "metadata": {},
   "source": [
    "A continuación buscamos nuestro gen de interes en las secuencias sacadas, en caso de no encontrarse, deberíamos de volver a ejecutar el comando esearch() hasta que nos lo diera, 40 entre todos los 3081"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4281d676",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OQ672413\n",
      "Plasmodium falciparum isolate MAO_27 chloroquine resistance transporter (crt) gene, partial cds\n"
     ]
    }
   ],
   "source": [
    "for rec in recs:\n",
    "    if rec.name == 'KM288867': #buscamos dicho gen entre los 40 registros presentados\n",
    "        break\n",
    "print(rec.name)\n",
    "print(rec.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3abb4cd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATTTATATTTTAAGTATTATTTATTTAAGTGTATCTGTAATGAATACAATTTTTGCTAAAAGAACTTTAAACAAAATTGGTAACTATAGTTTTG'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(rec.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf4d3a52",
   "metadata": {},
   "source": [
    "**OJO=Creo que no es el mismo**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fad8cc2c",
   "metadata": {},
   "source": [
    "## Alineación de secuencia por pares, comando pairwise2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75dcaf29",
   "metadata": {},
   "source": [
    "Biopython presenta dos métodos de alineamiento, el pairwise2 y el PairwiseAligner, que tanto uno como otro pueden llegar a realizar un alineamiento  global o local"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4047fffc",
   "metadata": {},
   "source": [
    "alignmenttype: local o global\n",
    "\n",
    "X: los parámetros de los aciertos. pudiendo ser estos:\n",
    "\n",
    "    x-indica que no hay parámetros, parametros iguales se puntúan como 1, de lo contrario 0\n",
    "    \n",
    "    m-una puntuacion de acierto se da con carácteres iguales, de lo contrario una puntuación de fallo (mismatch)\n",
    "    \n",
    "    d-Un diccionario establece las puntuaciones para cada par de caracteres, match_dict\n",
    "    \n",
    "    c-una función devuelve la puntucación establecida, match_fn\n",
    "    \n",
    "    \n",
    "Y: los parámetros de las penalizaciones por gaps\n",
    "\n",
    "    x- no hay penalización por gap\n",
    "    \n",
    "    s-misma sanción de gap tanto si se da una extensión como si se elimina material, en ambas secuencias. Open y extend\n",
    "    \n",
    "    d-las secuencias presentan diferentes puntuaciones de apertura y extensión. OpenA, extendA,openB y extendB\n",
    "    \n",
    "    c-una función devulve las puntaciones a restar de los gaps, gap_A_fn y gap_B_fn."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0fb1c6a",
   "metadata": {},
   "source": [
    "**Alineamiento global**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "16e5ba3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Gabriel\\anaconda3\\lib\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca3512a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCGGT\n",
      "| | ||\n",
      "A-C-GT\n",
      "  Score=4\n",
      "\n",
      "ACCGGT\n",
      "||  ||\n",
      "AC--GT\n",
      "  Score=4\n",
      "\n",
      "ACCGGT\n",
      "| || |\n",
      "A-CG-T\n",
      "  Score=4\n",
      "\n",
      "ACCGGT\n",
      "|| | |\n",
      "AC-G-T\n",
      "  Score=4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#globalxx -match 1, missmatch 0 y no hay penalización por gap\n",
    "alineamientos= pairwise2.align.globalxx(\"ACCGGT\",\"ACGT\")\n",
    "for alineamiento in alineamientos:\n",
    "    print(pairwise2.format_alignment(*alineamiento))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e6ffb959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCGGT\n",
      "| | ||\n",
      "A-C-GT\n",
      "  Score=8\n",
      "\n",
      "ACCGGT\n",
      "||  ||\n",
      "AC--GT\n",
      "  Score=8\n",
      "\n",
      "ACCGGT\n",
      "| || |\n",
      "A-CG-T\n",
      "  Score=8\n",
      "\n",
      "ACCGGT\n",
      "|| | |\n",
      "AC-G-T\n",
      "  Score=8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#globalmx- match 2, missmatch -1 y no hay penalización de gap\n",
    "alineamientos= pairwise2.align.globalmx(\"ACCGGT\",\"ACGT\",match=2,mismatch=-1)\n",
    "for alineamiento in alineamientos:\n",
    "    print(pairwise2.format_alignment(*alineamiento))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4875930e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCGGT\n",
      "||  ||\n",
      "AC--GT\n",
      "  Score=1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#globalxs - match 1, mismatch 0, opening gap -2, extended gap -1\n",
    "alineamientos= pairwise2.align.globalxs(\"ACCGGT\",\"ACGT\", open=-2,extend=-1)\n",
    "for alineamiento in alineamientos:\n",
    "    print(pairwise2.format_alignment(*alineamiento))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65509e4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEVLA\n",
      " ||| \n",
      "-EVL-\n",
      "  Score=13\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#globaldx - matching/mismatching scores read from blosum62 matrix, no gap penalty\n",
    "from Bio.Align import substitution_matrices\n",
    "matriz = substitution_matrices.load(\"BLOSUM62\") # matriz de puntuación blosum62 para el alineamiento de secuencias\n",
    "alineamientos=pairwise2.align.globaldx(\"KEVLA\",\"EVL\", match_dict=matriz)\n",
    "for alineamiento in alineamientos:\n",
    "    print(pairwise2.format_alignment(*alineamiento))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1adbcc90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCCCCGT\n",
      "||     .\n",
      "AC-----G\n",
      "  Score=9.94528\n",
      "\n",
      "ACCCCCGT\n",
      "||.     \n",
      "ACG-----\n",
      "  Score=9.94528\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#globalmc - puntuación matches 5, mismatches -4, penalización por gap definida a partir de una función llamada funcion_gap\n",
    "\n",
    "from math import log\n",
    "def funcion_gap(x,y): # x marca la posicion del gap en la secuencia y luego y es la longitud del gap\n",
    "    if y == 0: #no hay gap\n",
    "        return 0\n",
    "    elif y == 1: # penalización de gap por abertura\n",
    "        return -2\n",
    "    return -(2 + y/4.0 + log(y)/2.0) #en cualquier otro caso, es decir con la extensión de un gap\n",
    "\n",
    "alineamientos = pairwise2.align.globalmc(\"ACCCCCGT\",\"ACG\", match=5, mismatch=4,gap_A_fn=funcion_gap,gap_B_fn=funcion_gap)\n",
    "for alineamiento in alineamientos:\n",
    "    print(pairwise2.format_alignment(*alineamiento))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ba4094",
   "metadata": {},
   "source": [
    "**Ejecución de alineamiento de secuencias con BLAST - Alineamiento local**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3b31aaa",
   "metadata": {},
   "source": [
    "Recordemos que BLAST(Basic Local Alignment Search Tool) es un método de alineamiento local para la comparación de dos subsecuencias de interés.\n",
    "\n",
    "Para su aplicación utilizaremos el módulo Bio.Blast, se puede utilizar tanto de forma local como de forma global. a continuación veremos su forma de aplicación online.(.qblast())\n",
    "\n",
    "la función qblast() proviene del módulo Bio.Blast.NCBIWWW, y presenta 3 parámetros obligatorios:\n",
    "   \n",
    "   -tipo de programa blast a utilizar: blastn, blastp, blastx, tblast y tblastx\n",
    "   \n",
    "   -Base de datos en la que buscar\n",
    "   \n",
    "   -cadena de caracteres que contiene la secuencia solicitada, pudiendo ser la propia secuencia o un identificador como el número GI.\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "65ce5d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "137afe95",
   "metadata": {},
   "source": [
    "**ejemplo**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ace4de5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#programa de blast utilizado blastn\n",
    "#nt = nucleotide\n",
    "resultado_handle=NCBIWWW.qblast(\"blastn\",\"nt\",\"\"\"ggatttcattttactatggatacatacacccggatagaaaatggattgtcggcgtcggctaggaaataagtccgtaaaaaaatttttaatatttaagggggggatttccctctct\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a7d80e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# blast genera una salida en formato XML y tenemos que de alguna forma analizarla gramaticalmente(parse it)\n",
    "#para ello importamos NCBIXML\n",
    "from Bio.Blast import NCBIXML\n",
    "registros_blast=NCBIXML.parse(resultado_handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ee7ac627",
   "metadata": {},
   "outputs": [],
   "source": [
    "registros_blast=list(registros_blast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "072acb7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Bio.Blast.Record.Blast at 0x21d71e5ae80>]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registros_blast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03df7f24",
   "metadata": {},
   "source": [
    "Vemos a continuación que se encuentra en ese registro de blast guardado en memoria mediante una variable:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f7bd2cc",
   "metadata": {},
   "source": [
    "OJO: BLAST E value,es el número de resultados esperados de calidad similar (puntuación) que podrían encontrarse por casualidad."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "59eb0060",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 similar sequences in Blast output\n"
     ]
    }
   ],
   "source": [
    "E_VALUE_THRESH = 0.00000000001\n",
    "count = 0\n",
    "for registro in registros_blast:\n",
    "    for alineamiento in registro.alignments:\n",
    "        for hsp in alineamiento.hsps:\n",
    "            if hsp.expect < E_VALUE_THRESH:\n",
    "                count+=1\n",
    "                print(\"****Alineamiento****\")\n",
    "                print (\"secuencia:\", alignment. title) \n",
    "                print (\"longitud\". alignment.length)\n",
    "                print (hsp.query [0:75] + \"...\")\n",
    "                print (hsp.match [0:75] + \"...\")\n",
    "                print (hsp.sbjct [0:75] + \"...\")\n",
    "                print()\n",
    "\n",
    "print(f\"There are {count} similar sequences in Blast output\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5fc6895",
   "metadata": {},
   "source": [
    "## Creación de objetos Motif (motivos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbcca97e",
   "metadata": {},
   "source": [
    "Un motivo se trata de un patrón de secuencia de nucleótidos o amino ácidos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e14badbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e124c76a",
   "metadata": {},
   "source": [
    "Creamos un motivo de secuencia de DNA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "aea34a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "instancias= [\n",
    "    Seq(\"TACAA\"),\n",
    "    Seq(\"TACGC\"),\n",
    "    Seq(\"TACAC\"),\n",
    "    Seq(\"GACTA\"),\n",
    "    Seq(\"TACGA\"),\n",
    "    Seq(\"ATTGG\"),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7249fe26",
   "metadata": {},
   "outputs": [],
   "source": [
    "m = motifs.create(instancias)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fb5ef82b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACAA\n",
      "TACGC\n",
      "TACAC\n",
      "GACTA\n",
      "TACGA\n",
      "ATTGG\n"
     ]
    }
   ],
   "source": [
    "print(m)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5646bd0b",
   "metadata": {},
   "source": [
    "las intancias que creemos son guardadas en la variable en el atributo **.instances**. De forma que:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "33034e2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Gabriel\\anaconda3\\lib\\site-packages\\Bio\\motifs\\__init__.py:478: BiopythonDeprecationWarning: The instances attribute has been deprecated. Instead of mymotif.instances, please use mymotif.alignment.sequences.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[Seq('TACAA'),\n",
       " Seq('TACGC'),\n",
       " Seq('TACAC'),\n",
       " Seq('GACTA'),\n",
       " Seq('TACGA'),\n",
       " Seq('ATTGG')]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e3b3983c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Seq('TACAA'),\n",
       " Seq('TACGC'),\n",
       " Seq('TACAC'),\n",
       " Seq('GACTA'),\n",
       " Seq('TACGA'),\n",
       " Seq('ATTGG')]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#o también\n",
    "m.alignment.sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38570cab",
   "metadata": {},
   "source": [
    "Tambien existe un atributo llamado **.counts** que imprime una matriz con el número de veces que aparece una bases nitrogenada en las diferentes secuencias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e58ff1fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        0      1      2      3      4\n",
      "A:   1.00   5.00   0.00   2.00   3.00\n",
      "C:   0.00   0.00   5.00   0.00   2.00\n",
      "G:   1.00   0.00   0.00   3.00   1.00\n",
      "T:   4.00   1.00   1.00   1.00   0.00\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(m.counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e3b01b6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1.0, 5.0, 0.0, 2.0, 3.0]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#podemos accceder a la fila de un nucleótido en específico:\n",
    "m.counts[\"A\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f1dbcef7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 2.0, 'C': 0.0, 'G': 3.0, 'T': 1.0}"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#o incluso podemos llegar a acceder a una secuencia determinada de interés(OJO: con la posición como si fuera el índice de la lista)\n",
    "m.counts[:,3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "151b5705",
   "metadata": {},
   "source": [
    "También podemos obtener la secuencia consenso y la secuencia anticonsenso:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e36bf7ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGA\n",
      "CCACT\n"
     ]
    }
   ],
   "source": [
    "#secuencia consenso: secuencia que se establece \n",
    "print(m.consensus)\n",
    "\n",
    "#secuencia anticonsenso\n",
    "print(m.anticonsensus)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17256a6c",
   "metadata": {},
   "source": [
    "Dado que el consenso es ambiguo, para ser más específicos hacemos uso del comando degenerate_consensus, que siguiendo los codigos de ambiguedad de la UIPAC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ef79d876",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACRM')"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.degenerate_consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d16b77e0",
   "metadata": {},
   "source": [
    "Cada letra distinta a las iniciales de las bases, representa una posibilidad de bases distinta."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec79134",
   "metadata": {},
   "source": [
    "## Graficado de longitud de secuencias y contenido de GC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ef7afba",
   "metadata": {},
   "source": [
    "**Get data**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01d636b7",
   "metadata": {},
   "source": [
    "**Graficado de longitud de secuencias**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5172621b",
   "metadata": {},
   "source": [
    "**Graficado de contenido de GC**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ecb3825",
   "metadata": {},
   "source": [
    "## Lecturas rápidas de archivos FASTA/FASTQ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ae0bea",
   "metadata": {},
   "source": [
    "## Ordenación de los archivos FASTA/FASTQ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab01236",
   "metadata": {},
   "source": [
    "## Filtrado de archivos FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c80244",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
